---
title: "Untitled"
output: html_document
date: "2024-07-16"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
df0 <- fread("C:/docs/aou/sample_3-7-8.tsv") %>% as.data.frame()
df1 <- data.frame(`membership:sample_3-7-8_set_id` = paste0("samtools_", rep(c(1:4), length.out = nrow(df0))),
                  `sample_3-7-8` = df0$`entity:sample_3-7-8_id`, 
                  check.names = FALSE)
write.table(df1, "C:/docs/aou/samtools_memb.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```

```{r}
df0 <- fread("C:/docs/aou/sample_3-7-8.tsv") %>% as.data.frame()
df0$sample <- df0$cram %>% bname %>% str_remove(".cram$")
df1 <- fread("C:/docs/aou/sample_3-7-8_set_entity.tsv") %>% as.data.frame()

df_r1 <- data.frame(str_r1=readLines("C:/docs/locityper/terra/R1.fastq.gz.files"))
df_r1$r1_path <- lapply(df_r1$str_r1, function(x){
  x %>% str_extract("gs://.*1.fastq.gz") 
}) %>% unlist
df_r1$sample <- df_r1$r1_path %>% bname %>% str_remove("_1.fastq.gz")
df_r1$str_r1 <- NULL


df_r2 <- data.frame(str_r2=readLines("C:/docs/locityper/terra/R2.fastq.gz.files"))
df_r2$r2_path <- lapply(df_r2$str_r2, function(x){
  x %>% str_extract("gs://.*2.fastq.gz") 
}) %>% unlist
df_r2$sample <- df_r2$r2_path %>% bname %>% str_remove("_2.fastq.gz")
df_r2$str_r2 <- NULL

df2 <- merge(df0, df_r1, all.x = TRUE) %>% merge(df_r2, all.x=TRUE)
df2[is.na(df2)] <- ""

df3 <- data.frame(`membership:sample_3-7-8_set_id`  = "samtools_more_mem", 
                  `sample_3-7-8` = df2$`entity:sample_3-7-8_id`[df2$r1_path==""],
                  check.names = FALSE)

write(df3, "C:/docs/aou/memb_more_memory.memb")
write(df2[, c(2,1,3:9)], "C:/docs/aou/sample_20240719.tsv")

df4 <- data.frame(`membership:sample_3-7-8_set_id`  = "locityper_batch_01", 
                  `sample_3-7-8` = df2$`entity:sample_3-7-8_id`[df2$r1_path!=""],
                  check.names = FALSE)
write(df4, "C:/docs/aou/memb_first_batch.memb")
```


```{r}
df0 <- fread("C:/docs/aou/hifi-sample-t2t.tsv") %>% as.data.frame()
N_GRPS <- 3

df_memb <- data.frame(`membership:hifi-sample-t2t_set_id` = paste0("curve_callable_", rep(1:N_GRPS, nrow(df0))[seq_down(df0)]),
                      `hifi-sample-t2t` = df0$`entity:hifi-sample-t2t_id`,
                      check.names=FALSE)

write.table(df_memb, "C:/docs/aou/memb_curve_callable.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```

```{r}
h1 <- readLines("C:/docs/aou/callable_h1.fofn")
h2 <- readLines("C:/docs/aou/callable_h2.fofn")

h1_sample <- h1 %>% bname %>% str_remove("")


```



```{r}
df0 <- fread("C:/docs/aou/create_table/hifi-sample-t2t_set/hifi-sample-t2t_set_entity.tsv") %>% as.data.frame()
df0 <- df0[grepl("curve_callable_.$", df0$`entity:hifi-sample-t2t_set_id`), ]

df1 <- fread("C:/docs/aou/create_table/hifi-sample-t2t_set/hifi-sample-t2t_set_membership.tsv") %>% as.data.frame()

samples <- df1$`hifi-sample-t2t` %>% u

h1_files <- df0$output_h1_files %>% str_remove_all("\"|\\[|\\]") %>% str_split(",") %>% unlist
h2_files <- df0$output_h2_files %>% str_remove_all("\"|\\[|\\]") %>% str_split(",") %>% unlist

df_memb <- data.frame(`membership:hifi-sample-t2t_set_id` = "create_table_1", 
                      `hifi-sample-t2t` = samples,
                      check.names = FALSE)
write.table(df_memb, "C:/docs/aou/create_table/df_memb.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)

df_entity <- data.frame(`entity:hifi-sample-t2t_set_id` = "create_table_1", 
                        output_h1_files  = paste0("[\"", paste0(h1_files, collapse = "\",\""), "\"]"), 
                        output_h2_files  = paste0("[\"", paste0(h2_files, collapse = "\",\""), "\"]"), 
                        check.names = FALSE)
write.table(df_entity, "C:/docs/aou/create_table/df_ent_new.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)

write.table(df_memb$`hifi-sample-t2t`, "C:/docs/aou/create_table/samples_all.txt", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)

write.table(df_memb$`hifi-sample-t2t`, "C:/docs/aou/create_table/samples_all.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)
```



```{r}
df0 <- fread("C:/docs/aou/extra_samples.txt", header = FALSE) %>% as.data.frame()
df1 <- fread("C:/docs/aou/extra_samples_paths.txt", header = FALSE) %>% as.data.frame()

df1$sample <- df1$V1 %>% lapply(function(x){
  temp1 <- str_split(x, "/") %>% unlist
  return(temp1[9])
}) %>% unlist

df1 <- df1[df1$sample %in% df0$V1, ]
colnames(df1)  <- c("path", "sample")
df1_extra <- data.frame(path = c("/net/eichler/vol28/projects/long_read_archive/nobackups/pop/NA19240/alignments/PacBio_HiFi/CHM13_v2.0/GM19240.minimap2.bam", "/net/eichler/vol28/projects/long_read_archive/nobackups/pop/HG002/alignments/PacBio_HiFi/CHM13_v2.0/NA24385.minimap2.bam"), 
                        sample = c("NA19240", "HG002"))

df2 <- rbind(df1, df1_extra)

json_samples <- paste0("[\"", paste0(df2$sample, collapse = "\",\""), "\"]")
json_bams <- paste0("[\"", paste0(df2$path, collapse = "\",\""), "\"]")

script <- "{\n"
script <- c(script, "  \"discovery_curve.samtools_n_cpu\": \"4\",\n")
script <- c(script, "\"discovery_curve.callable_bed\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/collapsed.bed.gz\",\n")
script <- c(script, "\"discovery_curve.docker\": \"iwongbroadinstitute/disccurve:0.2\",\n")
script <- c(script, "\"discovery_curve.samtools_mem_gb\": \"8\",\n")
script <- c(script, "\"discovery_curve.sample_file\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/empty.txt\",\n")
script <- c(script, "\"discovery_curve.truvari_vcf\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/empty.txt\",\n")
script <- c(script, paste0("\"discovery_curve.input_samples\": ", json_samples, ",\n"))
script <- c(script, paste0("\"discovery_curve.input_bams\": ", json_bams, "\n"))
script <- c(script, "}")

write.table(script, "C:/docs/aou/new_callable.json", sep = " ", col.names = FALSE, row.names = FALSE, quote = FALSE)

script <- c(script, "")


write.table(df2[, c(2,1)], "C:/docs/aou/extra_samples_snakeman.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```
/net/eichler/vol28/projects/long_read_archive/nobackups/pop/NA19240/alignments/PacBio_HiFi/CHM13_v2.0/GM19240.minimap2.bam
/net/eichler/vol28/projects/long_read_archive/nobackups/pop/HG002/alignments/PacBio_HiFi/CHM13_v2.0/NA24385.minimap2.bam



```{r}
df0 <- fread("C:/docs/aou/depths.txt", header = FALSE) %>% as.data.frame()
df0$sample <- df0$V1 %>% bname %>% str_remove(".depth.txt$") %>% unlist

json_samples <- paste0("[\"", paste0(df0$sample, collapse = "\",\""), "\"]")
json_bams <- paste0("[\"", paste0(df0$V1, collapse = "\",\""), "\"]")


script <- "{\n"
script <- c(script, "  \"discovery_curve.samtools_n_cpu\": \"4\",\n")
script <- c(script, "\"discovery_curve.callable_bed\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/collapsed.bed.gz\",\n")
script <- c(script, "\"discovery_curve.docker\": \"iwongbroadinstitute/disccurve:0.2\",\n")
script <- c(script, "\"discovery_curve.samtools_mem_gb\": \"8\",\n")
script <- c(script, "\"discovery_curve.sample_file\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/empty.txt\",\n")
script <- c(script, "\"discovery_curve.truvari_vcf\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/empty.txt\",\n")
script <- c(script, paste0("\"discovery_curve.input_samples\": ", json_samples, ",\n"))
script <- c(script, paste0("\"discovery_curve.input_bams\": ", json_bams, "\n"))
script <- c(script, "}")

write.table(script, "C:/docs/aou/new_callable_skip_samtools.json", sep = " ", col.names = FALSE, row.names = FALSE, quote = FALSE)

```

