---
title: "Untitled"
output: html_document
date: "2024-07-16"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
df0 <- fread("C:/docs/aou/sample_3-7-8.tsv") %>% as.data.frame()
df1 <- data.frame(`membership:sample_3-7-8_set_id` = paste0("samtools_", rep(c(1:4), length.out = nrow(df0))),
                  `sample_3-7-8` = df0$`entity:sample_3-7-8_id`, 
                  check.names = FALSE)
write.table(df1, "C:/docs/aou/samtools_memb.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```

```{r}
df0 <- fread("C:/docs/aou/sample_3-7-8.tsv") %>% as.data.frame()
df0$sample <- df0$cram %>% bname %>% str_remove(".cram$")
df1 <- fread("C:/docs/aou/sample_3-7-8_set_entity.tsv") %>% as.data.frame()

df_r1 <- data.frame(str_r1=readLines("C:/docs/locityper/terra/R1.fastq.gz.files"))
df_r1$r1_path <- lapply(df_r1$str_r1, function(x){
  x %>% str_extract("gs://.*1.fastq.gz") 
}) %>% unlist
df_r1$sample <- df_r1$r1_path %>% bname %>% str_remove("_1.fastq.gz")
df_r1$str_r1 <- NULL


df_r2 <- data.frame(str_r2=readLines("C:/docs/locityper/terra/R2.fastq.gz.files"))
df_r2$r2_path <- lapply(df_r2$str_r2, function(x){
  x %>% str_extract("gs://.*2.fastq.gz") 
}) %>% unlist
df_r2$sample <- df_r2$r2_path %>% bname %>% str_remove("_2.fastq.gz")
df_r2$str_r2 <- NULL

df2 <- merge(df0, df_r1, all.x = TRUE) %>% merge(df_r2, all.x=TRUE)
df2[is.na(df2)] <- ""

df3 <- data.frame(`membership:sample_3-7-8_set_id`  = "samtools_more_mem", 
                  `sample_3-7-8` = df2$`entity:sample_3-7-8_id`[df2$r1_path==""],
                  check.names = FALSE)

write(df3, "C:/docs/aou/memb_more_memory.memb")
write(df2[, c(2,1,3:9)], "C:/docs/aou/sample_20240719.tsv")

df4 <- data.frame(`membership:sample_3-7-8_set_id`  = "locityper_batch_01", 
                  `sample_3-7-8` = df2$`entity:sample_3-7-8_id`[df2$r1_path!=""],
                  check.names = FALSE)
write(df4, "C:/docs/aou/memb_first_batch.memb")
```


```{r}
df0 <- fread("C:/docs/aou/hifi-sample-t2t.tsv") %>% as.data.frame()
N_GRPS <- 3

df_memb <- data.frame(`membership:hifi-sample-t2t_set_id` = paste0("curve_callable_", rep(1:N_GRPS, nrow(df0))[seq_down(df0)]),
                      `hifi-sample-t2t` = df0$`entity:hifi-sample-t2t_id`,
                      check.names=FALSE)

write.table(df_memb, "C:/docs/aou/memb_curve_callable.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```

```{r}
h1 <- readLines("C:/docs/aou/callable_h1.fofn")
h2 <- readLines("C:/docs/aou/callable_h2.fofn")

h1_sample <- h1 %>% bname %>% str_remove("")


```



```{r}
df0 <- fread("C:/docs/aou/create_table/hifi-sample-t2t_set/hifi-sample-t2t_set_entity.tsv") %>% as.data.frame()
df0 <- df0[grepl("curve_callable_.$", df0$`entity:hifi-sample-t2t_set_id`), ]

df1 <- fread("C:/docs/aou/create_table/hifi-sample-t2t_set/hifi-sample-t2t_set_membership.tsv") %>% as.data.frame()

samples <- df1$`hifi-sample-t2t` %>% u

h1_files <- df0$output_h1_files %>% str_remove_all("\"|\\[|\\]") %>% str_split(",") %>% unlist
h2_files <- df0$output_h2_files %>% str_remove_all("\"|\\[|\\]") %>% str_split(",") %>% unlist

df_memb <- data.frame(`membership:hifi-sample-t2t_set_id` = "create_table_1", 
                      `hifi-sample-t2t` = samples,
                      check.names = FALSE)
write.table(df_memb, "C:/docs/aou/create_table/df_memb.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)

df_entity <- data.frame(`entity:hifi-sample-t2t_set_id` = "create_table_1", 
                        output_h1_files  = paste0("[\"", paste0(h1_files, collapse = "\",\""), "\"]"), 
                        output_h2_files  = paste0("[\"", paste0(h2_files, collapse = "\",\""), "\"]"), 
                        check.names = FALSE)
write.table(df_entity, "C:/docs/aou/create_table/df_ent_new.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)

write.table(df_memb$`hifi-sample-t2t`, "C:/docs/aou/create_table/samples_all.txt", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)

write.table(df_memb$`hifi-sample-t2t`, "C:/docs/aou/create_table/samples_all.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)
```



```{r}
df0 <- fread("C:/docs/aou/extra_samples.txt", header = FALSE) %>% as.data.frame()
df1 <- fread("C:/docs/aou/extra_samples_paths.txt", header = FALSE) %>% as.data.frame()

df1$sample <- df1$V1 %>% lapply(function(x){
  temp1 <- str_split(x, "/") %>% unlist
  return(temp1[9])
}) %>% unlist

df1 <- df1[df1$sample %in% df0$V1, ]
colnames(df1)  <- c("path", "sample")
df1_extra <- data.frame(path = c("/net/eichler/vol28/projects/long_read_archive/nobackups/pop/NA19240/alignments/PacBio_HiFi/CHM13_v2.0/GM19240.minimap2.bam", "/net/eichler/vol28/projects/long_read_archive/nobackups/pop/HG002/alignments/PacBio_HiFi/CHM13_v2.0/NA24385.minimap2.bam"), 
                        sample = c("NA19240", "HG002"))

df2 <- rbind(df1, df1_extra)

json_samples <- paste0("[\"", paste0(df2$sample, collapse = "\",\""), "\"]")
json_bams <- paste0("[\"", paste0(df2$path, collapse = "\",\""), "\"]")

script <- "{\n"
script <- c(script, "  \"discovery_curve.samtools_n_cpu\": \"4\",\n")
script <- c(script, "\"discovery_curve.callable_bed\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/collapsed.bed.gz\",\n")
script <- c(script, "\"discovery_curve.docker\": \"iwongbroadinstitute/disccurve:0.2\",\n")
script <- c(script, "\"discovery_curve.samtools_mem_gb\": \"8\",\n")
script <- c(script, "\"discovery_curve.sample_file\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/empty.txt\",\n")
script <- c(script, "\"discovery_curve.truvari_vcf\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/empty.txt\",\n")
script <- c(script, paste0("\"discovery_curve.input_samples\": ", json_samples, ",\n"))
script <- c(script, paste0("\"discovery_curve.input_bams\": ", json_bams, "\n"))
script <- c(script, "}")

write.table(script, "C:/docs/aou/new_callable.json", sep = " ", col.names = FALSE, row.names = FALSE, quote = FALSE)

script <- c(script, "")


write.table(df2[, c(2,1)], "C:/docs/aou/extra_samples_snakeman.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```
/net/eichler/vol28/projects/long_read_archive/nobackups/pop/NA19240/alignments/PacBio_HiFi/CHM13_v2.0/GM19240.minimap2.bam
/net/eichler/vol28/projects/long_read_archive/nobackups/pop/HG002/alignments/PacBio_HiFi/CHM13_v2.0/NA24385.minimap2.bam



```{r}
df0 <- fread("C:/docs/aou/depths.txt", header = FALSE) %>% as.data.frame()
df0$sample <- df0$V1 %>% bname %>% str_remove(".depth.txt$") %>% unlist

json_samples <- paste0("[\"", paste0(df0$sample, collapse = "\",\""), "\"]")
json_bams <- paste0("[\"", paste0(df0$V1, collapse = "\",\""), "\"]")


script <- "{\n"
script <- c(script, "  \"discovery_curve.samtools_n_cpu\": \"4\",\n")
script <- c(script, "\"discovery_curve.callable_bed\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/collapsed.bed.gz\",\n")
script <- c(script, "\"discovery_curve.docker\": \"iwongbroadinstitute/disccurve:0.2\",\n")
script <- c(script, "\"discovery_curve.samtools_mem_gb\": \"8\",\n")
script <- c(script, "\"discovery_curve.sample_file\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/empty.txt\",\n")
script <- c(script, "\"discovery_curve.truvari_vcf\": \"/net/eichler/vol28/home/iwong1/nobackups/disc_curve/test_03/empty.txt\",\n")
script <- c(script, paste0("\"discovery_curve.input_samples\": ", json_samples, ",\n"))
script <- c(script, paste0("\"discovery_curve.input_bams\": ", json_bams, "\n"))
script <- c(script, "}")

write.table(script, "C:/docs/aou/new_callable_skip_samtools.json", sep = " ", col.names = FALSE, row.names = FALSE, quote = FALSE)

```

```{r}
files7 <- list.files("C:/docs/aou/counts_rd7/", pattern="counts$", full.names = TRUE)
df7 <- lapply(files7, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  rval <- data.frame(sample = df_temp$SAMPLE_ORDER,
                     poly_7 = df_temp$POLY, 
                     chr = x %>% str_extract("(?<=subset.).*(?=_out.counts)"))
  return(rval)
}) %>% do.call(rbind, .)

files3 <- list.files("C:/docs/aou/counts_rd3/", pattern="counts$", full.names = TRUE)
df3 <- lapply(files3, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  rval <- data.frame(sample = df_temp$SAMPLE_ORDER,
                     poly_3 = df_temp$POLY, 
                     chr = x %>% str_extract("(?<=subset.).*(?=_out.counts)"))
  return(rval)
}) %>% do.call(rbind, .)

df00 <- lapply(seq_down(df7), function(x){
  data.frame(sample = df7$sample[x],
             chr = df7$chr[x],
             poly_7 = df7$poly_7[x],
             poly_3 = df3$poly_3[df3$sample==df7$sample[x] & df3$chr==df7$chr[x]])
}) %>% do.call(rbind, .)

temp1 <- table(df00$poly_7 == df00$poly_3)
temp1

df00$diffs <- df00$poly_3 - df00$poly_7

ggplot(df00, aes(x=diffs)) + geom_histogram() + 
  scale_x_continuous(breaks=0:15) + 
  xlab("N polymorphic difference: 3 vs 7 depth")
  

```



```{r}
df_temp1 <- fread("C:/docs/aou/counts_rd3/subset.chr1_out.counts") %>% as.data.frame()
df_temp2 <- fread("C:/docs/aou/counts_rd7/subset.chr1_out.counts") %>% as.data.frame()
```



```{r}
df0 <- fread("C:/docs/aou/kiran_outputs/sr_sample.tsv") %>% as.data.frame()
df0$script <- paste0("echo ", df0$`entity:sr_sample_id`, "; ", 
                     # "mkdir -p ", df0$`entity:sr_sample_id`, "; ",
                     "mkdir -p ", df0$`entity:sr_sample_id`, "/panel_results; ",
                     # "gsutil cp ", df0$panel_all_results_tar_gz, " "  , df0$`entity:sr_sample_id`, "/panel_all_results/ ; ",
                     "gsutil cp ", df0$panel_all_summary_csv, " "  , df0$`entity:sr_sample_id`, "/panel_results/ ; ",
                     # "gsutil cp ", df0$panel_results_tar_gz, " "  , df0$`entity:sr_sample_id`, "/ ; "
)
splits <- split(df0$script, cut(seq_along(df0$script), breaks = 4, labels = FALSE))

write.table(splits[[1]], "C:/docs/aou/kiran_outputs/download_script_1.sh", sep=" ", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(splits[[2]], "C:/docs/aou/kiran_outputs/download_script_2.sh", sep=" ", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(splits[[3]], "C:/docs/aou/kiran_outputs/download_script_3.sh", sep=" ", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(splits[[4]], "C:/docs/aou/kiran_outputs/download_script_4.sh", sep=" ", col.names = FALSE, row.names = FALSE, quote=FALSE)

write.table(df0$script, "C:/docs/aou/kiran_outputs/download_script.sh", sep=" ", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r}
df0 <- fread("C:/docs/aou/Sample_sv.pheWAS.txt") %>% as.data.frame()
df0$svs_split <- df0$SV_IDs %>% str_split(",")
temp1 <- df0$svs_split %>% lapply(length) %>% unlist
loci_max <- max(temp1)
loci_min <- min(temp1)
loci_mean <- mean(temp1)
costs <- c(206.82, 134, 432, 175, 136, 158, 174, 170, 182, 193, 166, 149, 177, 147, 193, 169, 256, 264, 387, 404, 66 , 67)
nsamp <- c(256   , 500, 500, 500, 500, 500, 500, 500, 500, 500, 500, 500, 500, 500, 500, 500, 500, 499, 500, 500, 150, 99)
sum_costs <- sum(costs)
sum_samps <- sum(nsamp)
my_rate <- sum_costs / (sum_samps * loci_max)
my_rate2 <- sum_costs / (sum_samps * loci_mean)
my_rate3 <- sum_costs / (sum_samps * loci_min)

print(c(my_rate, my_rate2, my_rate3))
```

```{r}
df_crams <- data.frame(count = c(71, 96, 104, 111, 120, 130, 164, 198, 243, 263),
                       time  = c("Aug 26 11:15:45", "Aug 26 11:45:06", "Aug 26 11:55:38", "Aug 26 12:04:10", "Aug 26 12:14:52", "Aug 26 12:27:28", "Aug 26 13:11:19", "Aug 26 13:50:47", "Aug 26 14:47:33", "Aug 26 15:51:44") %>% as.POSIXct(format="%b %d %H:%M:%OS", tz="UTC"))

df_crams <- data.frame(count = c(2, 46, 91, 151, 175, 213, 301),
                       time  = c("Aug 29 11:10:00", "Aug 29 12:06:45", "Aug 29 13:02:36", "Aug 29 14:20:00", "Aug 29 14:48:30", "Aug 29 15:33:46", "Aug 29 17:18:24") %>% as.POSIXct(format="%b %d %H:%M:%OS", tz="UTC"))

df_crams <- data.frame(count = c(1, 20, 39, 70, 119, 151, 206),
                       time  = c("Sep  3 10:25:06", "Sep  3 10:47:41", "Sep  3 11:09:53", "Sep  3 11:46:55", "Sep  3 12:48:05", "Sep  3 13:26:35", "Sep  3 14:30:18") %>% as.POSIXct(format="%b %d %H:%M:%OS", tz="UTC"))


df_crams <- data.frame(count = c(13, 33, 43, 96, 160, 250, 341),
                       time  = c("Sep  8 09:56:36", "Sep  8 10:20:26", "Sep  8 10:32:39", "Sep  8 11:35:55", "Sep 8 12:50:18", "Sep  8 14:37:41", "Sep  8 16:26:29") %>% as.POSIXct(format="%b %d %H:%M:%OS", tz="UTC"))


n_crams <- df_crams$count[nrow(df_crams)] - df_crams$count[1]
n_time <- df_crams$time[nrow(df_crams)] - df_crams$time[1]
my_rate <- n_time/n_crams
total_time <- my_rate * 1040

final_time <- df_crams$time[1] + total_time 

final_time

df_crams$final_time <- lapply(seq_down(df_crams), function(x){
  if(x==1){return(final_time)}
  n_crams <- df_crams$count[x] - df_crams$count[1]
  n_time <- df_crams$time[x] - df_crams$time[1]
  my_rate <- n_time/n_crams
  total_time <- my_rate * 1040
  return(df_crams$time[1] + total_time )
}) %>% unlist %>% as.POSIXct(tz="UTC")

df_crams
```

as.POSIXct(df00$ts, format="%Y-%m-%dT%H:%M:%OSZ", tz="UTC")

```{r}
rolls <- data.frame(d1 = rep(1:6, 6),
                    d2 = rep(1:6, each=6))
rolls$total <- rolls$d1 + rolls$d2

roll <- function(){
  return(sum(sample(1:6, 2, TRUE)))
}
N <- 100000
charges <- data.frame(charge_distance = 1:12)
charges$roll <- lapply(charges$charge_distance, function(x){
  
  
  
})
```



```{r}
df00 <- fread("C:/docs/aou/sv_intersection/HA_GRCh38_SVs_intersection_CDS.tsv") %>% as.data.frame()
df00$genes_list <- df00$`GENE(S)` %>% str_split(",")
df_acmg <- fread("C:/docs/aou/sv_intersection/TargetedGeneLoci - ACMG_CHM13_Repeats.tsv") %>% as.data.frame()
df_cmrg <- fread("C:/docs/aou/sv_intersection/TargetedGeneLoci - CMRG_CHM13_Repeats.tsv") %>% as.data.frame()
df00$ACMG <- lapply(df00$genes_list, function(x){
  any(df_acmg$gene %in% x)
}) %>% unlist
df00$CMRG <- lapply(df00$genes_list, function(x){
  any(df_cmrg$gene %in% x)
}) %>% unlist

df00$`ACMG/CMRG` <- lapply(seq_down(df00), function(x){
  if(df00$ACMG[x] & df00$CMRG[x]){
    return("ACMG, CMRG")
  }
  if(df00$ACMG[x]){
    return("ACMG")
  }
  if(df00$CMRG[x]){
    return("CMRG")
  }
  return("-")
}) %>% unlist

df00$genes_list <- NULL
write.table(df00, "C:/docs/aou/sv_intersection/HA_GRCh38_SVs_intersection_CDS_iw.tsv", col.names = TRUE, row.names = FALSE, quote=FALSE, sep="\t")
```

```{r}
N <- 5000
big_N <- 50000
df00 <- fread("C:/docs/aou/locityper/full_cram_manifest_424207.tsv") %>% as.data.frame()
colnames(df00)[1] <- "entity:phase_ii_id"

df00$super_batch <- paste0("locityper_super_batch_", rep(seq(1,ceiling(nrow(df00)/big_N)),each=big_N)[1:nrow(df00)])

temp1 <- lapply(df00$super_batch %>% u, function(batch){
  super_batch_name <- str_remove(batch, "locityper_super_batch_")
  df_temp <- df00[df00$super_batch == batch, ]
  colnames(df_temp)[1] <- paste0("entity:phase_ii_sb", super_batch_name ,"_id")
  
  df_memb <- data.frame(memb = paste0("locityper_sb", super_batch_name, "_", rep(seq(1,ceiling(nrow(df_temp)/N)),each=N)[1:nrow(df_temp)]), 
                      phase_ii = df_temp[,1],
                      check.names = FALSE)
  colnames(df_memb) <- c(paste0("membership:phase_ii_sb", super_batch_name,"_set_id"), paste0("phase_ii_sb", super_batch_name))
  
  write.table(df_temp, paste0("C:/docs/aou/locityper/mans/phase-ii_", super_batch_name,"_sample.tsv"), sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
  write.table(df_memb, paste0("C:/docs/aou/locityper/mans/phase-ii_", super_batch_name,"_memb.tsv")  , sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
  
  
})

```



```{r}
df00 <- fread("C:/docs/aou/locityper/full_cram_manifest_424207.tsv") %>% as.data.frame()
df01 <- fread("C:/docs/aou/locityper/v1pt1b_aou12k_samples_allofus-drc-wgs-lr-prod_AoU_DRC_WGS_LongReads_ShortReads.tsv") %>% as.data.frame()
df01 <- df01[, c(1,7,8)]
df02 <- fread("C:/docs/aou/locityper/beri_sample_grch38_allofus-drc-wgs-lr-prod_AoU_DRC_WGS_LongReads_ShortReads.tsv") %>% as.data.frame()
df02 <- df02[, c(1,2,4,5,11,18)]
df04 <- fread("C:/docs/aou/locityper/ehr/SR_samples_both_GT_and_condition.txt", header=FALSE) %>% as.data.frame()
df03 <- fread("C:/docs/aou/locityper/ehr/SR_combined_manifest.massaged.tsv") %>% as.data.frame()

df03 <- df03[df03$research_id %in% df04$V1, ]
df03 <- df03[match(u(df03$biobankidsampleid), df03$biobankidsampleid), ]

all_crams <- u(c(df00$cram, df01$cram, df02$cram, df02$input_cram))

df03 <- df03[1:1000, ]
all_crams <- all_crams[1:1000]

#ran on cluster

df03$matching_cram <- lapply(seq_down(df03), function(x){
  
  match_1 <- all_crams[grepl(df03$biobankidsampleid[x], all_crams)]
  if(length(match_1) == 1){
    return(match_1)
  } 
  
  match_2 <-  all_crams[grepl(df03$biobank_id[x], all_crams)]
  if(length(match_2) == 1){
    return(match_2)
  }
  
  match_3 <-  all_crams[grepl(df03$sample_id[x], all_crams)]
  if(length(match_3) == 1){
    return(match_3)
  }
  
  match_4 <-  all_crams[grepl(df03$chipwellbarcode[x], all_crams)]
  if(length(match_4) == 1){
    return(match_4)
  }
  
  if(length(match_1) > 1){
    return(match_1)
  }
  
  if(length(match_2) > 1){
    return(match_2)
  }
  
  if(length(match_3) > 1){
    return(match_3)
  }
  
  if(length(match_3) > 1){
    return(match_3)
  }
  
  if(length(match_4) > 1){
    return(match_4)
  }
  
  return("NO_MATCH")
})

load("C:/docs/aou/locityper/ehr/df03_clust.rda")

n_matches <- df03$matching_cram %>% lapply(function(x){length(x)}) %>% unlist 
df05 <- df03[n_matches==1, ]
df05$matching_cram <- df05$matching_cram %>% unlist
df05 <- df05[df05$matching_cram != "NO_MATCH", ]
df05$vcf_path <- NULL
df05$crai <- paste0(df05$matching_cram, ".crai")


N <- 5000
big_N <- 50000

df00 <- data.frame(`entity:ehr_id` = df05$research_id, 
                   cram = df05$matching_cram, 
                   crai = df05$crai, 
                   chipwellbarcode = df05$chipwellbarcode,
                   biobank_id = df05$biobank_id, 
                   sample_id = df05$sample_id, 
                   biobankidsampleid = df05$biobankidsampleid,
                   check.names = FALSE)

df00$super_batch <- paste0("locityper_super_batch_", rep(seq(1,ceiling(nrow(df00)/big_N)),each=big_N)[1:nrow(df00)])

temp1 <- lapply(df00$super_batch %>% u, function(batch){
  super_batch_name <- str_remove(batch, "locityper_super_batch_")
  df_temp <- df00[df00$super_batch == batch, ]
  colnames(df_temp)[1] <- paste0("entity:ehr_sb", super_batch_name ,"_id")
  
  df_memb <- data.frame(memb = paste0("ehr_sb", super_batch_name, "_m", rep(seq(1,ceiling(nrow(df_temp)/N)),each=N)[1:nrow(df_temp)]), 
                      ehr = df_temp[,1],
                      check.names = FALSE)
  colnames(df_memb) <- c(paste0("membership:ehr_sb", super_batch_name,"_set_id"), paste0("ehr_sb", super_batch_name))
  
  write.table(df_temp, paste0("C:/docs/aou/locityper/mans_ehr/ehr_", super_batch_name,"_sample.tsv"), sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
  write.table(df_memb, paste0("C:/docs/aou/locityper/mans_ehr/ehr_", super_batch_name,"_memb.tsv")  , sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
  
  
})

```




```{r}
df00 <- fread("C:/docs/aou/locityper/1kg_benchmark/lt_svs_0129_GT.tsv", header=FALSE, sep="\t") %>% as.data.frame()
variantToSamples <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_var <- df00$V1[x]
  variantToSamples$put(curr_var, HashMap$new())
  temp2 <- lapply(2:ncol(df00), function(y){
    curr_string <- df00[x,y]
    curr_sample <- curr_string %>% str_extract(".*(?==)")
    curr_genotype <- curr_string %>% str_extract("(?<==).*")
    variantToSamples$get(curr_var)$put(curr_sample, curr_genotype)
  })
})

files <- list.files("C:/docs/aou/locityper/1kg_benchmark/gts_results/", full.names = TRUE, recursive = TRUE, pattern = ".csv")
df01 <- lapply(files, function(x){
  curr_sample <- x %>% str_extract("(?<=gts_results/gts.).*(?=.csv)")
  df_temp <- fread(x) %>% as.data.frame()
  df_temp$sample <- curr_sample
  return(df_temp)
}) %>% do.call(rbind, .)

df01$ref_gt <- lapply(seq_down(df01), function(x){
  curr_var <- df01$locus[x]
  curr_sample <- df01$sample[x]
  if(!variantToSamples$containsKey(curr_var)){return("NA")}
  if(variantToSamples$get(curr_var)$containsKey(curr_sample)){
    return(variantToSamples$get(curr_var)$get(curr_sample))
  } else {
    return("NA")
  }
}) %>% unlist

df01 <- df01[df01$ref_gt != "NA", ]
df01 <- df01[!grepl("\\.", df01$ref_gt), ]

df01$ltgt_to_gt <- lapply(seq_down(df01), function(x){
  rvals <- lapply(unlist(df01$genotype[x] %>% str_split(",")), function(gt){
    rval <- "ooblets"
    if(gt=="GRCh38"){
      rval <- "0"
    } else {
      which_hap  <- gt %>% str_extract("(?<=\\.).$") %>% as.numeric()
      which_samp <- gt %>% str_extract(".*(?=..$)")
      # print(paste0(gt,  ": ", which_samp, " ::: ", which_hap))
      rval <- unlist(str_split(variantToSamples$get(df01$locus[x])$get(which_samp), "\\|"))[which_hap]
    }
    return(rval)
  }) %>% unlist %>% paste0(collapse = "|")
  
  return(rvals)
}) %>% unlist

df01$ref_gt[df01$ref_gt=="0|1"] <- "1|0"
df01$ltgt_to_gt[df01$ltgt_to_gt=="0|1"] <- "1|0"

df01$match <- paste0(df01$ltgt_to_gt == df01$ref_gt)

g1 <- ggplot(df01) + 
  geom_point(data=df01, aes(x = quality, y=total_reads, color=match), alpha=0.7) + 
  ggtitle("locityper accuracy by call quality vs total_reads, ") + 
  labs(color="correct call") ; plot(g1)

ggplot(df01) + 
  geom_point(data=df01, aes(x=unexpl_reads, y=weight_dist, color=match), alpha=0.7)

ggplot(df01, aes(x=weight_dist, fill=match)) + 
  geom_histogram(alpha=0.7, position = "identity", aes(y = ..density..))

ggplot(df01, aes(x=unexpl_reads, fill=match)) + 
  geom_histogram(alpha=0.7, position = "identity", aes(y = ..density..))

df01$log_unexplained_reads <- log(1 + df01$unexpl_reads)
df01$log_weight_dist <- log(1 + df01$weight_dist)
ggplot(df01) + 
  geom_point(data=df01, aes(x=log_unexplained_reads, y=log_weight_dist, color=match), alpha=0.7)
g2 <- ggplot(df01, aes(x=log_weight_dist, fill=match)) + 
  geom_histogram(alpha=0.7, position = "identity", aes(y = ..density..)) +
  ggtitle("locityper accuracy by log(weight distribution) split by correct/not correct") + 
  labs(fill="correct gt call") ; plot(g2)
ggplot(df01, aes(x=log_unexplained_reads, fill=match)) + 
  geom_histogram(alpha=0.7, position = "identity", aes(y = ..density..))


ggsave("C:/docs/aou/locityper/benchmark_1kg_plot_01.png", plot=g1, width=3200, height=2000, units="px")
ggsave("C:/docs/aou/locityper/benchmark_1kg_plot_02.png", plot=g2, width=3200, height=2000, units="px")

```

g8 <- ggplot(data=df[df$size>0.125 & df$size<1500, ], aes(x=size_log2, fill=svtype)) +
  geom_histogram(alpha=0.7, position = "identity", aes(y = ..density..)) +
  scale_fill_manual(values=c(wes_palette("FantasticFox1")[5], wes_palette("FantasticFox1")[3])) +
  scale_x_continuous(breaks=-4:10, labels = 2^(-4:10)) + 
  xlab("CNV size (KB)") + 
  ggtitle("Raw CNV size distribution") 
g8


```{r}
df01$class <- lapply(seq_down(df01), function(x){
  if(df01$ltgt_to_gt[x] == "0|0"){
    if(df01$ref_gt[x]=="0|0"){
      return("TRUE_NEGATIVE")
    } else {
      return("FALSE_NEGATIVE")
    }
  } else {
    if(df01$ref_gt[x]=="0|0"){
      return("FALSE_POSITIVE")
    } else {
      return("TRUE_POSITIVE")
    }
  }
  return("NA")
}) %>% unlist()

MY_SENS <- sum(df01$class=="TRUE_POSITIVE")/sum(df01$class=="TRUE_POSITIVE" | df01$class=="FALSE_NEGATIVE")
MY_SPEC <- sum(df01$class=="TRUE_NEGATIVE")/sum(df01$class=="TRUE_NEGATIVE" | df01$class=="FALSE_POSITIVE")
MY_RECALL <- MY_SENS
MY_PRECISION <- sum(df01$class=="TRUE_POSITIVE")/sum(df01$class=="TRUE_POSITIVE" | df01$class=="FALSE_POSITIVE")
MY_F1 <- 2*(MY_PRECISION * MY_RECALL)/(MY_PRECISION + MY_RECALL)

list_by_variant <- lapply(df01$locus %>% u, function(x){
  df01[df01$locus==x, ]
})
names(list_by_variant) <- df01$locus %>% u

df_rval <- lapply(list_by_variant, function(x){
  print(x$locus[1])
  rval <- data.frame(name = x$locus[1], 
             accuracy = sum(x$match %>% as.logical) / nrow(x),
             sensitivity=sum(x$class=="TRUE_POSITIVE")/sum(x$class=="TRUE_POSITIVE" | x$class=="FALSE_NEGATIVE"),
             specificity=sum(x$class=="TRUE_NEGATIVE")/sum(x$class=="TRUE_NEGATIVE" | x$class=="FALSE_POSITIVE"),
             precision=sum(x$class=="TRUE_POSITIVE")/sum(x$class=="TRUE_POSITIVE" | x$class=="FALSE_POSITIVE"),
             n_samples = nrow(x),
             stringsAsFactors = FALSE)
  rval$F1 <- 2 * (rval$precision[1] * rval$sensitivity[1]) / (rval$precision[1] + rval$sensitivity[1])
  return(rval)
}) %>% do.call(rbind, .)
df_rval <- df_rval[order(df_rval$accuracy, decreasing = TRUE), ]
rownames(df_rval) <- seq_down(df_rval)
for(row in 1:nrow(df_rval)){
  for(col in 2:ncol(df_rval)){
    df_rval[row, col] <- round(df_rval[row, col], 3)
  }
}


df_panel <- fread("C:/docs/aou/locityper/1kg_benchmark/Locityper_SV_Panel - 1KG_TestRun.tsv") %>% as.data.frame()

df_rval$GENE <- lapply(df_rval$name, function(x){
  df_panel$GENE[df_panel$SVID==x]
}) %>% unlist

write.table(df_rval, "C:/docs/aou/locityper/1kg_benchmark/summary_stats_by_loci.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```



```{r}
x1 <- 800
y1 <- 42

x2 <- 146508
y2 <- (16*60+45.6) - (9*60+42)

rate <- (x2-x1)/(y2-y1)

est_time_min <- 200000 / rate

times <- u(c(times, est_time_min / 60))


paste0("recent: ", round(est_time_min/60, 3))
times %>% round(2) %>% sort
paste0("average: ", mean(times) %>% round(3))
```
 6.54 7.33 7.54 7.54 7.70 8.20 8.37 8.60 8.61

```{r}
df00 <- fread("C:/docs/aou/locityper/summary_gts_man/table.txt", sep="\t") %>% as.data.frame()
df00$ID <- lapply(df00$ID, function(x){
  unlist(str_split(x, ";"))[1]
}) %>% unlist

variantToSamples <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_var <- df00$ID[x]
  variantToSamples$put(curr_var, HashMap$new())
  temp2 <- lapply(4:ncol(df00), function(y){
    curr_sample <- colnames(df00)[y]
    curr_genotype <- df00[x,y]
    variantToSamples$get(curr_var)$put(curr_sample, curr_genotype)
  })
})

files <- list.files("C:/docs/aou/locityper/summary_gts_man/temp/", full.names = TRUE, recursive = TRUE, pattern = ".csv")
df01 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  return(df_temp)
}) %>% do.call(rbind, .)

df01$ltgt_to_gt <- lapply(seq_down(df01), function(x){
  rvals <- lapply(unlist(df01$genotype[x] %>% str_split(",")), function(gt){
    rval <- "ooblets"
    if(gt=="GRCh38"){
      rval <- "0"
    } else {
      which_hap  <- gt %>% str_extract("(?<=\\.).$") %>% as.numeric()
      which_samp <- gt %>% str_extract(".*(?=..$)")
      # print(paste0(gt,  ": ", which_samp, " ::: ", which_hap))
      rval <- unlist(str_split(variantToSamples$get(df01$locus[x])$get(which_samp), "\\|"))[which_hap]
    }
    return(rval)
  }) %>% unlist %>% paste0(collapse = "|")
  
  return(rvals)
}) %>% unlist

# df01$ltgt_to_gt[df01$ltgt_to_gt=="0|1"] <- "1|0"

df_map <- fread("C:/docs/aou/locityper/summary_gts_man/004-isaac-first-batch_SVs_of_interest_013025_CDS_name_map.tsv") %>% as.data.frame()
codeToName <- HashMap$new()
codeToName$populate(df_map, "new_name", "V1")

df01$converted_names <- lapply(seq_down(df01), function(x){
  rvals <- lapply(unlist(df01$genotype[x] %>% str_split(",")), function(gt){
    rval <- "GRCh38"
    if(gt!="GRCh38"){
      # which_hap  <- gt %>% str_extract("(?<=\\.).$") %>% as.numeric()
      which_samp <- gt %>% str_extract(".*(?=..$)")
      if(str_detect(which_samp, ":")){
        which_samp <- which_samp %>% str_extract("(?<=:).*?$")
      }
      # print(paste0(gt,  ": ", which_samp, " ::: ", which_hap))
      rval <- codeToName$get(which_samp)
      if(str_detect(rval, ":")){
        temp1 <- unlist(str_split(rval, ":"))
        rval <- temp1[length(temp1)]
      }
    }
    return(rval)
  }) %>% unlist %>% paste0(collapse = ",")
  
  return(rvals)
}) %>% unlist


write.table
```




```{r}
df00 <- fread("C:/docs/aou/locityper/summary_gts_man/locityper_ehr_gt_summary.tsv") %>% as.data.frame()
df_table <- fread("C:/docs/aou/locityper/summary_gts_man/table.txt") %>% as.data.frame()
df_sheet <- fread("C:/docs/aou/locityper/summary_gts_man/SV Intersections AoU Phase 2 - Top SVs of interest.tsv") %>% as.data.frame()
df_table <- df_table[, 1:3]
df_table$key1 <- lapply(df_table$ID, function(x){
  unlist(str_split(x, ";"))[1]
}) %>% unlist

df_table <- df_table[df_table$key1 %in% df00$locus, ]

df_table$value  <- lapply(seq_down(df_table), function(x){
  bool_chr <- df_table$CHROM[x] == df_sheet$CHROM
  bool_pos <- df_table$POS[x] == df_sheet$POS
  
  curr_svtype <- str_extract(df_table$key1[x], "(INS)|(DEL)")
  bool_svtype <- grepl(curr_svtype, df_sheet$ID)
  
  rval <- df_sheet$ID[bool_chr & bool_pos & bool_svtype]
  
  if(length(rval)==1){
    return(rval)
  } else {
    return("NONE")
  }
}) %>% unlist

df_table$gene <- lapply(df_table$value, function(x){
  if(x=="NONE"){
    return("")
  } else {
    return(df_sheet$`GENE(S)`[df_sheet$ID==x])
  }
}) %>% unlist

df_table$ID <- NULL
write.table(df_table, "C:/docs/aou/locityper/summary_gts_man/temp.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

# manually create table

df_table <- fread("C:/docs/aou/locityper/summary_gts_man/conversion_table_vars.txt") %>% as.data.frame()
keyToValue <- HashMap$new()
keyToValue$populate(df_table, "key1", "value")
keyToGene <- HashMap$new()
keyToGene$populate(df_table, "key1", "gene")
df00$gene <- lapply(df00$locus, function(x){keyToGene$get(x)}) %>% unlist
df00$name <- lapply(df00$locus, function(x){keyToValue$get(x)}) %>% unlist
write.table(df00, "C:/docs/aou/locityper/summary_gts_man/locityper_ehr_table_20260219.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```



```{r}
library(wesanderson)

df00$ltgt_to_gt[df00$ltgt_to_gt=="1|0"] <- "0|1"
df00$ltgt_to_gt[df00$ltgt_to_gt==".|1"] <- "1|."

list_by_locus <- lapply(u(df00$locus), function(x){
  df00[df00$locus == x, ]
})
names(list_by_locus) <- u(df00$locus)

my_plots <- lapply(list_by_locus, function(temp1){
  temp2 <- temp1$ltgt_to_gt %>% table
  df_temp <- data.frame(gts = names(temp2),
                        counts = unname(temp2),
                        stringsAsFactors = FALSE)
  df_temp$props <- temp2 %>% prop.table() %>% unname()
  g1 <- ggplot(temp1, aes(x = ltgt_to_gt, y = after_stat(prop), group = 1)) +
    geom_bar(fill = wes_palette("Zissou1")[1]) +
    scale_y_continuous(labels = scales::percent, breaks = seq(0,1,0.05)) +
    ylab("Proportion of Total") + 
    geom_hline(yintercept = 0.05, color = "red", size=1) + 
    ggtitle(paste0(temp1$name[1], ": ", temp1$gene[1], ";  N:", nrow(temp1))) + 
    xlab("genotype") + 
    geom_text(data = df_temp, aes(label=counts.Freq, x=gts, y=props+0.03))
  
  ggsave(paste0("C:/docs/aou/locityper/summary_gts_man/plots/", temp1$name[1], ".png"), plot=g1)
  
  
})
```

