---
title: "autism"
output: html_document
date: "2024-04-12"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
df <- fread("C:/home/autism/r_edits/temp.txt") %>% as.data.frame()

# for each row in the FORMAT column, get a list of which index corresponds to each field
my_FORMAT <- df$FORMAT %>% str_split(":")
index_GT <- lapply(my_FORMAT, function(x){which(x=="GT")}) %>% unlist
index_AD <- lapply(my_FORMAT, function(x){which(x=="AD")}) %>% unlist
index_DP <- lapply(my_FORMAT, function(x){which(x=="DP")}) %>% unlist


# get a vector of indexes for columns corresponding to fa, mo, p1, p2, etc
which_columns_to_split <- which(colnames(df) %in% c("FA", "MO") | grepl("\\.p[0-9]$", colnames(df)))
family_columns_to_cbind <- lapply(which_columns_to_split, function(COL_INDEX){
  split_member <- df[, COL_INDEX] %>% str_split(":")
  # identify which columns have more FORMAT fields than the FA column has fields, we won't be able to get good AD or DP infor for these
  bool_useable <- (my_FORMAT %>% lapply(length) %>% unlist)==(split_member %>% lapply(length) %>% unlist)
  columns_member <- lapply(seq_down(df), function(x){
    if(bool_useable[x]==FALSE){
      if(split_member[[x]][1] %>% grepl("\\/|\\|")){ # check to see if the first field contains GT information
        return(c(split_member[[x]][1], 0,0))
      } else {
        return(c("./.", 0, 0))
      }
    }
    
    # return the corresponding row from current family member, and the corresponding column for each corresponding field
    return(c(split_member[[x]][index_GT[x]],
             split_member[[x]][index_AD[x]],
             split_member[[x]][index_DP[x]]))
  }) %>% do.call(rbind, .) %>% as.data.frame()
  colnames(columns_member) <- paste0(colnames(df)[COL_INDEX], "_", c("GT", "AD", "DP"))
  return(columns_member)
}) %>% do.call(cbind, .) %>% as.data.frame()



```

# asm qc manifest
```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats")
# df0 <- fread("C:/home/autism/asm_qc/manifest_job_done1.tab") %>% as.data.frame()
# df0[is.na(df0)] <- "NA"

df0 <- read.table("C:/docs/autism/finished_20240617")
colnames(df0) <- "sample"


bool_child <- grepl("_(p.|s.)$", df0$sample)
df1 <- data.frame(SAMPLE = df0$sample, 
                  H1 = paste0(
                    "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/asm_iw/", 
                    df0$sample, "/", df0$sample, 
                    "/assemblies/hifiasm/",
                    ifelse(bool_child, "trio/", ""),
                    "0.16.1/",
                    df0$sample,
                    ".hifiasm.",
                    ifelse(bool_child, "dip", "bp"),
                    ".hap1.p_ctg.gfa.fasta"),
                  H2 = paste0(
                    "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/asm_iw/", 
                    df0$sample, "/", df0$sample, 
                    "/assemblies/hifiasm/",
                    ifelse(bool_child, "trio/", ""),
                    "0.16.1/",
                    df0$sample, 
                    ".hifiasm.",
                    ifelse(bool_child, "dip", "bp"),
                    ".hap2.p_ctg.gfa.fasta"),
                  ILLUMINA = df0$sample,
                  FOFN = paste0(
                    "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/fofn/Illumina/",
                    df0$sample, 
                    "-Illumina.fastq.fofn"),
                  TRIO = lapply(df0$sample, function(x){
                    if(df00$Family_Type[grep(x, df00$SP_ID)] == "Quad, Simplex"){
                      return("NO")
                    } else {
                      return("YES")
                    }}) %>% unlist,
                  MO_ID = ifelse(bool_child, 
                                  df0$sample %>% str_replace("_(p.|s.)$", "_mo"),
                                  "NA"),
                  FA_ID = ifelse(bool_child, 
                                  df0$sample %>% str_replace("_(p.|s.)$", "_fa"),
                                  "NA")
                  
                  )
write.table(df1, "C:/home/autism/asm_qc/manifest.asmqc.20240613.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```


```{r}
df2 <- read.table("C:/docs/autism/manifest.rerun.new", header=TRUE)
df3 <- df1[!df1$SAMPLE %in% df2$SAMPLE, ]
write.table(df3, "C:/home/autism/asm_qc/manifest.asmqc.20240617.asmqc.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```



```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=448025372", sheet = "FOFN_summary") %>% as.data.frame()
```
# update tracking sheet
```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet="Individual_Stats") %>% as.data.frame()

man_aqc <- fread("C:/docs/autism/asm_qc/manifest.tab.2024-04-22.finished") %>% as.data.frame()
df00$`asm_QC -- Isaac`[df00$SP_ID %in% man_aqc$SAMPLE] <- "Done"

man_pbsv <- fread("C:/docs/autism/manifest.tab.pbsv.2024-04-23.finished") %>% as.data.frame()
df00$`PBSV -- Isaac`[df00$SP_ID %in% man_pbsv$SAMPLE] <- "Done"

man_aln <- fread("C:/docs/autism/aln.samples") %>% as.data.frame()
df00$`aln -- Isaac`[df00$SP_ID %in% man_aln$sample] <- "Done"
df00$`aln_QC -- Isaac (VBI)`[df00$SP_ID %in% man_aln$sample] <- "Done"

write_sheet(df00, "https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", "temp_isaac")
```

```{r}
df1 <- fread("C:/docs/autism/asm_qc/merq.stats.qv.tsv") %>% as.data.frame()
df1 <- df1[!grepl("Both", df1$V1), ]
sample_ids <- df1$V1 %>% str_remove("_hap(1|2)") %>% u
df10 <- lapply(sample_ids, function(x){
  temp_rows <- df1[grepl(x, df1$V1), ]
  data.frame(sample = x, 
             QV_Merqury_hap1 = ifelse(any(grepl("hap1", temp_rows[,1])), temp_rows[grepl("hap1", temp_rows[,1]), 4] %>% round, "NA"), 
             QV_Merqury_hap2 = ifelse(any(grepl("hap2", temp_rows[,1])), temp_rows[grepl("hap2", temp_rows[,1]), 4] %>% round, "NA") 
             ) %>% return()
}) %>% do.call(rbind, . )
rm(df1)

df2 <- fread("C:/docs/autism/asm_qc/merq.stats.completeness.tsv") %>% as.data.frame()
df2 <- df2[!grepl("both", df2$V1), ]
sample_ids <- df2$V1 %>% str_remove("_hap(1|2)") %>% u
df20 <- lapply(sample_ids, function(x){
  temp_rows <- df2[grepl(x, df2$V1), ]
  data.frame(sample = x, 
             Completeness_hap1=ifelse(any(grepl("hap1", temp_rows[,1])), temp_rows[grepl("hap1", temp_rows[,1]), 5] %>% round, "NA"), 
             Completeness_hap2=ifelse(any(grepl("hap2", temp_rows[,1])), temp_rows[grepl("hap2", temp_rows[,1]), 5] %>% round, "NA") 
             ) %>% return()
}) %>% do.call(rbind, . )
rm(df2)

df3 <- fread("C:/docs/autism/asm_qc/contig_stats.tsv") %>% as.data.frame()
sample_ids <- df3$source %>% str_remove("_hap..n50.stats") %>% u
df30 <- lapply(sample_ids, function(x){
  temp_rows <- df3[grep(x, df3$source), ]
  data.frame(sample = x,
             Contig_N50_Mbp_hap1=
               ifelse(any(grepl("hap1", temp_rows[,1])), 
                      temp_rows[grepl("hap1", temp_rows[,1]) & grepl("N50 ", temp_rows[,2]), 2] %>% str_remove("N50 \\(Mbp\\):") %>% str_trim(), 
                      "NA"),
             Contig_N50_Mbp_hap2=
               ifelse(any(grepl("hap2", temp_rows[,1])), 
                      temp_rows[grepl("hap2", temp_rows[,1]) & grepl("N50 ", temp_rows[,2]), 2] %>% str_remove("N50 \\(Mbp\\):") %>% str_trim(), 
                      "NA")
             )
}) %>% do.call(rbind, .)
rm(df3)

df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet="Individual_Stats") %>% as.data.frame()


df123 <- merge(df10, df20, all = TRUE) %>% merge(df30, all=TRUE)
df123[is.na(df123)] <- "NA"

df01 <- merge(df00, df123, by.x = "SP_ID", by.y="sample", all=TRUE)
df01 <- df01[match(df00$SP_ID, df01$SP_ID), ]
write_sheet(df01, "https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", "temp_isaac")



bool_merq_resubmit <- df123$QV_Merqury_hap1=="NA" | df123$QV_Merqury_hap2=="NA"
bool_contig_resubmit <- df123$Contig_N50_Mbp_hap1 == "NA" | df123$Contig_N50_Mbp_hap2 == "NA"

```



```{r}
BOOLS <- c(TRUE, FALSE)
temp_n <- 15
temp_df <- data.frame(temp1 = sample(BOOLS, temp_n, replace = TRUE),
                      temp2 = sample(BOOLS, temp_n, replace = TRUE),
                      temp3 = sample(BOOLS, temp_n, replace = TRUE),
                      temp4 = sample(BOOLS, temp_n, replace = TRUE))
temp_df$ALL <- lapply(seq_down(temp_df), function(x){
  all(temp_df[x, ])
}) %>% unlist
```



```{r}
df0 <- read.table("C:/docs/autism/manifest.tab", sep="\t", header = TRUE)

df1 <- data.frame(SAMPLE = df0$SAMPLE, 
                  ROLE = ifelse(
                    df0$SAMPLE %>% grepl("_fa$", .), "father", ifelse(
                    df0$SAMPLE %>% grepl("_mo$", .), "mother", ifelse(
                    df0$SAMPLE %>% grepl("_p.$", .), "proband", "sibling"))),
                  FAMILY = df0$SAMPLE %>% gsub("_..$", "", .),
                  HAP1 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/saffire/results/",
                                df0$SAMPLE, "_hap1/alignments/", df0$SAMPLE, "_hap1.minimap2.paf"),
                  HAP2 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/saffire/results/",
                                df0$SAMPLE, "_hap2/alignments/", df0$SAMPLE, "_hap2.minimap2.paf"))
write.table(df1, "C:/docs/autism/manifest.concatchr.tab", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```






```{r}
df0 <- fread("C:/docs/autism/fofn/sample_list_fofn_to_Isaac.txt") %>% as.data.frame()

my_samples <- u(df0$SAMPLE)

lapply(my_samples, function(x){
  
  write.table(df0$FASTQ[df0$SAMPLE == x], paste0("C:/docs/autism/fofn/", x, ".fofn"), sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE)
  
  
})


```



```{r}
df0 <- fread("C:/docs/autism/ncdu_autism_sorted.tsv") %>% as.data.frame()
df0$notes <- NULL

file_types <- c(".gfa", ".bin", ".bed", ".yak")
my_regex <- paste0("assemblies/hifiasm/0.16.1/.*","(", paste0(file_types, collapse = "|"), ")$"  )
my_regex2 <- paste0("/asm_ys/.*_(s|p)1/.*.hifiasm.bp.hap")

df1 <- df0[grepl("^/net/eichler/vol28/projects/autism_genome_assembly/nobackups/asm_ys", df0$file), ]
df11 <- df1[grepl(my_regex, df1$file) | grepl(my_regex2, df1$file), ]
write.table(df11$file, "C:/docs/autism/to_delete.txt", sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r}
df1 <- readLines("C:/docs/autism/asm_files.txt") %>% as.data.frame()

file_types <- c(".gfa", ".bin", ".bed", ".yak")
my_regex <- paste0("assemblies/hifiasm/0.16.1/.*","(", paste0(file_types, collapse = "|"), ")$"  )
my_regex2 <- paste0("/asm_ys/.*_(s|p)1/.*.hifiasm.bp.hap")

df11 <- df1[grepl(my_regex, df1$.) | grepl(my_regex2, df1$.), ]
write.table(df11, "C:/docs/autism/to_delete.txt", sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r}
formals(write.table)$row.names <- FALSE
formals(write.table)$col.names <- TRUE
formals(write.table)$sep <- "\t"
formals(write.table)$quote <- FALSE

write.table(df12, "C:/Users/iwong1/Downloads/temp.tsv")
```


```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats")
df01 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "VBI_stats")

table(df00$SP_ID %in% df01$sample)
df00$SP_ID[!df00$SP_ID %in% df01$sample]


```
  "14444_fa" : ys
  "11569_mo" : na
  "13799_fa" : ys
  "12608_mo" : na


```{bash}
cat batch_2_samples.txt  | xargs -I {} sh -c "mkdir {}; cd {}; ln -s /net/eichler/vol26/7200/software/pipelines/hifiasm-smk/hifi/runlocal; cp ../hifiasm_trio.yaml a; cd .."
```


```{r}
input_file <- "C:/Users/iwong1/Downloads/batch_2_autism.txt"
df0 <- fread(input_file, header = FALSE) %>% as.data.frame()
fofn_dir <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/0_hifi_fofn/"
out_dir <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/asm_iw/"
temp1 <- lapply(df0$V1, function(x){
  fam_id <- x %>% str_remove("_..$")
  temp_df <- data.frame(
    sample = x,
    hifi_fofn = paste0(fofn_dir, x, ".fofn"),
    maternal_illumina = ifelse(grepl("(_p|_s)", x), paste0(fofn_dir, fam_id, "_mo.fofn"), "NA"),
    paternal_illumina = ifelse(grepl("(_p|_s)", x), paste0(fofn_dir, fam_id, "_fa.fofn"), "NA"),
    family_id = fam_id
  )
  # write.table(temp_df, paste0(out_dir, x, "/man.tab"), sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
  return(temp_df)
})
names(temp1) <- paste0(out_dir, df0$V1, "/man.tab")
```


```{r}
my_lines <- readLines("C:/docs/autism/outputs.txt")
new_breaks <- grep("____________", my_lines)
list_by_sample <- lapply(seq_along(new_breaks)[-c(1)], function(x){
  if(x == 1){
    return(my_lines[1:new_breaks[x]])
  } else if(x >= (length(new_breaks)-1)){
    return(my_lines[new_breaks[x]:length(my_lines)])
  } else {
    return(my_lines[new_breaks[x]:new_breaks[x+1]])    
  }
})

df0 <- lapply(list_by_sample, function(x){
  curr_sample <- x[2]
  correct_fofn <- grepl("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/0_hifi_fofn/", x) %>% any
  no_logs <- grepl("total 0", x) %>% any
  fastas <- grepl(".hifiasm.bp.hap2.p_ctg.gfa.fasta", x) %>% any
  
  data.frame(sample = curr_sample, 
             fofn = correct_fofn,
             logs = no_logs, 
             fasta = fastas) %>% return()
}) %>% do.call(rbind, .)

df_ys <- read.table("C:/docs/autism/finished_ys", header = FALSE)
df0$ys <- df0$sample %in% df_ys$V1

df0$FINISHED <- df0$ys | (df0$fofn & df0$logs & df0$fasta)

df0$sample[!df0$FINISHED] %>% write.table("C:/docs/autism/outs_fail.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)



files_all <- readLines("C:/docs/autism/all_files.txt")
df1 <- data.frame(sample = files_all %>% str_extract("(?<=./).*?(?=/)"), 
                  file = files_all)
df1 <- df1[!is.na(df1$sample), ]
df1 <- df1[df1$sample %in% df0$sample[df0$FINISHED], ]
files_to_delete <- c(".hifiasm.bp.hap1.p_ctg.gfa$",
                     ".hifiasm.bp.hap1.p_ctg.lowQ.bed$",
                     ".hifiasm.bp.hap1.p_ctg.noseq.gfa$",
                     ".hifiasm.bp.hap2.p_ctg.gfa$",
                     ".hifiasm.bp.hap2.p_ctg.lowQ.bed$",
                     ".hifiasm.bp.hap2.p_ctg.noseq.gfa$",
                     ".hifiasm.bp.p_ctg.gfa$",
                     ".hifiasm.bp.p_ctg.lowQ.bed$",
                     ".hifiasm.bp.p_ctg.noseq.gfa$",
                     ".hifiasm.bp.p_utg.gfa$",
                     ".hifiasm.bp.p_utg.lowQ.bed4",
                     ".hifiasm.bp.p_utg.noseq.gfa$",
                     ".hifiasm.bp.r_utg.gfa$",
                     ".hifiasm.bp.r_utg.lowQ.bed$",
                     ".hifiasm.bp.r_utg.noseq.gfa$",
                     ".hifiasm.dip.hap1.p_ctg.gfa$",
                     ".hifiasm.dip.hap1.p_ctg.lowQ.bed$",
                     ".hifiasm.dip.hap1.p_ctg.noseq.gfa$",
                     ".hifiasm.dip.hap2.p_ctg.gfa$",
                     ".hifiasm.dip.hap2.p_ctg.lowQ.bed$",
                     ".hifiasm.dip.hap2.p_ctg.noseq.gfa$",
                     ".hifiasm.dip.p_utg.gfa$",
                     ".hifiasm.dip.p_utg.lowQ.bed$",
                     ".hifiasm.dip.p_utg.noseq.gfa$",
                     ".hifiasm.dip.r_utg.gfa$",
                     ".hifiasm.dip.r_utg.lowQ.bed$",
                     ".hifiasm.dip.r_utg.noseq.gfa$",
                     ".hifiasm.ec.bin$",
                     ".hifiasm.ovlp.reverse.bin$",
                     ".hifiasm.ovlp.source.bin$",
                     ".hifiasm.bp.p_utg.lowQ.bed$") %>% paste(. , collapse = ")|(") %>% paste0("(", ., ")")
df1$to_delete <- lapply(df1$file, function(x){grepl(files_to_delete, x)}) %>% unlist
write.table(df1$file[df1$to_delete], "C:/docs/autism/files_to_delete.tsv", row.names = FALSE, quote=FALSE, col.names = FALSE)
```



get vcf counts per region
PAR:
chrY:10,000-2,781,479 and chrY:56,887,902-57,217,415
chrX:10,000-2,781,479 and chrX:155,701,382-156,030,895
```{r}
files_vcf <- list.files(path="/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pav", pattern = "vcf$", full.names = TRUE)


df0 <- lapply(files_vcf, function(x){
  sample_id <- x %>% bname %>% str_remove(".vcf.gz$")
  n_chrY_all <- system2("bcftools", paste0("view -H ",  x, " chrY | wc -l"), stdout=TRUE)
  n_chrY_PAR1 <- system2("bcftools", paste0("view -H ",  x, " chrY:10000-2781479 | wc -l"), stdout=TRUE)
  n_chrY_PAR2 <- system2("bcftools", paste0("view -H ",  x, " chrY:56887902-57217415 | wc -l"), stdout=TRUE)
  n_chrX_all <- system2("bcftools", paste0("view -H ",  x, " chrY | wc -l"), stdout=TRUE)
  n_chrX_PAR1 <- system2("bcftools", paste0("view -H ",  x, " chrX:10000-2781479 | wc -l"), stdout=TRUE)
  n_chrX_PAR2 <- system2("bcftools", paste0("view -H ",  x, " chrX:155701382-156030895 | wc -l"), stdout=TRUE)
  
  data.frame(sample = sample_id, 
             n_chrY_all = n_chrY_all,
             n_chrY_PAR1 = n_chrY_PAR1, 
             n_chrY_PAR2 = n_chrY_PAR2,
             n_chrX_all = n_chrX_all,
             n_chrX_PAR1 = n_chrX_PAR1, 
             n_chrX_PAR2 = n_chrX_PAR2,) %>% return
}) %>% do.call(rbind, .)

write.table(df0, "vcf_counts.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)



```


get F samples
```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats") 
df0 <- df00 %>% as.data.frame()
df1 <- df0$SP_ID[df0$Sex == "F"]
write.table(df1, "C:/docs/autism/F_samples.txt", quote=FALSE, row.names = FALSE)
```


as.POSIXct(paste(df0$year, df0$month, df0$day, df0$time, sep="-"), format = "%Y-%m-%d-%H:%M")

ggplot(df1, aes(x = dt_full, y = dt_day)) + 
  geom_point() + 
  coord_cartesian(ylim = c(as.POSIXct("9:00", format = "%H:%M"), as.POSIXct("17:00", format = "%H:%M"))) + 
  scale_y_continuous(breaks = as.POSIXct(9:17 %>% paste0(":00"), format = "%H:%M"), labels = paste0(c(9:12,1:5), c(rep(" AM",4), rep(" PM", 5)))) + 
  ylab("") + 
  geom_smooth(method='lm') +
  ggtitle("daily Lipton time")

```{r}
time <- c("04:15:52", "04:20:22") %>% as.POSIXct(format="%H:%M:%S")
available <- c(2138, 2096)

df_free <- data.frame(time = time,
                      free = available)

slope <- -(max(available) - min(available)) / as.numeric(max(time) - min(time))
b <- available[1] - slope * as.numeric(time[1])
deadline <- -b / (slope)


ggplot(df_free, aes(x=time, y = free)) + 
  geom_point() + 
  geom_smooth(method='lm') +
  coord_cartesian(ylim=c(0,2300), xlim=c(as.POSIXct("04:00:00", format="%H:%M:%S"), as.POSIXct("09:00:00", format="%H:%M:%S")))

```

```{r}
myLines <- readLines("C:/docs/autism/asm_qc/files.all.20240617.txt")
myLines <- myLines[grepl("/assemblies/hifiasm/.*\\....$", myLines)]
my_samples <- myLines %>% lapply(function(x){str_split(x, "/")[[1]][2]}) %>% unlist

all_samples <- u(my_samples)
list_files_by_sample <- lapply(all_samples, function(x){
  myLines[my_samples==x]
})
names(list_files_by_sample) <- all_samples


target_samples <- all_samples[grepl("_(p|s).",  all_samples)]
bool_has_base <- lapply(target_samples, function(x){
  has_base <- grepl("hifiasm.bp.hap1", list_files_by_sample[[x]]) %>% any
  has_trio <- grepl("trio", list_files_by_sample[[x]]) %>% any
  c(x, has_base, has_trio)
}) %>% do.call(rbind, .) %>% as.data.frame()

```



p1 and s1/2 samples without trio files but with normal files

# MAIN MANIFEST UPDATE LOG FILE PATHS 
```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats") 
df00 <- df00[-1, ]
df_comments <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=1859063258", sheet = "output-paths-iw") %>% as.data.frame()
df_comments[is.na(df_comments)] <- ""
df1 <- readLines("C:/docs/autism/asm_qc/files.fasta.iw.txt")
df2 <- readLines("C:/docs/autism/asm_qc/files.fasta.ys.txt")

all_samples <- df00$SP_ID
all_samples <- all_samples[!is.na(all_samples)]

df0 <- data.frame(sample = all_samples)
df0$type <- lapply(df0$sample, function(x){
  bool_is_quad <- grepl("Quad", df00$Family_Type[df00$SP_ID==x]) %>% unlist
  if(bool_is_quad){return("Quad")} else {return("Trio")}
}) %>% unlist

find_asm_output <- function(sample_id, hap){
    if(grepl("_(s|p).$", sample_id)){
    mypath <- paste0("asm_iw/", sample_id,"/", sample_id, "/assemblies/hifiasm/trio/0.16.1/", sample_id, ".hifiasm.dip.", hap,".p_ctg.gfa.fasta")
  } else {
    mypath <- paste0("asm_iw/", sample_id, "/", sample_id, "/assemblies/hifiasm/0.16.1/", sample_id, ".hifiasm.bp.", hap, ".p_ctg.gfa.fasta" )
  }
  
  if(mypath %in% df1){return(mypath)}
  
  if(grepl("_(s|p).$", sample_id)){
    mypath <- paste0("asm_ys/", sample_id, "/assemblies/hifiasm/trio/0.16.1/", sample_id, ".hifiasm.dip.", hap, ".p_ctg.gfa.fasta")
  } else {
    mypath <- paste0("asm_ys/", sample_id, "/assemblies/hifiasm/0.16.1/", sample_id, ".hifiasm.bp.", hap, ".p_ctg.gfa.fasta")
  }
  
  if(mypath %in% df2){return(mypath)}
  
  return("")
}
df0$hap1 <- lapply(df0$sample, function(x){
  find_asm_output(x, "hap1")
}) %>% unlist

df0$hap2 <- lapply(df0$sample, function(x){
  find_asm_output(x, "hap2")
}) %>% unlist

df3 <- readLines("C:/docs/autism/asm_qc/pbsv.grch38.files.txt")
df4 <- readLines("C:/docs/autism/asm_qc/pbsv.grch38_noY.files.txt")

find_pbsv_outputs <- function(sample_id, pbsv_str, line_src, pre_str){
  mypath <- paste0(pre_str, sample_id, "/pbsv_", sample_id, "_", pbsv_str,".vcf.gz")
  if(mypath %in% line_src){return(mypath)} else {return("")}
}

df0$PBSV_GRCH38_SV <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "sv", df3, "pbsv/GRCH38/results/")}) %>% unlist
df0$PBSV_GRCH38_DUP <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "dup", df3, "pbsv/GRCH38/results/")}) %>% unlist
df0$PBSV_GRCH38_BND <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "bnd", df3, "pbsv/GRCH38/results/")}) %>% unlist

df0$`PBSV_GRCH38-noY_SV` <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "sv", df4, "pbsv/GRCH38_noY/results/")}) %>% unlist
df0$`PBSV_GRCH38-noY_DUP` <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "dup", df4, "pbsv/GRCH38_noY/results/")}) %>% unlist
df0$`PBSV_GRCH38-noY_BND` <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "bnd", df4, "pbsv/GRCH38_noY/results/")}) %>% unlist


df5 <- readLines("C:/docs/autism/asm_qc/files_fcs_old.txt")
df6 <- readLines("C:/docs/autism/asm_qc/files_fcs_batch01.txt")
df7 <- readLines("C:/docs/autism/asm_qc/files_fcs_batch02.txt")
find_fcs_outputs <- function(sample_id, hap){
  
  
  mypath <- paste0("assembly_qc/QC_results/contamination_screening/results/", sample_id,"_", hap,"/fasta/", sample_id, "_", hap, ".fasta")
  if(mypath %in% df6){return(mypath)}
  
  mypath <- paste0("assembly_qc_GROUP2/QC_results/contamination_screening/results/", sample_id,"_", hap,"/fasta/", sample_id, "_", hap, ".fasta")
  if(mypath %in% df7){return(mypath)}
  
  mypath <- paste0("assembly_qc/QC_results_old/contamination_screening/results/", sample_id,"_", hap,"/fasta/", sample_id, "_", hap, ".fasta")
  if(mypath %in% df5){return(mypath)}

  return("")
}
df0$asm_qc_hap1 <- lapply(df0$sample, function(x){find_fcs_outputs(x, "hap1")}) %>% unlist
df0$asm_qc_hap2 <- lapply(df0$sample, function(x){find_fcs_outputs(x, "hap2")}) %>% unlist


df8 <- read.table("C:/docs/autism/asm_qc/1_hifialn.grch38.man.tsv", header=TRUE)
df9 <- read.table("C:/docs/autism/asm_qc/1_hifialn.grch38_noY.man.tsv", header=TRUE)
df0$aln_grch38 <- lapply(df0$sample, function(x){
  if(x %in% df8$SAMPLE){
    df8$BAM[df8$SAMPLE==x] %>% str_remove("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/") %>% return  
  } else {
    return("")
  }
}) %>% unlist
df0$`aln_grch38-noY` <- lapply(df0$sample, function(x){
  if(x %in% df9$SAMPLE){
    df9$BAM[df9$SAMPLE==x] %>% str_remove("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/") %>% return  
  } else {
    return("")
  }
}) %>% unlist


df10 <- read.table("C:/docs/autism/asm_qc/files.pav.txt")
df0$`pav_GRCH38` <- lapply(df0$sample, function(x){
  mypath <- paste0("pav_new/GRCH38/results_vcf/pav_", x, ".vcf.gz")
  if(mypath %in% df10$V1){return(mypath)}
  
  mypath <- paste0("pav_new/GRCH38/rescue/pav_", x, ".vcf.gz")
  if(mypath %in% df10$V1){return(mypath)}
  
  mypath <- paste0("pav_new/GRCH38_noY/results_vcf/pav_", x, ".vcf.gz")
  if(mypath %in% df10$V1){return(mypath)}
  
  mypath <- paste0("pav_new/GRCH38_noY/rescue/pav_", x, ".vcf.gz")
  if(mypath %in% df10$V1){return(mypath)}
  
  return("")
}) %>% unlist





df_comments <- df_comments[order(df_comments$sample), ]
df0 <- df0[order(df0$sample), ]
BOOL_CONTROL <- all(df_comments$sample == df0$sample)

if(BOOL_CONTROL){
  for(row in 1:nrow(df0)){
    for(col in 1:ncol(df0)){
      if(df0[row,col] == "" & df_comments[row,col]!="" & !grepl("/", df_comments[row,col])){
        df0[row,col] <- df_comments[row,col]
      } else if(df0[row,col] == "" & df_comments[row,col]!=""){
        print(paste0("change in ", row, ", ", col, ": ", df_comments[row,col]))
      }
    }
  }
  write.table(df0[order(df0$sample), ], "C:/docs/autism/asm_qc/autism_output_manifest_20240618.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
} else {
  print("warning new samples added!!")
}
```


```{r}
df_01 <- df0[!grepl("vcf.gz",df0$pav_GRCH38) & grepl("fasta", df0$asm_qc_hap1), c("sample", "asm_qc_hap1", "asm_qc_hap2", "pav_GRCH38")]
df_01_m <- df_01[df_01$sample %in% df00$SP_ID[df00$Sex=="M"], ]
df_01_f <- df_01[df_01$sample %in% df00$SP_ID[df00$Sex=="F"], ]


df_01_m_man <- data.frame(NAME = df_01_m$sample, 
                          HAP1 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_01_m$asm_qc_hap1), 
                          HAP2 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_01_m$asm_qc_hap1))

df_01_f_man <- data.frame(NAME = df_01_f$sample, 
                          HAP1 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_01_f$asm_qc_hap1), 
                          HAP2 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_01_f$asm_qc_hap1))

write(df_01_m_man, "C:/docs/autism/pav.manifest.m.rescue.tsv")
write(df_01_f_man, "C:/docs/autism/pav.manifest.f.rescue.tsv")


df_01 <- df0[!grepl("vcf.gz",df0$pav_GRCH38) & grepl("fasta", df0$asm_qc_hap1), c("sample", "hap1", "hap2")]
df_01_rescue <- data.frame(NAME = df_01$sample, 
                          HAP1 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_01$hap1), 
                          HAP2 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_01$hap2))

write(df_01_rescue, "C:/docs/autism/asm_qc/pav_man_rescue.tsv")
```





```{r}
df_asmqc_03 <- df0[grepl("fasta", df0$hap1) & !grepl("fasta", df0$asm_qc_hap1), c("sample", "hap1", "hap2", "asm_qc_hap1", "asm_qc_hap2")]
df_asmqc_03 <- df0[grepl("fasta", df0$hap1), c("sample", "hap1", "hap2", "asm_qc_hap1", "asm_qc_hap2")]
df_temp <- fread("C:/docs/autism/asm_qc/asm_qc_group3.txt") %>% as.data.frame()
df_asmqc_03 <- df_asmqc_03[df_asmqc_03$sample %in% df_temp$SP_ID, ]


df_asmqc_03_man <- data.frame(
  SAMPLE = df_asmqc_03$sample, 
  H1 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_asmqc_03$hap1), 
  H2 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_asmqc_03$hap2), 
  ILLUMINA = df_asmqc_03$sample, 
  FOFN = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/fofn/Illumina/", df_asmqc_03$sample,"-Illumina.fastq.fofn"), 
  TRIO = "NO", 
  MO_ID = ifelse(grepl("_(p|s).$", df_asmqc_03$sample), str_replace(df_asmqc_03$sample, "_(p|s).$", "_mo"), "NA"), 
  FA_ID = ifelse(grepl("_(p|s).$", df_asmqc_03$sample), str_replace(df_asmqc_03$sample, "_(p|s).$", "_fa"), "NA")
  )

df_temp <- readLines("C:/docs/autism/asm_qc/files.fix_sex.txt")
df_asmqc_03_man$H1[grepl("_fa", df_asmqc_03_man$SAMPLE)] <- df_asmqc_03_man$SAMPLE[grepl("_fa", df_asmqc_03_man$SAMPLE)] %>% lapply(function(x){
  paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_temp[grepl(paste0(x, "_hap1.fasta"), df_temp)])
}) %>% unlist

df_asmqc_03_man$H2[grepl("_fa", df_asmqc_03_man$SAMPLE)] <- df_asmqc_03_man$SAMPLE[grepl("_fa", df_asmqc_03_man$SAMPLE)] %>% lapply(function(x){
  paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_temp[grepl(paste0(x, "_hap2.fasta"), df_temp)])
}) %>% unlist

write(df_asmqc_03_man, "C:/docs/autism/asm_qc/asm_qc_man_20240710.tsv")
```


```{r}
df01 <- read.table("C:/docs/autism/asm_qc/files.fix_sex.txt")
df02 <- data.frame(sample = df01$V1 %>% str_extract("(?<=_fa/).*(?=_hap..fasta)") %>% u)

get_fix_sex <- function(x, hap){
  mypath <- paste0("fix_sex/results/", x, "/", x, "_", hap, ".fasta")
  if(mypath %in% df01$V1){return(mypath)}
  mypath <- paste0("fix_sex/results_old/", x, "/", x, "_", hap, ".fasta")
  if(mypath %in% df01$V1){return(mypath)}
  return("")
}

df02$hap1 <- df02$sample %>%  lapply(function(x){get_fix_sex(x, "hap1")})
df02$hap2 <- df02$sample %>%  lapply(function(x){get_fix_sex(x, "hap2")})

df02$hap1 <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df02$hap1)
df02$hap2 <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df02$hap2)

colnames(df02) <- c("NAME", "HAP1", "HAP2")
write(df02, "C:/docs/autism/asm_qc/pav.male.refresh.20240624.tsv")
```



```{r}
df_asmqc_01 <- read.table("C:/docs/autism/asm_qc/asmqc.batch01.man", header=TRUE)
df_asmqc_02 <- read.table("C:/docs/autism/asm_qc/asmqc.batch02.man", header=TRUE)

df_toasmqc <- df0[df0$asm_qc_hap1 == "" & df0$hap1 !="", ]

```

```{r}
my_comments <- df0$asm_qc_hap1[!grepl("/", df0$asm_qc_hap1)] %>% u
df_pav_man <- df0[!df0$asm_qc_hap1 %in% my_comments, c("sample", "asm_qc_hap1", "asm_qc_hap2")]
colnames(df_pav_man) <- c("NAME", "HAP1", "HAP2")
df_pav_man$HAP1 <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_pav_man$HAP1)
df_pav_man$HAP2 <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_pav_man$HAP2)



write(df_pav_man[df_pav_man$NAME %in% df00$SP_ID[df00$Sex=="M"], ], "C:/docs/autism/asm_qc/pav.male.refresh.20240618.tsv")
write(df_pav_man[df_pav_man$NAME %in% df00$SP_ID[df00$Sex=="F"], ], "C:/docs/autism/asm_qc/pav.female.refresh.20240618.tsv")
```









```{bash}
find pbsv/GRCH38_noY/results/ | grep .vcf.gz$ | xargs -I {} sh -c 'echo -n "{}/$(zcat {} | grep -v "^#" | wc -l) \n"' | awk -F "/" '{print $4 "\t" $5 "\t" $6}' > pbsv/counts_GRCH38_noY.tsv
```

```{r}
df0 <- fread("C:/docs/autism/asm_qc/counts_GRCH38.tsv") %>% as.data.frame()
samples <- df0$V1 %>% u
list_by_sample <- lapply(samples, function(x){df0[df0$V1==x, ]})
df_by_sample <- lapply(list_by_sample, function(x){
  count_bnd <- ifelse(1==length(grep("_bnd.vcf.gz", x$V2)), x$V3[grepl("_bnd.vcf.gz", x$V2)], "NA")   
  count_dup <- ifelse(1==length(grep("_dup.vcf.gz", x$V2)), x$V3[grepl("_dup.vcf.gz", x$V2)], "NA")   
  count_sv  <- ifelse(1==length(grep("_sv.vcf.gz" , x$V2)),  x$V3[grepl("_sv.vcf.gz" , x$V2)], "NA")   
  data.frame(SAMPLE=x$V1[1], BND=count_bnd, DUP=count_dup, SV=count_sv, TOTAL=count_sv+count_dup+count_bnd)
}) %>% do.call(rbind, .) %>% as.data.frame()
df_by_sample <- df_by_sample[order(df_by_sample$SAMPLE), ]
write(df_by_sample, "C:/docs/autism/asm_qc/pbsv_GRCH38_counts.tsv")

df1 <- fread("C:/docs/autism/asm_qc/counts_GRCH38_noY.tsv") %>% as.data.frame()
samples <- df1$V1 %>% u
list_by_sample <- lapply(samples, function(x){df1[df1$V1==x, ]})
df_by_sample <- lapply(list_by_sample, function(x){
  count_bnd <- ifelse(1==length(grep("_bnd.vcf.gz", x$V2)), x$V3[grepl("_bnd.vcf.gz", x$V2)], "NA")   
  count_dup <- ifelse(1==length(grep("_dup.vcf.gz", x$V2)), x$V3[grepl("_dup.vcf.gz", x$V2)], "NA")   
  count_sv  <- ifelse(1==length(grep("_sv.vcf.gz" , x$V2)),  x$V3[grepl("_sv.vcf.gz" , x$V2)], "NA")   
  data.frame(SAMPLE=x$V1[1], BND=count_bnd, DUP=count_dup, SV=count_sv, TOTAL=count_sv+count_dup+count_bnd)
}) %>% do.call(rbind, .) %>% as.data.frame()
df_by_sample <- df_by_sample[order(df_by_sample$SAMPLE), ]
write(df_by_sample, "C:/docs/autism/asm_qc/pbsv_GRCH38_counts_noY.tsv")

```



```{r}
MY_NAME <- "ISAAC"


df_issac <- data.frame(
  letters = str_split(MY_NAME, "") %>% unlist
)


hash_codons <- HashMap$new()
hash_codons$put("I", "ATT,ATC,ATA")
hash_codons$put("S", "AGT,AGC,TCT,TCA,TCC,TCG")
hash_codons$put("A", "GCA,GCT,GCC,GCG")
hash_codons$put("C", "TGT,TGC")
hash_codons$put("T", "ACT,ACA,ACG,ACC")
hash_codons$put("R", "CGT,CGA,CGC,CGG")
hash_codons$put("H", "CAT,CAC")
hash_codons$put("W", "TGG")
hash_codons$put("N", "AAT,AAc")
hash_codons$put("G", "GGT,GGA,GGC,GGG")
hash_codons$put("Y", "TAT,TAC")

df_issac$codons <- lapply(df_issac$letters, function(x){hash_codons$get(x)})

df_issac$list <- lapply(df_issac$codons, function(x){str_split(x, ",") %>% unlist})

df_temp <- expand.grid(df_issac$list, stringsAsFactors = FALSE)
name_strings <- lapply(seq_down(df_temp), function(x){
  paste0(df_temp[x, ]) %>% paste(collapse = "") %>% str_to_lower()
}) %>% unlist 

library(seqinr)
df_output <- data.frame(dna = name_strings)
df_output$temp <- lapply(name_strings, function(x){unlist(str_split(x, ""))})
df_output$frame153 <- lapply(df_output$temp, function(x){translate(x, frame=0, sens="F") %>% paste(collapse = "")}) %>% unlist
df_output$frame253 <- lapply(df_output$temp, function(x){translate(x, frame=1, sens="F") %>% paste(collapse = "")}) %>% unlist
df_output$frame353 <- lapply(df_output$temp, function(x){translate(x, frame=2, sens="F") %>% paste(collapse = "")}) %>% unlist
df_output$frame135 <- lapply(df_output$temp, function(x){translate(x, frame=0, sens="R") %>% paste(collapse = "")}) %>% unlist
df_output$frame235 <- lapply(df_output$temp, function(x){translate(x, frame=1, sens="R") %>% paste(collapse = "")}) %>% unlist
df_output$frame335 <- lapply(df_output$temp, function(x){translate(x, frame=2, sens="R") %>% paste(collapse = "")}) %>% unlist
df_output$temp <- NULL
write(df_output, "C:/Users/iwong1/Downloads/isaac.txt")
```




```{r}
df0 <- read.table("C:/docs/autism/asm_qc/files.txt")

df1 <- data.frame(sample = df0$V1 %>% str_extract("(?<=n50/).*(?=_hap..n50.stats$)") %>% u)
df1$hap1_path <- df1$sample %>% lapply(function(x){
  temp1 <- df0$V1[grepl(paste0(x, ".*hap1"), df0$V1)]
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/contig_stats/n50/",x,"_hap1.n50.stats")
  if(mypath %in% df0$V1){return(mypath)}
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc_GROUP2/QC_results/contig_stats/n50/",x,"_hap1.n50.stats")
  if(mypath %in% df0$V1){return(mypath)}
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results_old/contig_stats/n50/",x,"_hap1.n50.stats")
  if(mypath %in% df0$V1){return(mypath)}
  return("")
}) %>% unlist

df1$hap2_path <- df1$sample %>% lapply(function(x){
  temp1 <- df0$V1[grepl(paste0(x, ".*hap1"), df0$V1)]
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/contig_stats/n50/",x,"_hap2.n50.stats")
  if(mypath %in% df0$V1){return(mypath)}
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc_GROUP2/QC_results/contig_stats/n50/",x,"_hap2.n50.stats")
  if(mypath %in% df0$V1){return(mypath)}
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results_old/contig_stats/n50/",x,"_hap2.n50.stats")
  if(mypath %in% df0$V1){return(mypath)}
  return("")
}) %>% unlist

df1$n50_hap1 <- df1$hap1_path %>% lapply(function(x){
  temp1 <- readLines(x)
  temp1[grepl("N50", temp1)] %>% str_extract("(?<=N50 \\(Mbp\\):).*") %>% str_trim()
}) %>% unlist

df1$n50_hap2 <- df1$hap2_path %>% lapply(function(x){
  temp1 <- readLines(x)
  temp1[grepl("N50", temp1)] %>% str_extract("(?<=N50 \\(Mbp\\):).*") %>% str_trim()
}) %>% unlist


write.table()
```


```{r}
df0 <- read.table("C:/docs/autism/asm_qc/files1.txt")

df1 <- data.frame(sample = df0$V1 %>% str_extract("(?<=_(fa|mo|p.|s.)/).*(?=.qv)") %>% u)
df1$qv_path <- df1$sample %>% lapply(function(x){
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/merqury/results/", x, "/",x,".qv")
  if(mypath %in% df0$V1){return(mypath)}
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc_GROUP2/QC_results/merqury/results/", x, "/",x,".qv")
  if(mypath %in% df0$V1){return(mypath)}
  mypath <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results_old/merqury/results/", x, "/",x,".qv")
  if(mypath %in% df0$V1){return(mypath)}
  return("")
}) %>% unlist

df1$qv_hap1 <- df1$qv_path %>% lapply(function(x){
  temp1 <- read.table(x)
  if(any(grepl("_hap1", temp1$V1))){
    temp1$V4[grepl("_hap1", temp1$V1)]    
  } else {
    return("")
  }
}) %>% unlist

df1$qv_hap2 <- df1$qv_path %>% lapply(function(x){
  temp1 <- read.table(x)
  if(any(grepl("_hap2", temp1$V1))){
    temp1$V4[grepl("_hap2", temp1$V1)]    
  } else {
    return("")
  }
}) %>% unlist



```

ataagcgcggcttgt	ISAAC	*ARL	KRGL	TSRAY	QAAL	KPRL
acacggatttcgcatgcacgcgcttat     TRISHARAY       HGFRMHAL        TDFACTRL        ISACMRNPC       *ARACEIR        KRVHAKSV

/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/contamination_screening/results/12832_p1_hap1/fasta/12832_p1_hap1.fasta


software/pipelines/ss_analyses/202004
- align stranseq data to the refrence (bwa or winnowmap)
run breakpointr and/or saarclust

requests: 
- add strandphsaser tool
- add automatic library selection
- add strandseq pool genotyping

strandseq protocol
- merge bams 
  - samtools merge -@4 merged.bam {input.bam}
- call veraibale sites with bcftools
  - bcftools mpileup etc
- 
- get watson crick (WC) regions in each strandseq seq cell
  - exportRegions(datapath= {breakpointr.data})
  - trying to ask which sample differes the lest from the given cell haplotype, then that is the sample of origin for the given cell. 
  
will need to meet with david seperately later to go over pipeline?????????????????????????
   
GT:AD:AO:DP:PL:QA:QR:RO
0/0:8,0:0:9:0,24,165:0:204:8
0/1:8,4:4:13:63,0,132:115:218:8 
0/1:12,4:4:17:38,0,220:110:329:12
```{r}
df_temp <- c("GT:AD:AO:DP:PL:QA:QR:RO",
"0/0:8,0:0:9:0,24,165:0:204:8",
"0/1:8,4:4:13:63,0,132:115:218:8", 
"0/1:12,4:4:17:38,0,220:110:329:12") %>% str_split(":") %>% do.call(rbind, .) %>% as.data.frame()
```








```{r}
df0 <- fread("C:/docs/lab_meeting_presentation/0_summary_stats.txt") %>% as.data.frame()
list_by_family <- lapply(u(df0$Family_ID), function(x){
  return(df0[df0$Family_ID == x, ])
})


g1 <- ggplot(df0, aes(x=asm_n50_hap1, y=asm_qv_hap1)) + 
  geom_point(aes(color=Family_ID )) + 
  ggtitle("Hap1 HiFi Assembly") + 
  xlab("Contig N50 (Mbp)") + 
  ylab("Quality Value (Merqury)")
g1

g2 <- ggplot(df0, aes(x=asm_n50_hap2, y=asm_qv_hap2)) + 
  geom_point(aes(color=Family_ID )) + 
  ggtitle("Hap2 HiFi Assembly") + 
  xlab("Contig N50 (Mbp)") + 
  ylab("Quality Value (Merqury)")
g2


g3 <- ggplot(df0, aes(x=read_cov_X, y=`PBSV_SV>49bp`)) + 
  geom_point(aes(color=Family_ID )) + 
  ggtitle("Read Coverage vs PBSV Calls") + 
  xlab("Read Coverage") + 
  ylab("PBSV Calls > 49bp") + 
  scale_x_continuous(trans="log10") + 
  scale_y_continuous(trans="log10") 
g3


g4 <- ggplot(df0, aes(x=asm_n50_hap1, y=`PAV_SV>49bp`)) + 
  geom_point(aes(color=Family_ID )) + 
  ggtitle("Assembly N50 vs PAV Calls") + 
  xlab("Assembly N50") + 
  ylab("PAV Calls > 49bp") + 
  scale_x_continuous(trans="log10") 
g4



g5 <- ggplot(df0[df0$`PAV_SV>49bp` < 22000, ], aes(x=asm_n50_hap1, y=asm_qv_hap1)) + 
  geom_point(aes(color=SP_ID )) + 
  ggtitle("Assembly N50 vs PAV Calls") + 
  xlab("Assembly N50") + 
  ylab("PAV Calls > 49bp") + 
  scale_x_continuous(trans="log10") 
g5




```



# Measure X Coverage
```{r}
FAI_HAP1_FILE <- "C:/docs/autism/fix_sex/BK143_fa_hap1.fasta.fai"
FAI_HAP2_FILE <- "C:/docs/autism/fix_sex/BK143_fa_hap2.fasta.fai"
PAF_HAP1_FILE <- "C:/docs/autism/fix_sex/BK143_fa_hap1.paf"
PAF_HAP2_FILE <- "C:/docs/autism/fix_sex/BK143_fa_hap2.paf"
REASSIGN_FILE <- "C:/docs/autism/fix_sex/BK143_fa_reassign.tab"
CTG_HAP1_FILE <- "C:/docs/autism/fix_sex/BK143_fa_hap1.fasta.contigs"
CTG_HAP2_FILE <- "C:/docs/autism/fix_sex/BK143_fa_hap2.fasta.contigs"

df_fai_1 <- read.table(FAI_HAP1_FILE, header=FALSE, sep="\t")
df_fai_2 <- read.table(FAI_HAP2_FILE, header=FALSE, sep="\t")

df_paf_1 <- fread(PAF_HAP1_FILE, header=FALSE, fill=TRUE) %>% as.data.frame()
df_paf_2 <- fread(PAF_HAP2_FILE, header=FALSE, fill=TRUE) %>% as.data.frame()

df_reass <- read.table(REASSIGN_FILE, header=TRUE, sep="\t")

df_ctg_1 <- read.table(CTG_HAP1_FILE, header = FALSE)$V1 %>% str_remove(">")
df_ctg_2 <- read.table(CTG_HAP2_FILE, header = FALSE)$V1 %>% str_remove(">")


table(df_reass$ctg[df_reass$dest=="hap1"] %in% df_ctg_1)
table(df_reass$ctg[df_reass$dest=="hap1"] %in% df_ctg_2)
table(df_reass$ctg[df_reass$dest=="hap2"] %in% df_ctg_1)
table(df_reass$ctg[df_reass$dest=="hap2"] %in% df_ctg_2)


df_reass$final_location <- lapply(seq_down(df_reass), function(x){
  results <- ""
  
  if(grepl(df_reass$ctg[x], df_ctg_1) %>% any){
    results <- c(results, "HAP1")
  }
  if(grepl(df_reass$ctg[x], df_ctg_2) %>% any){
    results <- c(results, "HAP2")
  }
  return(results)
}) 


```




```{r}
library(GenomicRanges)
library(IRanges)

PAF_FILE <- "C:/docs/autism/fix_sex/BK143_fa_hap1.paf"
df0 <- fread(PAF_FILE, header=FALSE, fill=TRUE) %>% as.data.frame()
df0$V24 <- NULL
df0$V25 <- NULL

df1 <- lapply(df0$V1 %>% u, function(x){
  temp1 <- df0[df0$V1==x, ]
  AS_SCORE <- temp1$V15 %>% str_remove("AS:i:") %>% as.numeric()
  return(temp1[which(AS_SCORE==max(AS_SCORE)), ])
}) %>% do.call(rbind, .) 

gr1 <- GRanges(seqnames = df1$V6, 
               ranges=IRanges(start=df1$V8,
                              end=df1$V9))
gr1 <- sortSeqlevels(gr1)
gr1 <- sort(gr1)

df2 <- reduce(gr1) %>% as.data.frame(stringsAsFactors=FALSE)
df2$seqnames <- df2$seqnames %>% as.character()


```


```{r}
df0 <- fread("C:/docs/autism/asm_qc/chromosome_coverage_percentage_fixed.tsv") %>% as.data.frame()
lapply(2:25, function(x){
  
  hist(df0[[x]]) %>% plot
  
})
```



```{r}
df01 <- data.frame(SAMPLE=all_samples, 
                   ROLE=all_samples %>% ifelse(grepl("_p.", all_samples), "proband", 
                                               ifelse(grepl("_s.", all_samples), "sibling", 
                                                      ifelse(grepl("_mo", all_samples), "mother", "father"))))
```


```{r}
df0 <- fread("C:/docs/autism/files_paf.fofn", header=FALSE) %>% as.data.frame()
df_paf1 <- data.frame(paf=df0$V1[grepl("_hap1", df0$V1)])
df_paf2 <- data.frame(paf=df0$V1[grepl("_hap2", df0$V1)])
df_paf1$sample <- df_paf1$paf %>% bname %>% str_remove("_hap.*")
df_paf2$sample <- df_paf2$paf %>% bname %>% str_remove("_hap.*")

df1 <- data.frame(SAMPLE = df_paf1$sample[df_paf1$sample %in% df_paf2$sample] %>% sort %>% u)

df1$ROLE <- ifelse(grepl("_p.", df1$SAMPLE), "proband", 
                   ifelse(grepl("_s.", df1$SAMPLE), "sibling", 
                          ifelse(grepl("_mo", df1$SAMPLE), "mother", "father")))

df1$FAMILY <- df1$SAMPLE %>% str_remove("_..$")
df1 <- df1[order(df1$FAMILY), ]

df1$HAP1 <- lapply(df1$SAMPLE, function(x){
  df_paf1$paf[df_paf1$sample==x][1]
}) %>% unlist

df1$HAP2 <- lapply(df1$SAMPLE, function(x){
  df_paf2$paf[df_paf2$sample==x][1]
}) %>% unlist

temp1 <- srt(df1$FAMILY)
df2 <- df1[df1$FAMILY %in% names(temp1)[temp1==4], ]

write.table(df2, "C:/docs/autism/concat_chrom.man.test.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)


df3 <- data.frame(SAMPLE=c(paste0(df2$SAMPLE), paste0(df2$SAMPLE)), 
           ROLE="sibling", 
           FAMILY=c(paste0(df2$FAMILY), paste0(df2$FAMILY)),
           HAP1=c(df2$HAP1, df2$HAP2),
           HAP2=c(df2$HAP1, df2$HAP2))
df3 <- df3[!grepl("(_fa)|(_mo)", df3$SAMPLE), ]
write.table(df3, "C:/docs/autism/man.ver2.test.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)


df4 <- data.frame(SAMPLE=df2$SAMPLE[df2$ROLE=="sibling"])

```



```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats")
df01 <- df00 %>% as.data.frame()
df02 <- df01[grepl("batch1", df01$Deposit_HiFi), ]

df02_pro <- df02[grep("_p", df02$SP_ID), ]
df02_sib <- df02[grep("_s", df02$SP_ID), ]

df0 <- fread("C:/docs/autism/paf/hg38_paf.txt", header=FALSE) %>% as.data.frame()
df_paf1 <- data.frame(paf=df0$V1[grepl("_hap1", df0$V1)])
df_paf2 <- data.frame(paf=df0$V1[grepl("_hap2", df0$V1)])
df_paf1$sample <- df_paf1$paf %>% bname %>% str_remove("_hap.*")
df_paf2$sample <- df_paf2$paf %>% bname %>% str_remove("_hap.*")

# df02_pro <- df02_sib
df_pro <- data.frame(SAMPLE = df02_pro$SP_ID %>% paste0(rep(c("hap1", "hap2"), each=nrow(df02_pro))),
                     ROLE = paste0(rep(c("hap1", "hap2"), each=nrow(df02_pro))))
df_pro$FAMILY <- df_pro$SAMPLE %>% str_extract(".*(?=_)")
df_pro$HAP1 <- seq_down(df_pro) %>% lapply(function(x){
  sample_id <- df_pro$SAMPLE[x] %>% str_remove(df_pro$ROLE[x])
  if(df_pro$ROLE[x] == "hap1"){
    rval <- df_paf1$paf[df_paf1$sample==sample_id]
  } else {
    rval <- df_paf2$paf[df_paf2$sample==sample_id]
  }
  
  if(length(rval)==0){
    return("NA")
  } else {
    return(rval)
  }
}) %>% unlist

df_pro$HAP2 <- seq_down(df_pro) %>% lapply(function(x){
  sample_id <- df_pro$SAMPLE[x] %>% str_remove(df_pro$ROLE[x])
  if(df_pro$ROLE[x] == "hap1"){
    rval <- df_paf1$paf[df_paf1$sample==sample_id]
  } else {
    rval <- df_paf2$paf[df_paf2$sample==sample_id]
  }
  
  if(length(rval)==0){
    return("NA")
  } else {
    return(rval)
  }
}) %>% unlist

write.table(df_pro, "C:/docs/autism/paf/concat_chrom_pro.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)
# write.table(df_pro[df_pro$HAP1!="NA", ], "C:/docs/autism/paf/concat_chrom_sib.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)
```


```{r}
df_col <- read.table("C:/docs/etc/my_colors_most_uniq.txt", header=FALSE, sep="\t")
df_col$xkcd <- paste0("'xkcd:", df_col$V1, "'")

paste(df_col$xkcd, collapse = ", ")

```

```{r}

files <- list.files(".", full.names = TRUE, recursive = TRUE, pattern = ".tsv")
df01 <- lapply(files, function(x){
  df_temp <- fread(x, header=FALSE) %>% as.data.frame()
  df_temp <- df_temp[df_temp$V7 >= 0.95 & df0$V4 < 3, ]
  rval <- data.frame(sample = bname(x), 
                     coverage = sum(df1$V3 - df1$V2))
  return(rval)
}) %>% do.call(rbind, .)
write.table(df01, )

```

```{r}
df0 <- fread("C:/docs/autism/batch_1_samples.txt") %>% as.data.frame()
df1 <- data.frame(SAMPLE = paste0(rep(df0$sample, 2), rep(c("hap1", "hap2"), each=nrow(df0))),
                  ROLE = rep(c("hap1", "hap2"), each=nrow(df0)), 
                  FAMILY = rep(df0$sample, 2) %>% str_remove("_.."), 
                  HAP1 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/saffire_hg38/QC_results/saffire/results/", rep(df0$sample, 2), "_", rep(c("hap1", "hap2"), each=nrow(df0)), "/alignments/", rep(df0$sample, 2), "_", rep(c("hap1", "hap2"), each=nrow(df0)), ".minimap2.paf"))
df1$HAP2 <- df1$HAP1

write.table(df1, "C:/docs/autism/concat_chrom_batch_1_samples.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```




```{r}
files <- list.files("C:/docs/autism/coverage/results_cyto/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
COLNAMES <- c("chr", "start", "end", "name", "gieStain", "n_features", "n_bases", "length", "percent")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX")

df_coordinates <- fread("C:/docs/autism/coverage/cyto.bed") %>% as.data.frame()
df_coordinates <- df_coordinates[!grepl("_|M|Y", df_coordinates$`#chrom`), ]

chr_lengths <- lapply(df_coordinates$`#chrom` %>% u, function(x){
  max(df_coordinates$end[df_coordinates$`#chrom`==x]) %>% as.numeric
})
names(chr_lengths) <- df_coordinates$`#chrom` %>% u
df_map <- data.frame(chr = CHR_ORDER, 
                     end = chr_lengths[CHR_ORDER] %>% unlist %>% unname())
df_map$cum_end <- c(0, df_map$end %>% as.numeric() %>% cumsum())[-24]

df_coordinates$plot_start <- lapply(seq_down(df_coordinates), function(x){
  df_coordinates$start[x] + df_map$cum_end[which(df_map$chr==df_coordinates$`#chrom`[x])]
}) %>% unlist

df_coordinates$plot_end <- lapply(seq_down(df_coordinates), function(x){
  df_coordinates$end[x] + df_map$cum_end[which(df_map$chr==df_coordinates$`#chrom`[x])]
}) %>% unlist

df_coordinates$hash <- paste0(df_coordinates$`#chrom`, ":", df_coordinates$start, "_", df_coordinates$end, "__", df_coordinates$name)
hashToStart <- HashMap$new()
hashToEnd <- HashMap$new()
hashToStart$populate(df_coordinates, "hash", "plot_start")
hashToEnd$populate(df_coordinates, "hash", "plot_end")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  df_temp$SOURCE <- x
  return(df_temp)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- df00$sample %>% str_extract("(?<=_)..(?=_)")
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M|Y", df00$chr), ]
df00 <- df00[df00$n_features < 4, ]

df00$hash <- paste0(df00$chr, ":", df00$start, "_", df00$end, "__", df00$name)
df00$plot_start <- df00$hash %>% lapply(function(hash){hashToStart$get(hash)}) %>% unlist
df00$plot_end   <- df00$hash %>% lapply(function(hash){hashToEnd$get(hash)}) %>% unlist
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


g1 <- ggplot(df01, aes(x = plot_start, y = percent, colour = Role, shape = HAP)) +
  geom_point(position = position_dodge(0.1), alpha=0.7) + 
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=df_map$cum_end, labels = df_map$chr) + 
  ylab("contig coverage") + 
  xlab("Genomic Location / Cytogenetic Bands") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + 
  geom_vline(xintercept=df_map$cum_end, linetype="dotted") + 
  coord_cartesian(xlim=c(min(df00$plot_start), max(df00$plot_end))); g1
ggsave("C:/docs/autism/coverage/coverage_batch_02.pdf", g1, width = 15, height = 7, units = "in")
```


```{r}
CURR_GENOME <- "hg38"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/autism/coverage/results_hg38/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    if(x == "chrX"){
      to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
        BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]   
        BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
        BOOL_OL <- BOOL_START & BOOL_END
        return(which(BOOL_OL))
      }) %>% unlist %>% u
      temp_chr <- temp_chr[-to_remove, ]
    }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- ifelse(grepl("^H", df00$sample), "p1", df00$sample %>% str_extract("(?<=_)..(?=_)"))
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")
df00 <- df00[!df00$role %in% c("fa", "mo"), ]

df00$family <- ifelse(grepl("^H", df00$sample), 
                      df00$sample %>% str_remove("_hap."), 
                      df00$sample %>% str_extract("^.....(?=_)")) 
TARGET_FAMILIES <- read.table("C:/docs/autism/coverage/batch_1_families_main.tsv")
df00 <- df00[df00$family %in% TARGET_FAMILIES$V1, ]

write.table(df00, "C:/docs/autism/coverage/hg38_coverage_ggplot_20241231.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/autism/coverage/hg38_coverage_main_20241231.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


df_sex <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()
df01$sex <- lapply(df01$sample %>% str_remove("_hap."), function(x){
  df_sex$Sex[df_sex$SP_ID==x]
}) %>% unlist

df01_main <- df01[!grepl("X|Y", df01$chr), ]
df01_x <- df01[grepl("X", df01$chr) & !(df01$sex=="M"  & grepl("hap1", df01$sample)) , ]
df01_y <- df01[grepl("Y", df01$chr) & df01$sex=="M" & grepl("hap1", df01$sample), ]
df02 <- rbind(df01_main, rbind(df01_x, df01_y))

df_save <- df02
# df02 <- df02[!grepl("^H", df02$sample), ]

# df02 <- df01
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.003, alpha=0.8) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly hg38 Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/autism/coverage/coverage_hg38_window_20241231.pdf", g2, width = 15, height = 7, units = "in")
```

```{r}
CURR_GENOME <- "t2t"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/autism/coverage/results_T2T_window/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    if(x == "chrX"){
      to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
        BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]   
        BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
        BOOL_OL <- BOOL_START & BOOL_END
        return(which(BOOL_OL))
      }) %>% unlist %>% u
      temp_chr <- temp_chr[-to_remove, ]
    }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- ifelse(grepl("^H", df00$sample), "p1", df00$sample %>% str_extract("(?<=_)..(?=_)"))
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")
df00 <- df00[!df00$role %in% c("fa", "mo"), ]

df00$family <- ifelse(grepl("^H", df00$sample), 
                      df00$sample %>% str_remove("_hap."), 
                      df00$sample %>% str_extract("^.....(?=_)")) 
TARGET_FAMILIES <- read.table("C:/docs/autism/coverage/batch_1_families_main.tsv")
df00 <- df00[df00$family %in% TARGET_FAMILIES$V1, ]

write.table(df00, "C:/docs/autism/coverage/T2T_coverage_ggplot_20241231.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/autism/coverage/T2T_coverage_main_20241231.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


df_sex <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()
df01$sex <- lapply(df01$sample %>% str_remove("_hap."), function(x){
  df_sex$Sex[df_sex$SP_ID==x]
}) %>% unlist

df01_main <- df01[!grepl("X|Y", df01$chr), ]
df01_x <- df01[grepl("X", df01$chr) & !(df01$sex=="M"  & grepl("hap1", df01$sample)) , ]
df01_y <- df01[grepl("Y", df01$chr) & df01$sex=="M" & grepl("hap1", df01$sample), ]
df02 <- rbind(df01_main, rbind(df01_x, df01_y))

df_save <- df02
# df02 <- df02[!grepl("^H", df02$sample), ]

# df02 <- df01
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.003, alpha=0.8) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly T2T Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/autism/coverage/coverage_T2T_window_20241231.pdf", g2, width = 15, height = 7, units = "in")
```



```{r}
CURR_GENOME <- "t2t"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/autism/coverage/results_T2T_window/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    if(x == "chrX"){
      to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
        BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]   
        BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
        BOOL_OL <- BOOL_START & BOOL_END
        return(which(BOOL_OL))
      }) %>% unlist %>% u
      print(to_remove)
      temp_chr <- temp_chr[-to_remove, ]
    }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- df00$sample %>% str_extract("(?<=_)..(?=_)")
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")
df00 <- df00[!df00$role %in% c("fa", "mo"), ]

df00$family <- df00$sample %>% str_extract("^.....(?=_)")
TARGET_FAMILIES <- read.table("C:/docs/autism/coverage/batch_1_families_main.tsv")
df00 <- df00[df00$family %in% TARGET_FAMILIES$V1, ]

write.table(df00, "C:/docs/autism/coverage/T2T_coverage_ggplot3.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/autism/coverage/T2T_coverage_main3.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


df_sex <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()
df01$sex <- lapply(df01$sample %>% str_remove("_hap."), function(x){
  df_sex$Sex[df_sex$SP_ID==x]
}) %>% unlist

df01_main <- df01[!grepl("X|Y", df01$chr), ]
df01_x <- df01[grepl("X", df01$chr) & !(df01$sex=="M"  & grepl("hap1", df01$sample)) , ]
df01_y <- df01[grepl("Y", df01$chr) & df01$sex=="M" & grepl("hap1", df01$sample), ]
df02 <- rbind(df01_main, rbind(df01_x, df01_y))

# df02 <- df01
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.003, alpha=0.8) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly T2T Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/autism/coverage/coverage_T2T_window3.pdf", g2, width = 15, height = 7, units = "in")
```

```{r}
files <- list.files("C:/docs/autism/coverage/results_T2T_window/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")
  
df_temp1 <- fread(files[1]) %>% as.data.frame()
colnames(df_temp1) <- COLNAMES
df_temp1 <- df_temp1[df_temp1$chr == "chrX", ]

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  df_temp <- df_temp[df_temp$chr == "chrX", ]
  rval <- ifelse((df_temp$percent >= 0.95) & (df_temp$n_features <= 3), 1, 0)
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame() 
colnames(df00) <- paste0(df_temp1$chr, ":", df_temp1$start, "-", df_temp1$end)
df00$sample <- files %>% bname %>% str_remove(".tsv$")
my_sums_all <- colSums(df00[, -which(colnames(df00)=="sample")])
df_all <- my_sums_all %>% as.data.frame()
colnames(df_all) <- "all"
df_all$coord <- rownames(df_all)

df01 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  df_temp <- df_temp[df_temp$chr == "chrX", ]
  rval <- ifelse(df_temp$percent >= 0.95, 1, 0)
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame() 
colnames(df01) <- paste0(df_temp1$chr, ":", df_temp1$start, "-", df_temp1$end)
df01$sample <- files %>% bname %>% str_remove(".tsv$")
my_sums_percent <- colSums(df01[, -which(colnames(df01)=="sample")])
df_percent <- my_sums_percent %>% as.data.frame()
colnames(df_percent) <- "percent"
df_percent$coord <- rownames(df_percent)

df02 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  df_temp <- df_temp[df_temp$chr == "chrX", ]
  rval <- ifelse(df_temp$n_features <= 3, 1, 0)
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame() 
colnames(df02) <- paste0(df_temp1$chr, ":", df_temp1$start, "-", df_temp1$end)
df02$sample <- files %>% bname %>% str_remove(".tsv$")
my_sums_contigs <- colSums(df02[, -which(colnames(df02)=="sample")])
df_contigs <- my_sums_contigs %>% as.data.frame()
colnames(df_contigs) <- "contigs"
df_contigs$coord <- rownames(df_contigs)


df03 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  df_temp <- df_temp[df_temp$chr == "chrX", ]
  rval <- df_temp$percent
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame() 
colnames(df03) <- paste0(df_temp1$chr, ":", df_temp1$start, "-", df_temp1$end)
df03[df03 >=0.95] <- NA
df03$sample <- files %>% bname %>% str_remove(".tsv$")
my_sums_percent_mean <- colMeans(df03[, -which(colnames(df03)=="sample")], na.rm = TRUE)
df_percent_mean <- my_sums_percent_mean %>% as.data.frame()
colnames(df_percent_mean) <- "percent_mean"
df_percent_mean$coord <- rownames(df_percent_mean)


df_final <- merge(df_all, merge(df_contigs, df_percent, by = "coord"), by="coord")
df_final$coord_end <- df_final$coord %>% str_extract("(?<=-).*") %>% as.numeric()
df_final <- df_final[order(df_final$coord_end, decreasing = FALSE), ]
df_final$coord_end <- NULL

df_final_0 <- merge(df_final, df_percent_mean, by="coord", all.x = TRUE, all.y = TRUE)

df_final <- df_final_0

df_final$coord_end <- df_final$coord %>% str_extract("(?<=-).*") %>% as.numeric()
df_final <- df_final[order(df_final$coord_end, decreasing = FALSE), ]
df_final$coord_end <- NULL
write.table(df_final, "C:/docs/autism/coverage/chr_x_bins_coord_sorted.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
write.table(df_final[order(df_final$all, df_final$coord, decreasing = FALSE), ], "C:/docs/autism/coverage/chr_x_bins_count_sorted.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```






```{r}
my_fams <- read.table("C:/docs/autism/coverage/batch_1_families.txt", header = FALSE)

paste(my_fams$V1, collapse = "', '")

```

```{r}
df0 <- fread("C:/docs/autism/assemblies.txt2", header=FALSE) %>% as.data.frame()
df0$sample <- df0$V1 %>% str_extract("(?<=fasta/)....._..")
df1 <- df0[!grepl("(old|redo|_fa)", df0$V1), ]

dff <- fread("C:/docs/autism/assemblies.txt3", header=FALSE) %>% as.data.frame()
dff$sample <- dff$V1 %>% str_extract("(?<=results/)....._fa")

df2 <- rbind(df1, dff)

LRA <- "/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/"
child_paths <- ifelse(grepl("(_p.|_s.)", df2$sample), "trio/0.16.1/", "0.16.1/")
df2$destination <- paste0(LRA, df2$sample, "/assemblies/hifiasm/", child_paths, df2$V1 %>% bname)

script <- paste0("cp ", df2$V1, " ", df2$destination)
df_man <- data.frame(SAMPLE = df2$sample, 
                     SOURCE = df2$V1, 
                     DEST = df2$destination)
df_man <- df_man[order(df_man$SAMPLE), ]
df_man$SAMPLE <- paste0(df_man$SAMPLE, seq_down(df_man))


write.table(script, "C:/docs/autism/asm_mig.sh", col.names = FALSE, row.names = FALSE, quote=FALSE, sep="\t")
write.table(df_man, "C:/docs/autism/asm_mig.man", col.names = TRUE, row.names = FALSE, quote=FALSE, sep="\t")

```

```{r}
df0 <- fread("C:/docs/autism/rett_alns.fofn", header=FALSE) %>% as.data.frame()
df0$sample <- df0$V1 %>% bname %>% str_remove(".all.sorted.bam")
df0 <- df0[order(df0$V1), ]
df0$ref <- df0$V1 %>% lapply(function(x){
  unlist(str_split(x, "/"))[12]
}) %>% unlist
df0$dest_dir <- paste0("/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/", df0$sample, "/alignments/PacBio_HiFi/", df0$ref)
df0$final_bam <- paste0(df0$dest_dir, "/alignment.bam")
df0$final_bai <- paste0(df0$dest_dir, "/alignment.bam.bai")

script_dir <- paste0("mkdir -p ", df0$dest_dir)
script_cp_bam <- paste0("cp ", df0$V1, " ", df0$final_bam)
script_cp_bai <- paste0("cp ", df0$V1, ".bai ",  df0$final_bai)
script_remove_bam <- paste0("rm -rf ", df0$V1)
script_remove_bai <- paste0("rm -rf ", df0$V1, ".bai")
script_link_bam <- paste0("ln -s ", df0$final_bam, " ", df0$V1)
script_link_bai <- paste0("ln -s ", df0$final_bai, " ", df0$V1, ".bai")


final_script <- c("#!/bin/bash\nset -euxo pipefail", script_dir, script_cp_bam, script_cp_bai, script_remove_bam, script_remove_bai, script_link_bam, script_link_bai)
write.table(final_script, "C:/docs/autism/migrate_rett_samples.sh", sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/11414_mo/alignments/PacBio_HiFi/GRCh38


```{r}
df0 <- fread("C:/docs/autism/autism_samples.txt") %>% as.data.frame()
df0$i <- seq_along(df0)
df0 <- df0[grepl("_", df0$SP_ID), ]
df1 <- fread("C:/docs/autism/files.txt3", header=FALSE) %>% as.data.frame()
colnames(df1) <- "path"

df1$sample <- df1$path %>% lapply(function(x){
  temp1 <- str_split(x, "/") %>% unlist
  return(temp1[9])
}) %>% unlist

df1$hap <- ifelse(grepl("hap1", df1$path), "HAP1", ifelse(grepl("hap2", df1$path), "HAP2", "NONE"))
df1 <- df1[df1$hap != "NONE", ]
df1 <- df1[df1$sample %in% df0$SP_ID, ]
df1 <- df1[!grepl("p_ctg", df1$path), ]



df2 <- data.frame(SAMPLE = df0$SP_ID, 
                  HAP1 = lapply(df0$SP_ID, function(x){
                    temp1 <- df1$path[df1$sample == x & df1$hap == "HAP1"]
                    if(length(temp1)==1){
                      return(temp1)
                    } else {
                      return("toomany")
                    }
                  }) %>% unlist, 
                  HAP2 = lapply(df0$SP_ID, function(x){
                    temp1 <- df1$path[df1$sample == x & df1$hap == "HAP2"]
                    if(length(temp1)==1){
                      return(temp1)
                    } else {
                      return("toomany")
                    }
                  }) %>% unlist)


write.table(df2[df2$HAP1 != "toomany", ], "C:/docs/autism/asm_qc_lra_manifest.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

df3 <- df2[!grepl("_fa", df2$SAMPLE), ]



```


```{r}
dfr <- fread("C:/docs/autism/files.rettlike.asmqc.txt", header=FALSE) %>% as.data.frame()
dfr$sample <- dfr$V1 %>% bname %>% str_remove("_hap..fasta")
dfr$hap <- ifelse(grepl("hap1", dfr$V1), "HAP1", ifelse(grepl("hap2", dfr$V1), "HAP2", "NONE"))

df0 <- dfr
df0 <- df0[order(df0$sample), ]
df0$dest_dir <- paste0("/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/", df0$sample, "/assemblies/hifiasm/0.16.1")
df0$final_fasta <- paste0(df0$dest_dir, "/", bname(df0$V1))
df0$final_fai <- paste0(df0$final_fasta, ".fai")

script_dir <- paste0("mkdir -p ", df0$dest_dir)
script_cp_fasta <- paste0("cp ", df0$V1, " ", df0$final_fasta)
script_cp_fai <- paste0("cp ", df0$V1, ".fai ",  df0$final_fai)
script_remove_fasta <- paste0("rm -rf ", df0$V1)
script_remove_fai <- paste0("rm -rf ", df0$V1, ".fai")
script_link_fasta <- paste0("ln -s ", df0$final_fasta, " ", df0$V1)
script_link_fai <- paste0("ln -s ", df0$final_fai, " ", df0$V1, ".fai")


final_script <- c("#!/bin/bash\nset -euxo pipefail", script_dir, script_cp_fasta, script_cp_fai, script_remove_fasta, script_remove_fai, script_link_fasta, script_link_fai)
write.table(final_script, "C:/docs/autism/migrate_rett_asm_samples.sh", sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE)



rett_samples <- dfr$sample %>% u %>% sort
df_r <- data.frame(SAMPLE = rett_samples, 
                  HAP1 = lapply(rett_samples, function(x){
                    temp1 <- dfr$V1[dfr$sample == x & dfr$hap == "HAP1"]
                    return(temp1[1])
                  }) %>% unlist, 
                  HAP2 = lapply(rett_samples, function(x){
                    temp1 <- dfr$V1[dfr$sample == x & dfr$hap == "HAP2"]
                    return(temp1[1])
                  }) %>% unlist)

df_r$new_hap1 <- paste0("/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/", df_r$SAMPLE, "/assemblies/hifiasm/0.16.1/", bname(df_r$HAP1))
df_r$new_hap2 <- paste0("/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/", df_r$SAMPLE, "/assemblies/hifiasm/0.16.1/", bname(df_r$HAP2))


write.table(df_r[, c(1,4,5)], "C:/docs/autism/rett_asms_lra.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```



```{r}
df00 <- fread("C:/docs/autism/fa_asms.txt", header=FALSE) %>% as.data.frame()
df00$sample <- df00$V1 %>% lapply(function(x){
  temp1 <- str_split(x, "/") %>% unlist
  return(temp1[11])
}) %>% unlist 
```



```{r}
df00 <- fread("C:/docs/autism/asm_qc_lra_manifest_updated.tsv.txt") %>% as.data.frame()
df01 <- fread("C:/docs/autism/manifest.autism.aln.GRCh38.tsv") %>% as.data.frame()
df02 <- fread("C:/docs/autism/manifest.autism.aln.GRCh38_noY.tsv") %>% as.data.frame()
df03 <- fread("C:/docs/autism/rett-hg38-aln.txt", header=FALSE) %>% as.data.frame()
df04 <- fread("C:/docs/autism/rett-hg38-noY-aln.txt", header = FALSE) %>% as.data.frame()

df03$sample <- df03$V1 %>% lapply(function(x){
  unlist(str_split(x, "/"))[9]
}) %>% unlist
df04$sample <- df04$V1 %>% lapply(function(x){
  unlist(str_split(x, "/"))[9]
}) %>% unlist


all_samples <- u(c(df00$SAMPLE, df01$SAMPLE, df02$SAMPLE, df03$sample, df04$sample))

df_final <- data.frame(sample = sort(all_samples))
df_final$HAP1 <- df_final$sample %>% lapply(function(x){
  if(x %in% df00$SAMPLE){
    return(df00$HAP1[df00$SAMPLE == x])
  } else {
    return("missing_sequence_files")
  }
}) %>% unlist
df_final$HAP2 <- df_final$sample %>% lapply(function(x){
  if(x %in% df00$SAMPLE){
    return(df00$HAP2[df00$SAMPLE == x])
  } else {
    return("missing_sequence_files")
  }
}) %>% unlist
df_final$alignment <- df_final$sample %>% lapply(function(x){
  if(x %in% df02$SAMPLE){
    df02$BAM[df02$SAMPLE == x]
  } else if(x %in% df01$SAMPLE){
    df01$BAM[df01$SAMPLE == x]
  } else if(x %in% df03$sample){
    df03$V1[df03$sample == x]
  } else if(x %in% df04$sample){
    df04$V1[df04$sample == x]
  } else {
    return("ERROR")
  }
}) %>% unlist

colnames(df_final) <- c("SAMPLE", "ASM_QC_HAP1",  "ASM_QC_HAP2", "ALIGNMENT")
write.table(df_final, "C:/docs/autism/manifest.autism.asm.aln.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```


```{r}
CURR_GENOME <- "t2t"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/autism/coverage/results_t2t_exclude//", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    # if(x == "chrX"){
    #   to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
    #     BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]   
    #     BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
    #     BOOL_OL <- BOOL_START & BOOL_END
    #     return(which(BOOL_OL))
    #   }) %>% unlist %>% u
    #   # print(to_remove)
    #   temp_chr <- temp_chr[-to_remove, ]
    # }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- df00$sample %>% str_extract("(?<=_)..(?=_)")
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")
df00 <- df00[!df00$role %in% c("fa", "mo"), ]

df00$family <- df00$sample %>% str_extract("^.....(?=_)")
TARGET_FAMILIES <- read.table("C:/docs/autism/coverage/batch_1_families_main.tsv")
df00 <- df00[df00$family %in% TARGET_FAMILIES$V1, ]

write.table(df00, "C:/docs/autism/coverage/T2T_coverage_ggplot_exclude.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/autism/coverage/T2T_coverage_main_exclude.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


df_sex <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()
df01$sex <- lapply(df01$sample %>% str_remove("_hap."), function(x){
  df_sex$Sex[df_sex$SP_ID==x]
}) %>% unlist

df01_main <- df01[!grepl("X|Y", df01$chr), ]
df01_x <- df01[grepl("X", df01$chr) & !(df01$sex=="M"  & grepl("hap1", df01$sample)) , ]
df01_y <- df01[grepl("Y", df01$chr) & df01$sex=="M" & grepl("hap1", df01$sample), ]
df02 <- rbind(df01_main, rbind(df01_x, df01_y))

# df02 <- df01
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.003, alpha=0.8) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly T2T Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/autism/coverage/coverage_T2T_window_exclude.pdf", g2, width = 15, height = 7, units = "in")
```


```{r}
CURR_GENOME <- "hg38"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/autism/coverage/results_hg38_exclude/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    # if(x == "chrX"){
    #   to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
    #     BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]   
    #     BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
    #     BOOL_OL <- BOOL_START & BOOL_END
    #     return(which(BOOL_OL))
    #   }) %>% unlist %>% u
    #   print(to_remove)
    #   temp_chr <- temp_chr[-to_remove, ]
    # }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- df00$sample %>% str_extract("(?<=_)..(?=_)")
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")
df00 <- df00[!df00$role %in% c("fa", "mo"), ]

df00$family <- df00$sample %>% str_extract("^.....(?=_)")
TARGET_FAMILIES <- read.table("C:/docs/autism/coverage/batch_1_families_main.tsv")
df00 <- df00[df00$family %in% TARGET_FAMILIES$V1, ]

write.table(df00, "C:/docs/autism/coverage/hg38_coverage_ggplot_exclude.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/autism/coverage/hg38_coverage_main_exclude.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


df_sex <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()
df01$sex <- lapply(df01$sample %>% str_remove("_hap."), function(x){
  df_sex$Sex[df_sex$SP_ID==x]
}) %>% unlist

df01_main <- df01[!grepl("X|Y", df01$chr), ]
df01_x <- df01[grepl("X", df01$chr) & !(df01$sex=="M"  & grepl("hap1", df01$sample)) , ]
df01_y <- df01[grepl("Y", df01$chr) & df01$sex=="M" & grepl("hap1", df01$sample), ]
df02 <- rbind(df01_main, rbind(df01_x, df01_y))

# df02 <- df01
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.003, alpha=0.8) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly HG38 Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/autism/coverage/coverage_hg38_window_exclude.pdf", g2, width = 15, height = 7, units = "in")
```

```{r}
CURR_GENOME <- "hg38"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/autism/coverage/results_hg38_exclude_big/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    # if(x == "chrX"){
    #   to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
    #     BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]   
    #     BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
    #     BOOL_OL <- BOOL_START & BOOL_END
    #     return(which(BOOL_OL))
    #   }) %>% unlist %>% u
    #   print(to_remove)
    #   temp_chr <- temp_chr[-to_remove, ]
    # }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- df00$sample %>% str_extract("(?<=_)..(?=_)")
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")
df00 <- df00[!df00$role %in% c("fa", "mo"), ]

df00$family <- df00$sample %>% str_extract("^.....(?=_)")
TARGET_FAMILIES <- read.table("C:/docs/autism/coverage/batch_1_families_main.tsv")
df00 <- df00[df00$family %in% TARGET_FAMILIES$V1, ]

write.table(df00, "C:/docs/autism/coverage/hg38_coverage_ggplot_exclude_big.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/autism/coverage/hg38_coverage_main_exclude_big.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


df_sex <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()
df01$sex <- lapply(df01$sample %>% str_remove("_hap."), function(x){
  df_sex$Sex[df_sex$SP_ID==x]
}) %>% unlist

df01_main <- df01[!grepl("X|Y", df01$chr), ]
df01_x <- df01[grepl("X", df01$chr) & !(df01$sex=="M"  & grepl("hap1", df01$sample)) , ]
df01_y <- df01[grepl("Y", df01$chr) & df01$sex=="M" & grepl("hap1", df01$sample), ]
df02 <- rbind(df01_main, rbind(df01_x, df01_y))

# df02 <- df01
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.003, alpha=0.8) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly HG38 Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/autism/coverage/coverage_hg38_window_exclude_big.pdf", g2, width = 15, height = 7, units = "in")
```


```{r}
CURR_GENOME <- "hg38"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/autism/coverage/results_hg38_exclude_small/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    # if(x == "chrX"){
    #   to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
    #     BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]   
    #     BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
    #     BOOL_OL <- BOOL_START & BOOL_END
    #     return(which(BOOL_OL))
    #   }) %>% unlist %>% u
    #   print(to_remove)
    #   temp_chr <- temp_chr[-to_remove, ]
    # }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- df00$sample %>% str_extract("(?<=_)..(?=_)")
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")
df00 <- df00[!df00$role %in% c("fa", "mo"), ]

df00$family <- df00$sample %>% str_extract("^.....(?=_)")
TARGET_FAMILIES <- read.table("C:/docs/autism/coverage/batch_1_families_main.tsv")
df00 <- df00[df00$family %in% TARGET_FAMILIES$V1, ]

write.table(df00, "C:/docs/autism/coverage/hg38_coverage_ggplot_exclude_small.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/autism/coverage/hg38_coverage_main_exclude_small.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


df_sex <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()
df01$sex <- lapply(df01$sample %>% str_remove("_hap."), function(x){
  df_sex$Sex[df_sex$SP_ID==x]
}) %>% unlist

df01_main <- df01[!grepl("X|Y", df01$chr), ]
df01_x <- df01[grepl("X", df01$chr) & !(df01$sex=="M"  & grepl("hap1", df01$sample)) , ]
df01_y <- df01[grepl("Y", df01$chr) & df01$sex=="M" & grepl("hap1", df01$sample), ]
df02 <- rbind(df01_main, rbind(df01_x, df01_y))

# df02 <- df01
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.003, alpha=0.8) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly HG38 Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/autism/coverage/coverage_hg38_window_exclude_small.pdf", g2, width = 15, height = 7, units = "in")
```


```{r}
df_temp <- data.frame(`HYZ-204` = c("n50 (Mbp)", "# contigs"), 
                      hap1 = c("8.24", "4001"), 
                      hap2 = c("7.67", "3867"), 
                      check.names = FALSE)

df_temp <- data.frame(`HYZ-207` = c("n50 (Mbp)", "# contigs"), 
                      hap1 = c("5.98", "4277"), 
                      hap2 = c("6.33", "4177"), 
                      check.names = FALSE)
```

```{r}
df_temp <- data.frame(`HYZ-204` = c("n50 (Mbp)", "# contigs"), 
                      hap1 = c("8.24", "4001"), 
                      hap2 = c("7.67", "3867"), 
                      check.names = FALSE)

df_temp <- data.frame(`HYZ-207` = c("n50 (Mbp)", "# contigs"), 
                      hap1 = c("5.98", "4277"), 
                      hap2 = c("6.33", "4177"), 
                      check.names = FALSE)
```


```{r}
df0 <- fread("C:/docs/autism/qv/temp.txt") %>% as.data.frame()
df1 <- fread("C:/docs/autism/manifest.autism.asm.aln.tsv") %>% as.data.frame()
my_dir <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/Rett-like/Isaac/asm_qc_2/QC_results/merqury/meryl/"
my_res <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/Rett-like/Isaac/asm_qc_2/QC_results/merqury/results/"

my_scripts <- lapply(df0$sample, function(x){
  my_script <- paste0("#!/usr/bin/env bash\n",
         "set -euxo pipefail\n",
         "mkdir -p ", my_res, x, "\n",
         "cd ", my_res, x, "\n",
         "merqury.sh ", my_dir, x, "/", x, "_all.meryl ",
         df1$ASM_QC_HAP1[df1$SAMPLE==x], " ",
         df1$ASM_QC_HAP2[df1$SAMPLE==x], " ", x)
  write.table(my_script, paste0("C:/docs/autism/qv/qv_script_", x, ".sh"), sep="", col.names = FALSE, row.names = FALSE, quote=FALSE)
})
names(my_scripts) <- df0$sample

```



5/19/2025
```{r}
CURR_GENOME <- "t2t"

files <- list.files("C:/docs/autism/coverage/results_2025_05_19/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- df00$sample %>% str_extract("(?<=_)..(?=_)")
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("p", df00$role), "proband", "sibling")
df00 <- df00[!df00$role %in% c("fa", "mo"), ]

df00$family <- df00$sample %>% str_extract("^.....(?=_)")
TARGET_FAMILIES <- read.table("C:/docs/autism/coverage/batch_1_families_main.tsv")
df00 <- df00[df00$family %in% TARGET_FAMILIES$V1, ]

write.table(df00, "C:/docs/autism/coverage/T2T_coverage_ggplot_20250519.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/autism/coverage/T2T_coverage_main_exclude_20250519.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

my_samples <- read.table("C:/docs/autism/coverage/temp.txt")
my_samples$sample <- my_samples$V1 %>% str_remove("hap.")

df01 <- df00[str_remove(df00$sample, "_hap.") %in% my_samples$sample, ]


df_sex <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()
df01$sex <- lapply(df01$sample %>% str_remove("_hap."), function(x){
  df_sex$Sex[df_sex$SP_ID==x]
}) %>% unlist

df01_main <- df01[!grepl("X|Y", df01$chr), ]
df01_x <- df01[grepl("X", df01$chr) & !(df01$sex=="M"  & grepl("hap1", df01$sample)) , ]
df01_y <- df01[grepl("Y", df01$chr) & df01$sex=="M" & grepl("hap1", df01$sample), ]
df02 <- rbind(df01_main, rbind(df01_x, df01_y))

# df02 <- df01
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.003, alpha=0.8) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly T2T Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/autism/coverage/coverage_T2T_window_exclude_20250519.pdf", g2, width = 15, height = 7, units = "in")
```


```{r}
df00 <- fread("C:/docs/autism/sawfish_hapdiff/manifest.autism.aln.tsv") %>% as.data.frame()
df01 <- fread("C:/docs/autism/sawfish_hapdiff/manifest.maskedY.tsv") %>% as.data.frame()
df02 <- fread("C:/docs/autism/sawfish_hapdiff/manifest.maskedYpar.tsv") %>% as.data.frame()
df03 <- fread("C:/docs/autism/sawfish_hapdiff/manifest.T2T-CHM13v2.tsv") %>% as.data.frame()

df01$REF <- "/net/eichler/vol28/eee_shared/assemblies/CHM13/T2T/v2.0/T2T-CHM13v2.masked_Y.fasta"
df02$REF <- "/net/eichler/vol28/eee_shared/assemblies/CHM13/T2T/v2.0/T2T-CHM13v2.masked_Ypars.fasta"
df03$REF <- "/net/eichler/vol28/eee_shared/assemblies/CHM13/T2T/v2.0/T2T-CHM13v2.fasta"

df10 <- rbind(df01, df02, df03)
df00$ALIGNMENT_HG38 <- df00$ALIGNMENT
df00$ALIGNMENT <- NULL
df00$ALIGNMENT_T2T <- lapply(df00$SAMPLE, function(x){
  if(any(x %in% df10$SAMPLE)){
    df10$ALN[df10$SAMPLE == x][1]  
  } else {
    "missing_sequence_files"
  }
}) %>% unlist
df00$REF <- lapply(df00$SAMPLE, function(x){
  if(any(x %in% df10$SAMPLE)){
    df10$REF[df10$SAMPLE == x][1]  
  } else {
    "missing_sequence_files"
  }
}) %>% unlist

df1 <- read.table("C:/docs/autism/coverage/temp.txt") %>% as.data.frame()
df1 <- fread("C:/docs/autism/batch_1_samples.txt") %>% as.data.frame()
batch_1_samples <- df1$V1 %>% str_remove("hap.") %>% u

write.table(df00, "C:/docs/autism/sawfish_hapdiff/manifest.sawfish.hapdiff.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```

```{r}
df_delly <- fread("C:/docs/autism/manifest_2025_05_23/autism_delly.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_delly$path <- df_delly$path %>% str_replace("^.", "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/delly")
df_delly$sample <- df_delly$path %>% bname %>% str_remove(".delly.vcf")
df_delly$reference <- df_delly$path %>% str_extract("T2T-CHM[^/]+")
df_delly <- df_delly[order(df_delly$reference), ]

df_hapdiff <- fread("C:/docs/autism/manifest_2025_05_23/autism_hapdiff.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_hapdiff <- df_hapdiff[grepl("hapdiff_phased.vcf.gz$"  , df_hapdiff$path), , drop = FALSE]
df_hapdiff$sample <- df_hapdiff$path %>% str_remove("/hapdiff_phased.vcf.gz") %>% bname

df_pav <- fread("C:/docs/autism/manifest_2025_05_23/autism_pav.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pav <- df_pav[grepl("shard.*.vcf.gz$", df_pav$path), , drop=FALSE]
df_pav$path <- df_pav$path %>% str_replace("^.", "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pav_new/CHM13v2.0")
df_pav$sample <- df_pav$path %>% bname() %>% str_remove(".vcf.gz") %>% str_remove("pav_")

df_pav_noY <- fread("C:/docs/autism/manifest_2025_05_23/autism_SV_manifest_20250502.tsv") %>% as.data.frame()
df_pav_noY <- df_pav_noY[, c("SAMPLE", "pav_CHM13v2.0_noY")]


df_pbsv_bnd <- fread("C:/docs/autism/manifest_2025_05_23/autism_pbsv.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_bnd <- df_pbsv_bnd[grepl("_bnd.vcf.gz$", df_pbsv_bnd$path), , drop=FALSE]
df_pbsv_bnd$path <- df_pbsv_bnd$path %>% str_replace("^.", "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0")
df_pbsv_bnd$sample <- df_pbsv_bnd$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_bnd.vcf.gz)")
df_pbsv_bnd$reference <- df_pbsv_bnd$path %>% str_extract("(?<=pbsv/CHM13v2.0/)[^/]+(?=/results)")
df_pbsv_bnd <- df_pbsv_bnd[order(df_pbsv_bnd$reference), ]

df_pbsv_dup <- fread("C:/docs/autism/manifest_2025_05_23/autism_pbsv.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_dup <- df_pbsv_dup[grepl("_dup.vcf.gz$", df_pbsv_dup$path), , drop=FALSE]
df_pbsv_dup$path <- df_pbsv_dup$path %>% str_replace("^.", "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0")
df_pbsv_dup$sample <- df_pbsv_dup$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_dup.vcf.gz)")
df_pbsv_dup$reference <- df_pbsv_dup$path %>% str_extract("(?<=pbsv/CHM13v2.0/)[^/]+(?=/results)")
df_pbsv_dup <- df_pbsv_dup[order(df_pbsv_dup$reference), ]

df_pbsv_sv <- fread("C:/docs/autism/manifest_2025_05_23/autism_pbsv.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_sv <- df_pbsv_sv[grepl("_sv.vcf.gz$", df_pbsv_sv$path), , drop=FALSE]
df_pbsv_sv$path <- df_pbsv_sv$path %>% str_replace("^.", "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0")
df_pbsv_sv$sample <- df_pbsv_sv$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_sv.vcf.gz)")
df_pbsv_sv$reference <- df_pbsv_sv$path %>% str_extract("(?<=pbsv/CHM13v2.0/)[^/]+(?=/results)")
df_pbsv_sv <- df_pbsv_sv[order(df_pbsv_sv$reference), ]

df_sawfish <- fread("C:/docs/autism/manifest_2025_05_23/autism_sawfish.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_sawfish <- df_sawfish[grepl("genotyped.sv.vcf.gz$", df_sawfish$path), , drop=FALSE]
df_sawfish$sample <- df_sawfish$path %>% str_extract("(?<=/sawfish/)[^/]+(?=/sawfish_call/)")
df_temp_sawfish <- fread("C:/docs/autism/manifest_2025_05_23/manifest.sawfish.hapdiff.input.tsv") %>% as.data.frame()
df_sawfish$reference <- df_sawfish$sample %>% lapply(function(x){df_temp_sawfish$REF[df_temp_sawfish$SAMPLE==x]}) %>% unlist
df_sawfish <- df_sawfish[order(df_sawfish$reference), ]

df_sniffles <- fread("C:/docs/autism/manifest_2025_05_23/autism_sniffles.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_sniffles <- df_sniffles[grepl(".sniffles.vcf$", df_sniffles$path), , drop=FALSE]
df_sniffles$path <- df_sniffles$path %>% str_replace("^.", "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/sniffles")
df_sniffles$sample <- df_sniffles$path %>% bname %>% str_remove(".sniffles.vcf")
df_sniffles$reference <- df_sniffles$path %>% str_extract("T2T-CHM[^/]+")
df_sniffles <- df_sniffles[order(df_sniffles$reference), ]

df_temp <- fread("C:/docs/autism/manifest_2025_05_23/manifest.autism.aln.tsv") %>% as.data.frame()
df_temp2 <- fread("C:/docs/autism/manifest_2025_05_23/manifest.sawfish.hapdiff.input.tsv") %>% as.data.frame()

df00 <- data.frame(SAMPLE = u(c(df_delly$sample, df_hapdiff$sample, df_pav$sample, df_pbsv_bnd$sample, df_sawfish$sample, df_sniffles$sample, df_pav_noY$SAMPLE, df_temp_sawfish$SAMPLE, df_temp$SAMPLE)) %>% sort)

df00$ASM_QC_HAP1 <-lapply(df00$SAMPLE, function(x){if(x %in% df_temp$SAMPLE){df_temp$ASM_QC_HAP1[df_temp$SAMPLE==x][1]}else{""}}) %>% unlist
df00$ASM_QC_HAP2 <-lapply(df00$SAMPLE, function(x){if(x %in% df_temp$SAMPLE){df_temp$ASM_QC_HAP2[df_temp$SAMPLE==x][1]}else{""}}) %>% unlist
df00$ALN_HG38 <-lapply(df00$SAMPLE, function(x){if(x %in% df_temp2$SAMPLE){df_temp2$ALIGNMENT_HG38[df_temp2$SAMPLE==x][1]}else{""}}) %>% unlist
df00$ALN_T2T <-lapply(df00$SAMPLE, function(x){if(x %in% df_temp2$SAMPLE){df_temp2$ALIGNMENT_T2T[df_temp2$SAMPLE==x][1]}else{""}}) %>% unlist
df00$ALN_T2T_REF <-lapply(df00$SAMPLE, function(x){if(x %in% df_temp2$SAMPLE){df_temp2$REF[df_temp2$SAMPLE==x][1]}else{""}}) %>% unlist

df00$DELLY <- lapply(df00$SAMPLE, function(x){if(x %in% df_delly$sample){tail(df_delly$path[df_delly$sample==x], 1)}else{""}}) %>% unlist
df00$DELLY_REF <- lapply(df00$SAMPLE, function(x){if(x %in% df_delly$sample){tail(df_delly$reference[df_delly$sample==x], 1)}else{""}}) %>% unlist

df00$HAPDIFF <- lapply(df00$SAMPLE, function(x){if(x %in% df_hapdiff$sample){tail(df_hapdiff$path[df_hapdiff$sample==x], 1)}else{""}}) %>% unlist

df00$PAV_T2T <- lapply(df00$SAMPLE, function(x){if(x %in% df_pav$sample){tail(df_pav$path[df_pav$sample==x], 1)}else{""}}) %>% unlist
df00$PAV_T2T_NoY <- lapply(df00$SAMPLE, function(x){if(x %in% df_pav_noY$SAMPLE ){tail(df_pav_noY$pav_CHM13v2.0_noY[df_pav_noY$SAMPLE ==x], 1)}else{""}}) %>% unlist

df00$PBSV_BND_T2T <- lapply(df00$SAMPLE, function(x){if(x %in% df_pbsv_bnd$sample){tail(df_pbsv_bnd$path[df_pbsv_bnd$sample==x], 1)}else{""}}) %>% unlist
df00$PBSV_BND_T2T_REF <- lapply(df00$SAMPLE, function(x){if(x %in% df_pbsv_bnd$sample){tail(df_pbsv_bnd$reference[df_pbsv_bnd$sample==x], 1)}else{""}}) %>% unlist

df00$PBSV_DUP_T2T <- lapply(df00$SAMPLE, function(x){if(x %in% df_pbsv_dup$sample){tail(df_pbsv_dup$path[df_pbsv_dup$sample==x], 1)}else{""}}) %>% unlist
df00$PBSV_DUP_T2T_REF <- lapply(df00$SAMPLE, function(x){if(x %in% df_pbsv_dup$sample){tail(df_pbsv_dup$reference[df_pbsv_dup$sample==x], 1)}else{""}}) %>% unlist

df00$PBSV_SV_T2T <- lapply(df00$SAMPLE, function(x){if(x %in% df_pbsv_sv$sample){tail(df_pbsv_sv$path[df_pbsv_sv$sample==x], 1)}else{""}}) %>% unlist
df00$PBSV_SV_T2T_REF <- lapply(df00$SAMPLE, function(x){if(x %in% df_pbsv_sv$sample){tail(df_pbsv_sv$reference[df_pbsv_sv$sample==x], 1)}else{""}}) %>% unlist

df00$SAWFISH <- lapply(df00$SAMPLE, function(x){if(x %in% df_sawfish$sample){tail(df_sawfish$path[df_sawfish$sample==x], 1)}else{""}}) %>% unlist
df00$SAWFISH_REF <- lapply(df00$SAMPLE, function(x){if(x %in% df_sawfish$sample){tail(df_sawfish$reference[df_sawfish$sample==x], 1)}else{""}}) %>% unlist

df00$SNIFFLES <- lapply(df00$SAMPLE, function(x){if(x %in% df_sniffles$sample){tail(df_sniffles$path[df_sniffles$sample==x], 1)}else{""}}) %>% unlist
df00$SNIFFLES_REF <- lapply(df00$SAMPLE, function(x){if(x %in% df_sniffles$sample){tail(df_sniffles$reference[df_sniffles$sample==x], 1)}else{""}}) %>% unlist

# write.table(df00, "C:/docs/autism/manifest_2025_05_23/manifest.autism.t2t.2025_05_27.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```


```{r}
df01 <- fread("C:/docs/autism/manifest_2025_05_23/manifest.autism.t2t.2025_05_30.tsv") %>% as.data.frame()
df01$DELLY[df01$SAMPLE == "11611_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/delly/T2T-CHM13v2/results/11611_fa/11611_fa.delly.vcf"
df01$DELLY[df01$SAMPLE == "13561_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/delly/T2T-CHM13v2/results/13561_fa/13561_fa.delly.vcf"
df01$DELLY_REF[df01$SAMPLE == "11611_fa"] <- "T2T-CHM13v2"
df01$DELLY_REF[df01$SAMPLE == "13561_fa"] <- "T2T-CHM13v2"
write.table(df01, "C:/docs/autism/manifest_2025_05_23/manifest.autism.t2t.2025_06_03.tsv", col.names = TRUE, row.names = FALSE, quote=FALSE, sep = "\t")
```


```{r}
df01 <- fread("C:/docs/autism/manifest_2025_05_23/manifest.autism.t2t.2025_06_03.tsv") %>% as.data.frame()
df01$PBSV_BND_T2T[df01$SAMPLE=="11611_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0/T2T-CHM13v2/results/11611_fa/pbsv_11611_fa_bnd.vcf.gz"
df01$PBSV_DUP_T2T[df01$SAMPLE=="11611_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0/T2T-CHM13v2/results/11611_fa/pbsv_11611_fa_dup.vcf.gz"
df01$PBSV_SV_T2T[df01$SAMPLE=="11611_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0/T2T-CHM13v2/results/11611_fa/pbsv_11611_fa_sv.vcf.gz"
df01$PBSV_BND_T2T_REF[df01$SAMPLE=="11611_fa"] <- "T2T-CHM13v2"
df01$PBSV_DUP_T2T_REF[df01$SAMPLE=="11611_fa"] <- "T2T-CHM13v2"
df01$PBSV_SV_T2T_REF[df01$SAMPLE=="11611_fa"] <- "T2T-CHM13v2"

df01$PBSV_BND_T2T[df01$SAMPLE=="13561_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0/T2T-CHM13v2/results/13561_fa/pbsv_13561_fa_bnd.vcf.gz"
df01$PBSV_DUP_T2T[df01$SAMPLE=="13561_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0/T2T-CHM13v2/results/13561_fa/pbsv_13561_fa_dup.vcf.gz"
df01$PBSV_SV_T2T[df01$SAMPLE=="13561_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pbsv/CHM13v2.0/T2T-CHM13v2/results/13561_fa/pbsv_13561_fa_sv.vcf.gz"
df01$PBSV_BND_T2T_REF[df01$SAMPLE=="13561_fa"] <- "T2T-CHM13v2"
df01$PBSV_DUP_T2T_REF[df01$SAMPLE=="13561_fa"] <- "T2T-CHM13v2"
df01$PBSV_SV_T2T_REF[df01$SAMPLE=="13561_fa"] <- "T2T-CHM13v2"

df01$SNIFFLES[df01$SAMPLE=="11611_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/sniffles/T2T-CHM13v2/results/11611_fa/sniffles/2.3.2/11611_fa.sniffles.vcf.gz"
df01$SNIFFLES[df01$SAMPLE=="13561_fa"] <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/sniffles/T2T-CHM13v2/results/13561_fa/sniffles/2.3.2/13561_fa.sniffles.vcf.gz"
df01$SNIFFLES_REF[df01$SAMPLE=="11611_fa"] <- "T2T-CHM13v2"
df01$SNIFFLES_REF[df01$SAMPLE=="13561_fa"] <- "T2T-CHM13v2"

write.table(df01, "C:/docs/autism/manifest_2025_05_23/manifest.autism.t2t.2025_06_09.tsv", col.names = TRUE, row.names = FALSE, quote=FALSE, sep = "\t")
```

```{r}
df1 <- fread("C:/Users/iwong1/Downloads/manifest.autism.asm.aln (3).tsv") %>% as.data.frame()
df2 <- fread("C:/Users/iwong1/Downloads/autism_samples_main.tsv.txt", header = FALSE) %>% as.data.frame()
labs <- fread("C:/docs/autism/coverage/sex_labels.txt") %>% as.data.frame()

df00 <- data.frame(SAMPLE=df1$SAMPLE[df1$SAMPLE %in% df2$V1], 
                   INPUT = df1$ALIGNMENT[df1$SAMPLE %in% df2$V1], 
                   BAI = paste0(df1$ALIGNMENT[df1$SAMPLE %in% df2$V1], ".bai"))
df01 <- df00[df00$SAMPLE %in% labs$SP_ID[labs$Sex=="F"], ]
# write.table(df00, "C:/docs/autism/manifest_2025_07_02/delly_hg38.manifest.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
# write.table(df01, "C:/docs/autism/manifest_2025_07_02/delly_hg38_noy.manifest.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```

```{r}
df00 <- df1[df1$SAMPLE %in% df2$V1, ]
df00$LABEL <- df00$SAMPLE %>% lapply(function(x){labs$Sex[labs$SP_ID==x]}) %>% unlist
```


```{r}
df_delly_hg38 <- fread("C:/docs/autism/manifest_2025_07_02/delly_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_delly_hg38$sample <- df_delly_hg38$path %>% bname %>% str_remove(".delly.vcf")
df_delly_hg38noy <- fread("C:/docs/autism/manifest_2025_07_02/delly_hg38noy.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_delly_hg38noy$sample <- df_delly_hg38noy$path %>% bname %>% str_remove(".delly.vcf")

df_hapdiff_hg38 <- fread("C:/docs/autism/manifest_2025_07_02/hapdiff_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_hapdiff_hg38 <- df_hapdiff_hg38[grepl("hapdiff_phased.vcf.gz$"  , df_hapdiff_hg38$path), , drop = FALSE]
df_hapdiff_hg38$sample <- df_hapdiff_hg38$path %>% str_remove("/hapdiff_phased.vcf.gz") %>% bname

df_hapdiff_hg38noy <- fread("C:/docs/autism/manifest_2025_07_02/hapdiff_hg38noy.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_hapdiff_hg38noy <- df_hapdiff_hg38noy[grepl("hapdiff_phased.vcf.gz$"  , df_hapdiff_hg38noy$path), , drop = FALSE]
df_hapdiff_hg38noy$sample <- df_hapdiff_hg38noy$path %>% str_remove("/hapdiff_phased.vcf.gz") %>% bname

df_pbsv_bnd_hg38 <- fread("C:/docs/autism/manifest_2025_07_02/pbsv_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_bnd_hg38 <- df_pbsv_bnd_hg38[grepl("_bnd.vcf.gz$", df_pbsv_bnd_hg38$path), , drop=FALSE]
df_pbsv_bnd_hg38$sample <- df_pbsv_bnd_hg38$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_bnd.vcf.gz)")

df_pbsv_dup_hg38 <- fread("C:/docs/autism/manifest_2025_07_02/pbsv_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_dup_hg38 <- df_pbsv_dup_hg38[grepl("_dup.vcf.gz$", df_pbsv_dup_hg38$path), , drop=FALSE]
df_pbsv_dup_hg38$sample <- df_pbsv_dup_hg38$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_dup.vcf.gz)")

df_pbsv_sv_hg38 <- fread("C:/docs/autism/manifest_2025_07_02/pbsv_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_sv_hg38 <- df_pbsv_sv_hg38[grepl("_sv.vcf.gz$", df_pbsv_sv_hg38$path), , drop=FALSE]
df_pbsv_sv_hg38$sample <- df_pbsv_sv_hg38$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_sv.vcf.gz)")

df_pbsv_bnd_hg38noy <- fread("C:/docs/autism/manifest_2025_07_02/pbsv_hg38noy.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_bnd_hg38noy <- df_pbsv_bnd_hg38noy[grepl("_bnd.vcf.gz$", df_pbsv_bnd_hg38noy$path), , drop=FALSE]
df_pbsv_bnd_hg38noy$sample <- df_pbsv_bnd_hg38noy$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_bnd.vcf.gz)")

df_pbsv_dup_hg38noy <- fread("C:/docs/autism/manifest_2025_07_02/pbsv_hg38noy.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_dup_hg38noy <- df_pbsv_dup_hg38noy[grepl("_dup.vcf.gz$", df_pbsv_dup_hg38noy$path), , drop=FALSE]
df_pbsv_dup_hg38noy$sample <- df_pbsv_dup_hg38noy$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_dup.vcf.gz)")

df_pbsv_sv_hg38noy <- fread("C:/docs/autism/manifest_2025_07_02/pbsv_hg38noy.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pbsv_sv_hg38noy <- df_pbsv_sv_hg38noy[grepl("_sv.vcf.gz$", df_pbsv_sv_hg38noy$path), , drop=FALSE]
df_pbsv_sv_hg38noy$sample <- df_pbsv_sv_hg38noy$path %>% bname %>% str_extract("(?<=pbsv_).*(?=_sv.vcf.gz)")

df_sawfish_hg38 <- fread("C:/docs/autism/manifest_2025_07_02/sawfish_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_sawfish_hg38 <- df_sawfish_hg38[grepl("genotyped.sv.vcf.gz$", df_sawfish_hg38$path), , drop=FALSE]
df_sawfish_hg38$sample <- df_sawfish_hg38$path %>% str_extract("(?<=/sawfish/)[^/]+(?=/sawfish_call/)")

df_sawfish_hg38noy <- fread("C:/docs/autism/manifest_2025_07_02/sawfish_hg38noy.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_sawfish_hg38noy <- df_sawfish_hg38noy[grepl("genotyped.sv.vcf.gz$", df_sawfish_hg38noy$path), , drop=FALSE]
df_sawfish_hg38noy$sample <- df_sawfish_hg38noy$path %>% str_extract("(?<=/sawfish/)[^/]+(?=/sawfish_call/)")

df_sniffles_hg38 <- fread("C:/docs/autism/manifest_2025_07_02/sniffles_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_sniffles_hg38 <- df_sniffles_hg38[grepl(".sniffles.vcf", df_sniffles_hg38$path), , drop=FALSE]
df_sniffles_hg38$sample <- df_sniffles_hg38$path %>% bname %>% str_remove(".sniffles.vcf.gz")

df_sniffles_hg38noy <- fread("C:/docs/autism/manifest_2025_07_02/sniffles_hg38noy.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_sniffles_hg38noy <- df_sniffles_hg38noy[grepl(".sniffles.vcf", df_sniffles_hg38noy$path), , drop=FALSE]
df_sniffles_hg38noy$sample <- df_sniffles_hg38noy$path %>% bname %>% str_remove(".sniffles.vcf.gz")

df_pav <- fread("C:/docs/autism/manifest_2025_07_02/pav_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_pav <- df_pav[grepl(".vcf.gz", df_pav$path), , drop=FALSE]
df_pav$sample <- df_pav$path %>% bname %>% str_remove(".vcf.gz") %>% str_remove("^pav_")


df_cutesv_hg38 <- fread("C:/docs/autism/manifest_2025_07_02/cutesv_hg38.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_cutesv_hg38 <- df_cutesv_hg38[grepl(".cutesv.vcf.gz$"  , df_cutesv_hg38$path), , drop = FALSE]
df_cutesv_hg38$sample <- df_cutesv_hg38$path %>% str_remove(".cutesv.vcf.gz") %>% bname

df_cutesv_hg38noy <- fread("C:/docs/autism/manifest_2025_07_02/cutesv_hg38noy.txt", header=FALSE, col.names = "path") %>% as.data.frame()
df_cutesv_hg38noy <- df_cutesv_hg38noy[grepl(".cutesv.vcf.gz$"  , df_cutesv_hg38noy$path), , drop = FALSE]
df_cutesv_hg38noy$sample <- df_cutesv_hg38noy$path %>% str_remove(".cutesv.vcf.gz") %>% bname
 

df00$DELLY <- lapply(df00$SAMPLE, function(x){
  if(df00$LABEL[df00$SAMPLE==x] == "F"){
    if(x %in% df_delly_hg38noy$sample){tail(df_delly_hg38noy$path[df_delly_hg38noy$sample==x], 1)}else{""}
  } else {
    if(x %in% df_delly_hg38$sample)   {tail(df_delly_hg38$path[df_delly_hg38$sample==x]      , 1)}else{""}
  }
}) %>% unlist

df00$HAPDIFF <- lapply(df00$SAMPLE, function(x){
  if(df00$LABEL[df00$SAMPLE==x] == "F"){
    if(x %in% df_hapdiff_hg38noy$sample){tail(df_hapdiff_hg38noy$path[df_hapdiff_hg38noy$sample==x], 1)}else{""}
  } else {
    if(x %in% df_hapdiff_hg38$sample){tail(df_hapdiff_hg38$path[df_hapdiff_hg38$sample==x],          1)}else{""}
  }
}) %>% unlist

df00$PBSV_BND <- lapply(df00$SAMPLE, function(x){
  if(df00$LABEL[df00$SAMPLE==x] == "F"){
    if(x %in% df_pbsv_bnd_hg38noy$sample){tail(df_pbsv_bnd_hg38noy$path[df_pbsv_bnd_hg38noy$sample==x], 1)}else{""}
  } else {
    if(x %in% df_pbsv_bnd_hg38$sample)   {tail(df_pbsv_bnd_hg38$path[df_pbsv_bnd_hg38$sample==x]      , 1)}else{""}
  }
}) %>% unlist

df00$PBSV_DUP <- lapply(df00$SAMPLE, function(x){
  if(df00$LABEL[df00$SAMPLE==x] == "F"){
    if(x %in% df_pbsv_dup_hg38noy$sample){tail(df_pbsv_dup_hg38noy$path[df_pbsv_dup_hg38noy$sample==x], 1)}else{""}
  } else {
    if(x %in% df_pbsv_dup_hg38$sample)   {tail(df_pbsv_dup_hg38$path[df_pbsv_dup_hg38$sample==x]      , 1)}else{""}
  }
}) %>% unlist

df00$PBSV_SV <- lapply(df00$SAMPLE, function(x){
  if(df00$LABEL[df00$SAMPLE==x] == "F"){
    if(x %in% df_pbsv_sv_hg38noy$sample){tail(df_pbsv_sv_hg38noy$path[df_pbsv_sv_hg38noy$sample==x], 1)}else{""}
  } else {
    if(x %in% df_pbsv_sv_hg38$sample)   {tail(df_pbsv_sv_hg38$path[df_pbsv_sv_hg38$sample==x]      , 1)}else{""}
  }
}) %>% unlist

df00$SAWFISH <- lapply(df00$SAMPLE, function(x){
  if(df00$LABEL[df00$SAMPLE==x] == "F"){
    if(x %in% df_sawfish_hg38noy$sample){tail(df_sawfish_hg38noy$path[df_sawfish_hg38noy$sample==x], 1)}else{""}
  } else {
    if(x %in% df_sawfish_hg38$sample)   {tail(df_sawfish_hg38$path[df_sawfish_hg38$sample==x]      , 1)}else{""}
  }
}) %>% unlist

df00$SNIFFLES <- lapply(df00$SAMPLE, function(x){
  if(df00$LABEL[df00$SAMPLE==x] == "F"){
    if(x %in% df_sniffles_hg38noy$sample){tail(df_sniffles_hg38noy$path[df_sniffles_hg38noy$sample==x], 1)}else{""}
  } else {
    if(x %in% df_sniffles_hg38$sample)   {tail(df_sniffles_hg38$path[df_sniffles_hg38$sample==x]      , 1)}else{""}
  }
}) %>% unlist

df00$PAV <- lapply(df00$SAMPLE, function(x){
  if(x %in% df_pav$sample){tail(df_pav$path[df_pav$sample==x], 1)}else{""}
}) %>% unlist

df00$CUTESV <- lapply(df00$SAMPLE, function(x){
  if(df00$LABEL[df00$SAMPLE==x] == "F"){
    if(x %in% df_cutesv_hg38noy$sample){tail(df_cutesv_hg38noy$path[df_cutesv_hg38noy$sample==x], 1)}else{""}
  } else {
    if(x %in% df_cutesv_hg38$sample)   {tail(df_cutesv_hg38$path[df_cutesv_hg38$sample==x]      , 1)}else{""}
  }
}) %>% unlist

write.table(df00, "C:/docs/autism/manifest_2025_07_02/manifest.autism.hg38.2025_07_21.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```


```{r}
df1 <- fread("C:/docs/autism/manifest_2025_07_02/DatasetS1.tsv") %>% as.data.frame()
df2 <- fread("C:/docs/autism/manifest_2025_07_02/callable.txt2", header=FALSE, col.names="PATH") %>% as.data.frame()
df2$SAMPLE <- df2$PATH %>% str_extract("(?<=results/).*(?=/callable/)")
df2$HAP1 <- grepl("callable_regions_h1_500.bed.gz", df2$PATH)
df2$HAP2 <- grepl("callable_regions_h2_500.bed.gz", df2$PATH)

df10 <- df1[, c(4,5,6,8)]
df10$callable_h1 <- seq_down(df10) %>% lapply(function(x){
  rval <- ""
  if(df10$Sample[x] %in% df2$SAMPLE){
    rval <- df2$PATH[df2$HAP1 & (df2$SAMPLE==df10$Sample[x])]
  } else if(df10$cohort_ID[x] %in% df2$SAMPLE) {
    rval <- df2$PATH[df2$HAP1 & (df2$SAMPLE==df10$cohort_ID[x])]
  } else {
    print(df10$Sample[x])
    return("NA")
  }
  if(length(rval!=1)){
    rval <- rval[which(max(nchar(rval)) == nchar(rval))]
  }
  return(rval)
}) %>% unlist

df10$callable_h2 <- seq_down(df10) %>% lapply(function(x){
  rval <- ""
  if(df10$Sample[x] %in% df2$SAMPLE){
    rval <- df2$PATH[df2$HAP2 & (df2$SAMPLE==df10$Sample[x])]
  } else if(df10$cohort_ID[x] %in% df2$SAMPLE) {
    rval <- df2$PATH[df2$HAP2 & (df2$SAMPLE==df10$cohort_ID[x])]
  } else {
    print(df10$Sample[x])
    return("NA")
  }
  if(length(rval!=1)){
    rval <- rval[which(max(nchar(rval)) == nchar(rval))]
  }
  return(rval)
}) %>% unlist

write.table(df10, "C:/docs/autism/manifest_2025_07_02/pav_hg38_callable.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

# temp1 <- df10$callable_h1 %>% lapply(length) %>% unlist
# df10$callable_h1[temp1!=1]

```


```{r}
df0 <- fread("C:/docs/autism/manifest_2025_10_06/bams.txt") %>% as.data.frame()
df0$sample <- df0$path %>% bname %>% str_remove(".all.sorted.bam")
df0$ref <- lapply(df0$path, function(x){
  unlist(str_split(x, "/"))[10]
}) %>% unlist
migrate <- function(source, sample, ref, curr, max){
  dest_bam <- paste0("/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/", sample, "/alignments/PacBio_HiFi/", ref, "/alignment.bam")
  dest_bai <- paste0("/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/", sample, "/alignments/PacBio_HiFi/", ref, "/alignment.bam.bai")
  script_000 <- paste0("mkdir -p ", "/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/", sample, "/alignments/PacBio_HiFi/", ref)
  script_1 <- paste0("mv ", source, " ", dest_bam)
  script_2 <- paste0("mv ", source, ".bai ", dest_bai)
  script_3 <- paste0("ln -s ", dest_bam, " ", source)
  script_4 <- paste0("ln -s ", dest_bai, " ", source, ".bai")
  script_0 <- paste0("echo ", curr, "/", max, ": ", source)
  script_00 <- "date"
  c(script_00, script_0, script_000, script_1, script_2, script_3, script_4, "\n")
}
bign <- nrow(df0)
temp2 <- lapply(seq_down(df0), function(x){
  migrate(df0$path[x], df0$sample[x], df0$ref[x], x, bign)
}) %>% unlist

temp2 <- c("#!/bin/bash", "set -euo pipefail", temp2)
write.table(temp2, "C:/docs/autism/manifest_2025_10_06/migration.sh", col.names = FALSE, row.names = FALSE, quote=FALSE)

temp3 <- temp2[grepl("mv.*long_read_archive.*bam$", temp2)] %>% str_extract("/net/eichler/vol28/projects/long_read_archive/nobackups/clinical/.*bam$")
df1 <- data.frame(sample = temp3 %>% lapply(function(x){unlist(str_split(x, "/"))[9]}) %>% unlist %>% u )
MY_REFS <- df0$ref %>% u
for(CURR_REF in MY_REFS){
  new_col <- lapply(df1$sample, function(curr_sample){
    temp3[grepl(paste0(curr_sample, ".*/", CURR_REF, "/"), temp3)]
  })  %>% unlist
  df1 <- cbind(df1, new_col)
  colnames(df1)[ncol(df1)] <- CURR_REF
}
write.table(df1, "C:/docs/autism/manifest_2025_10_06/bams.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```

