---
title: "autism"
output: html_document
date: "2024-04-12"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
df <- fread("C:/home/autism/r_edits/temp.txt") %>% as.data.frame()

# for each row in the FORMAT column, get a list of which index corresponds to each field
my_FORMAT <- df$FORMAT %>% str_split(":")
index_GT <- lapply(my_FORMAT, function(x){which(x=="GT")}) %>% unlist
index_AD <- lapply(my_FORMAT, function(x){which(x=="AD")}) %>% unlist
index_DP <- lapply(my_FORMAT, function(x){which(x=="DP")}) %>% unlist


# get a vector of indexes for columns corresponding to fa, mo, p1, p2, etc
which_columns_to_split <- which(colnames(df) %in% c("FA", "MO") | grepl("\\.p[0-9]$", colnames(df)))
family_columns_to_cbind <- lapply(which_columns_to_split, function(COL_INDEX){
  split_member <- df[, COL_INDEX] %>% str_split(":")
  # identify which columns have more FORMAT fields than the FA column has fields, we won't be able to get good AD or DP infor for these
  bool_useable <- (my_FORMAT %>% lapply(length) %>% unlist)==(split_member %>% lapply(length) %>% unlist)
  columns_member <- lapply(seq_down(df), function(x){
    if(bool_useable[x]==FALSE){
      if(split_member[[x]][1] %>% grepl("\\/|\\|")){ # check to see if the first field contains GT information
        return(c(split_member[[x]][1], 0,0))
      } else {
        return(c("./.", 0, 0))
      }
    }
    
    # return the corresponding row from current family member, and the corresponding column for each corresponding field
    return(c(split_member[[x]][index_GT[x]],
             split_member[[x]][index_AD[x]],
             split_member[[x]][index_DP[x]]))
  }) %>% do.call(rbind, .) %>% as.data.frame()
  colnames(columns_member) <- paste0(colnames(df)[COL_INDEX], "_", c("GT", "AD", "DP"))
  return(columns_member)
}) %>% do.call(cbind, .) %>% as.data.frame()



```

# asm qc manifest
```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats")
# df0 <- fread("C:/home/autism/asm_qc/manifest_job_done1.tab") %>% as.data.frame()
# df0[is.na(df0)] <- "NA"

df0 <- read.table("C:/docs/autism/finished_20240617")
colnames(df0) <- "sample"


bool_child <- grepl("_(p.|s.)$", df0$sample)
df1 <- data.frame(SAMPLE = df0$sample, 
                  H1 = paste0(
                    "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/asm_iw/", 
                    df0$sample, "/", df0$sample, 
                    "/assemblies/hifiasm/",
                    ifelse(bool_child, "trio/", ""),
                    "0.16.1/",
                    df0$sample,
                    ".hifiasm.",
                    ifelse(bool_child, "dip", "bp"),
                    ".hap1.p_ctg.gfa.fasta"),
                  H2 = paste0(
                    "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/asm_iw/", 
                    df0$sample, "/", df0$sample, 
                    "/assemblies/hifiasm/",
                    ifelse(bool_child, "trio/", ""),
                    "0.16.1/",
                    df0$sample, 
                    ".hifiasm.",
                    ifelse(bool_child, "dip", "bp"),
                    ".hap2.p_ctg.gfa.fasta"),
                  ILLUMINA = df0$sample,
                  FOFN = paste0(
                    "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/fofn/Illumina/",
                    df0$sample, 
                    "-Illumina.fastq.fofn"),
                  TRIO = lapply(df0$sample, function(x){
                    if(df00$Family_Type[grep(x, df00$SP_ID)] == "Quad, Simplex"){
                      return("NO")
                    } else {
                      return("YES")
                    }}) %>% unlist,
                  MO_ID = ifelse(bool_child, 
                                  df0$sample %>% str_replace("_(p.|s.)$", "_mo"),
                                  "NA"),
                  FA_ID = ifelse(bool_child, 
                                  df0$sample %>% str_replace("_(p.|s.)$", "_fa"),
                                  "NA")
                  
                  )
write.table(df1, "C:/home/autism/asm_qc/manifest.asmqc.20240613.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```


```{r}
df2 <- read.table("C:/docs/autism/manifest.rerun.new", header=TRUE)
df3 <- df1[!df1$SAMPLE %in% df2$SAMPLE, ]
write.table(df3, "C:/home/autism/asm_qc/manifest.asmqc.20240617.asmqc.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```



```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=448025372", sheet = "FOFN_summary") %>% as.data.frame()
```
# update tracking sheet
```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet="Individual_Stats") %>% as.data.frame()

man_aqc <- fread("C:/docs/autism/asm_qc/manifest.tab.2024-04-22.finished") %>% as.data.frame()
df00$`asm_QC -- Isaac`[df00$SP_ID %in% man_aqc$SAMPLE] <- "Done"

man_pbsv <- fread("C:/docs/autism/manifest.tab.pbsv.2024-04-23.finished") %>% as.data.frame()
df00$`PBSV -- Isaac`[df00$SP_ID %in% man_pbsv$SAMPLE] <- "Done"

man_aln <- fread("C:/docs/autism/aln.samples") %>% as.data.frame()
df00$`aln -- Isaac`[df00$SP_ID %in% man_aln$sample] <- "Done"
df00$`aln_QC -- Isaac (VBI)`[df00$SP_ID %in% man_aln$sample] <- "Done"

write_sheet(df00, "https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", "temp_isaac")
```

```{r}
df1 <- fread("C:/docs/autism/asm_qc/merq.stats.qv.tsv") %>% as.data.frame()
df1 <- df1[!grepl("Both", df1$V1), ]
sample_ids <- df1$V1 %>% str_remove("_hap(1|2)") %>% u
df10 <- lapply(sample_ids, function(x){
  temp_rows <- df1[grepl(x, df1$V1), ]
  data.frame(sample = x, 
             QV_Merqury_hap1 = ifelse(any(grepl("hap1", temp_rows[,1])), temp_rows[grepl("hap1", temp_rows[,1]), 4] %>% round, "NA"), 
             QV_Merqury_hap2 = ifelse(any(grepl("hap2", temp_rows[,1])), temp_rows[grepl("hap2", temp_rows[,1]), 4] %>% round, "NA") 
             ) %>% return()
}) %>% do.call(rbind, . )
rm(df1)

df2 <- fread("C:/docs/autism/asm_qc/merq.stats.completeness.tsv") %>% as.data.frame()
df2 <- df2[!grepl("both", df2$V1), ]
sample_ids <- df2$V1 %>% str_remove("_hap(1|2)") %>% u
df20 <- lapply(sample_ids, function(x){
  temp_rows <- df2[grepl(x, df2$V1), ]
  data.frame(sample = x, 
             Completeness_hap1=ifelse(any(grepl("hap1", temp_rows[,1])), temp_rows[grepl("hap1", temp_rows[,1]), 5] %>% round, "NA"), 
             Completeness_hap2=ifelse(any(grepl("hap2", temp_rows[,1])), temp_rows[grepl("hap2", temp_rows[,1]), 5] %>% round, "NA") 
             ) %>% return()
}) %>% do.call(rbind, . )
rm(df2)

df3 <- fread("C:/docs/autism/asm_qc/contig_stats.tsv") %>% as.data.frame()
sample_ids <- df3$source %>% str_remove("_hap..n50.stats") %>% u
df30 <- lapply(sample_ids, function(x){
  temp_rows <- df3[grep(x, df3$source), ]
  data.frame(sample = x,
             Contig_N50_Mbp_hap1=
               ifelse(any(grepl("hap1", temp_rows[,1])), 
                      temp_rows[grepl("hap1", temp_rows[,1]) & grepl("N50 ", temp_rows[,2]), 2] %>% str_remove("N50 \\(Mbp\\):") %>% str_trim(), 
                      "NA"),
             Contig_N50_Mbp_hap2=
               ifelse(any(grepl("hap2", temp_rows[,1])), 
                      temp_rows[grepl("hap2", temp_rows[,1]) & grepl("N50 ", temp_rows[,2]), 2] %>% str_remove("N50 \\(Mbp\\):") %>% str_trim(), 
                      "NA")
             )
}) %>% do.call(rbind, .)
rm(df3)

df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet="Individual_Stats") %>% as.data.frame()


df123 <- merge(df10, df20, all = TRUE) %>% merge(df30, all=TRUE)
df123[is.na(df123)] <- "NA"

df01 <- merge(df00, df123, by.x = "SP_ID", by.y="sample", all=TRUE)
df01 <- df01[match(df00$SP_ID, df01$SP_ID), ]
write_sheet(df01, "https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", "temp_isaac")



bool_merq_resubmit <- df123$QV_Merqury_hap1=="NA" | df123$QV_Merqury_hap2=="NA"
bool_contig_resubmit <- df123$Contig_N50_Mbp_hap1 == "NA" | df123$Contig_N50_Mbp_hap2 == "NA"

```



```{r}
BOOLS <- c(TRUE, FALSE)
temp_n <- 15
temp_df <- data.frame(temp1 = sample(BOOLS, temp_n, replace = TRUE),
                      temp2 = sample(BOOLS, temp_n, replace = TRUE),
                      temp3 = sample(BOOLS, temp_n, replace = TRUE),
                      temp4 = sample(BOOLS, temp_n, replace = TRUE))
temp_df$ALL <- lapply(seq_down(temp_df), function(x){
  all(temp_df[x, ])
}) %>% unlist
```



```{r}
df0 <- read.table("C:/docs/autism/manifest.tab", sep="\t", header = TRUE)

df1 <- data.frame(SAMPLE = df0$SAMPLE, 
                  ROLE = ifelse(
                    df0$SAMPLE %>% grepl("_fa$", .), "father", ifelse(
                    df0$SAMPLE %>% grepl("_mo$", .), "mother", ifelse(
                    df0$SAMPLE %>% grepl("_p.$", .), "proband", "sibling"))),
                  FAMILY = df0$SAMPLE %>% gsub("_..$", "", .),
                  HAP1 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/saffire/results/",
                                df0$SAMPLE, "_hap1/alignments/", df0$SAMPLE, "_hap1.minimap2.paf"),
                  HAP2 = paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/saffire/results/",
                                df0$SAMPLE, "_hap2/alignments/", df0$SAMPLE, "_hap2.minimap2.paf"))
write.table(df1, "C:/docs/autism/manifest.concatchr.tab", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```






```{r}
df0 <- fread("C:/docs/autism/fofn/sample_list_fofn_to_Isaac.txt") %>% as.data.frame()

my_samples <- u(df0$SAMPLE)

lapply(my_samples, function(x){
  
  write.table(df0$FASTQ[df0$SAMPLE == x], paste0("C:/docs/autism/fofn/", x, ".fofn"), sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE)
  
  
})


```



```{r}
df0 <- fread("C:/docs/autism/ncdu_autism_sorted.tsv") %>% as.data.frame()
df0$notes <- NULL

file_types <- c(".gfa", ".bin", ".bed", ".yak")
my_regex <- paste0("assemblies/hifiasm/0.16.1/.*","(", paste0(file_types, collapse = "|"), ")$"  )
my_regex2 <- paste0("/asm_ys/.*_(s|p)1/.*.hifiasm.bp.hap")

df1 <- df0[grepl("^/net/eichler/vol28/projects/autism_genome_assembly/nobackups/asm_ys", df0$file), ]
df11 <- df1[grepl(my_regex, df1$file) | grepl(my_regex2, df1$file), ]
write.table(df11$file, "C:/docs/autism/to_delete.txt", sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r}
df1 <- readLines("C:/docs/autism/asm_files.txt") %>% as.data.frame()

file_types <- c(".gfa", ".bin", ".bed", ".yak")
my_regex <- paste0("assemblies/hifiasm/0.16.1/.*","(", paste0(file_types, collapse = "|"), ")$"  )
my_regex2 <- paste0("/asm_ys/.*_(s|p)1/.*.hifiasm.bp.hap")

df11 <- df1[grepl(my_regex, df1$.) | grepl(my_regex2, df1$.), ]
write.table(df11, "C:/docs/autism/to_delete.txt", sep="\t", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r}
formals(write.table)$row.names <- FALSE
formals(write.table)$col.names <- TRUE
formals(write.table)$sep <- "\t"
formals(write.table)$quote <- FALSE

write.table(df12, "C:/Users/iwong1/Downloads/temp.tsv")
```


```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats")
df01 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "VBI_stats")

table(df00$SP_ID %in% df01$sample)
df00$SP_ID[!df00$SP_ID %in% df01$sample]


```
  "14444_fa" : ys
  "11569_mo" : na
  "13799_fa" : ys
  "12608_mo" : na


```{bash}
cat batch_2_samples.txt  | xargs -I {} sh -c "mkdir {}; cd {}; ln -s /net/eichler/vol26/7200/software/pipelines/hifiasm-smk/hifi/runlocal; cp ../hifiasm_trio.yaml a; cd .."
```


```{r}
input_file <- "C:/Users/iwong1/Downloads/batch_2_autism.txt"
df0 <- fread(input_file, header = FALSE) %>% as.data.frame()
fofn_dir <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/0_hifi_fofn/"
out_dir <- "/net/eichler/vol28/projects/autism_genome_assembly/nobackups/asm_iw/"
temp1 <- lapply(df0$V1, function(x){
  fam_id <- x %>% str_remove("_..$")
  temp_df <- data.frame(
    sample = x,
    hifi_fofn = paste0(fofn_dir, x, ".fofn"),
    maternal_illumina = ifelse(grepl("(_p|_s)", x), paste0(fofn_dir, fam_id, "_mo.fofn"), "NA"),
    paternal_illumina = ifelse(grepl("(_p|_s)", x), paste0(fofn_dir, fam_id, "_fa.fofn"), "NA"),
    family_id = fam_id
  )
  # write.table(temp_df, paste0(out_dir, x, "/man.tab"), sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
  return(temp_df)
})
names(temp1) <- paste0(out_dir, df0$V1, "/man.tab")
```


```{r}
my_lines <- readLines("C:/docs/autism/outputs.txt")
new_breaks <- grep("____________", my_lines)
list_by_sample <- lapply(seq_along(new_breaks)[-c(1)], function(x){
  if(x == 1){
    return(my_lines[1:new_breaks[x]])
  } else if(x >= (length(new_breaks)-1)){
    return(my_lines[new_breaks[x]:length(my_lines)])
  } else {
    return(my_lines[new_breaks[x]:new_breaks[x+1]])    
  }
})

df0 <- lapply(list_by_sample, function(x){
  curr_sample <- x[2]
  correct_fofn <- grepl("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/0_hifi_fofn/", x) %>% any
  no_logs <- grepl("total 0", x) %>% any
  fastas <- grepl(".hifiasm.bp.hap2.p_ctg.gfa.fasta", x) %>% any
  
  data.frame(sample = curr_sample, 
             fofn = correct_fofn,
             logs = no_logs, 
             fasta = fastas) %>% return()
}) %>% do.call(rbind, .)

df_ys <- read.table("C:/docs/autism/finished_ys", header = FALSE)
df0$ys <- df0$sample %in% df_ys$V1

df0$FINISHED <- df0$ys | (df0$fofn & df0$logs & df0$fasta)

df0$sample[!df0$FINISHED] %>% write.table("C:/docs/autism/outs_fail.txt", col.names = FALSE, row.names = FALSE, quote=FALSE)



files_all <- readLines("C:/docs/autism/all_files.txt")
df1 <- data.frame(sample = files_all %>% str_extract("(?<=./).*?(?=/)"), 
                  file = files_all)
df1 <- df1[!is.na(df1$sample), ]
df1 <- df1[df1$sample %in% df0$sample[df0$FINISHED], ]
files_to_delete <- c(".hifiasm.bp.hap1.p_ctg.gfa$",
                     ".hifiasm.bp.hap1.p_ctg.lowQ.bed$",
                     ".hifiasm.bp.hap1.p_ctg.noseq.gfa$",
                     ".hifiasm.bp.hap2.p_ctg.gfa$",
                     ".hifiasm.bp.hap2.p_ctg.lowQ.bed$",
                     ".hifiasm.bp.hap2.p_ctg.noseq.gfa$",
                     ".hifiasm.bp.p_ctg.gfa$",
                     ".hifiasm.bp.p_ctg.lowQ.bed$",
                     ".hifiasm.bp.p_ctg.noseq.gfa$",
                     ".hifiasm.bp.p_utg.gfa$",
                     ".hifiasm.bp.p_utg.lowQ.bed4",
                     ".hifiasm.bp.p_utg.noseq.gfa$",
                     ".hifiasm.bp.r_utg.gfa$",
                     ".hifiasm.bp.r_utg.lowQ.bed$",
                     ".hifiasm.bp.r_utg.noseq.gfa$",
                     ".hifiasm.dip.hap1.p_ctg.gfa$",
                     ".hifiasm.dip.hap1.p_ctg.lowQ.bed$",
                     ".hifiasm.dip.hap1.p_ctg.noseq.gfa$",
                     ".hifiasm.dip.hap2.p_ctg.gfa$",
                     ".hifiasm.dip.hap2.p_ctg.lowQ.bed$",
                     ".hifiasm.dip.hap2.p_ctg.noseq.gfa$",
                     ".hifiasm.dip.p_utg.gfa$",
                     ".hifiasm.dip.p_utg.lowQ.bed$",
                     ".hifiasm.dip.p_utg.noseq.gfa$",
                     ".hifiasm.dip.r_utg.gfa$",
                     ".hifiasm.dip.r_utg.lowQ.bed$",
                     ".hifiasm.dip.r_utg.noseq.gfa$",
                     ".hifiasm.ec.bin$",
                     ".hifiasm.ovlp.reverse.bin$",
                     ".hifiasm.ovlp.source.bin$",
                     ".hifiasm.bp.p_utg.lowQ.bed$") %>% paste(. , collapse = ")|(") %>% paste0("(", ., ")")
df1$to_delete <- lapply(df1$file, function(x){grepl(files_to_delete, x)}) %>% unlist
write.table(df1$file[df1$to_delete], "C:/docs/autism/files_to_delete.tsv", row.names = FALSE, quote=FALSE, col.names = FALSE)
```



get vcf counts per region
PAR:
chrY:10,000-2,781,479 and chrY:56,887,902-57,217,415
chrX:10,000-2,781,479 and chrX:155,701,382-156,030,895
```{r}
files_vcf <- list.files(path="/net/eichler/vol28/projects/autism_genome_assembly/nobackups/pav", pattern = "vcf$", full.names = TRUE)


df0 <- lapply(files_vcf, function(x){
  sample_id <- x %>% bname %>% str_remove(".vcf.gz$")
  n_chrY_all <- system2("bcftools", paste0("view -H ",  x, " chrY | wc -l"), stdout=TRUE)
  n_chrY_PAR1 <- system2("bcftools", paste0("view -H ",  x, " chrY:10000-2781479 | wc -l"), stdout=TRUE)
  n_chrY_PAR2 <- system2("bcftools", paste0("view -H ",  x, " chrY:56887902-57217415 | wc -l"), stdout=TRUE)
  n_chrX_all <- system2("bcftools", paste0("view -H ",  x, " chrY | wc -l"), stdout=TRUE)
  n_chrX_PAR1 <- system2("bcftools", paste0("view -H ",  x, " chrX:10000-2781479 | wc -l"), stdout=TRUE)
  n_chrX_PAR2 <- system2("bcftools", paste0("view -H ",  x, " chrX:155701382-156030895 | wc -l"), stdout=TRUE)
  
  data.frame(sample = sample_id, 
             n_chrY_all = n_chrY_all,
             n_chrY_PAR1 = n_chrY_PAR1, 
             n_chrY_PAR2 = n_chrY_PAR2,
             n_chrX_all = n_chrX_all,
             n_chrX_PAR1 = n_chrX_PAR1, 
             n_chrX_PAR2 = n_chrX_PAR2,) %>% return
}) %>% do.call(rbind, .)

write.table(df0, "vcf_counts.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)



```


get F samples
```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats") 
df0 <- df00 %>% as.data.frame()
df1 <- df0$SP_ID[df0$Sex == "F"]
write.table(df1, "C:/docs/autism/F_samples.txt", quote=FALSE, row.names = FALSE)
```


as.POSIXct(paste(df0$year, df0$month, df0$day, df0$time, sep="-"), format = "%Y-%m-%d-%H:%M")

ggplot(df1, aes(x = dt_full, y = dt_day)) + 
  geom_point() + 
  coord_cartesian(ylim = c(as.POSIXct("9:00", format = "%H:%M"), as.POSIXct("17:00", format = "%H:%M"))) + 
  scale_y_continuous(breaks = as.POSIXct(9:17 %>% paste0(":00"), format = "%H:%M"), labels = paste0(c(9:12,1:5), c(rep(" AM",4), rep(" PM", 5)))) + 
  ylab("") + 
  geom_smooth(method='lm') +
  ggtitle("daily Lipton time")

```{r}
time <- c("04:15:52", "04:20:22") %>% as.POSIXct(format="%H:%M:%S")
available <- c(2138, 2096)

df_free <- data.frame(time = time,
                      free = available)

slope <- -(max(available) - min(available)) / as.numeric(max(time) - min(time))
b <- available[1] - slope * as.numeric(time[1])
deadline <- -b / (slope)


ggplot(df_free, aes(x=time, y = free)) + 
  geom_point() + 
  geom_smooth(method='lm') +
  coord_cartesian(ylim=c(0,2300), xlim=c(as.POSIXct("04:00:00", format="%H:%M:%S"), as.POSIXct("09:00:00", format="%H:%M:%S")))

```

```{r}
myLines <- readLines("C:/docs/autism/asm_qc/files.all.20240617.txt")
myLines <- myLines[grepl("/assemblies/hifiasm/.*\\....$", myLines)]
my_samples <- myLines %>% lapply(function(x){str_split(x, "/")[[1]][2]}) %>% unlist

all_samples <- u(my_samples)
list_files_by_sample <- lapply(all_samples, function(x){
  myLines[my_samples==x]
})
names(list_files_by_sample) <- all_samples


target_samples <- all_samples[grepl("_(p|s).",  all_samples)]
bool_has_base <- lapply(target_samples, function(x){
  has_base <- grepl("hifiasm.bp.hap1", list_files_by_sample[[x]]) %>% any
  has_trio <- grepl("trio", list_files_by_sample[[x]]) %>% any
  c(x, has_base, has_trio)
}) %>% do.call(rbind, .) %>% as.data.frame()

```



p1 and s1/2 samples without trio files but with normal files


```{r}
df00 <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=0", sheet = "Individual_Stats")
df_comments <- read_sheet("https://docs.google.com/spreadsheets/d/1NYBlpsY9rizPnUMT_qE1oHg8ZGnYocCaMk7CXciZqyk/edit#gid=1859063258", sheet = "output-paths-iw") %>% as.data.frame()
df_comments[is.na(df_comments)] <- ""
df1 <- readLines("C:/docs/autism/asm_qc/files.fasta.iw.txt")
df2 <- readLines("C:/docs/autism/asm_qc/files.fasta.ys.txt")

all_samples <- df00$SP_ID
all_samples <- all_samples[!is.na(all_samples)]

df0 <- data.frame(sample = all_samples)
find_asm_output <- function(sample_id, hap){
    if(grepl("_(s|p).$", sample_id)){
    mypath <- paste0("asm_iw/", sample_id,"/", sample_id, "/assemblies/hifiasm/trio/0.16.1/", sample_id, ".hifiasm.dip.", hap,".p_ctg.gfa.fasta")
  } else {
    mypath <- paste0("asm_iw/", sample_id, "/", sample_id, "/assemblies/hifiasm/0.16.1/", sample_id, ".hifiasm.bp.", hap, ".p_ctg.gfa.fasta" )
  }
  
  if(mypath %in% df1){return(mypath)}
  
  if(grepl("_(s|p).$", sample_id)){
    mypath <- paste0("asm_ys/", sample_id, "/assemblies/hifiasm/trio/0.16.1/", sample_id, ".hifiasm.dip.", hap, ".p_ctg.gfa.fasta")
  } else {
    mypath <- paste0("asm_ys/", sample_id, "/assemblies/hifiasm/0.16.1/", sample_id, ".hifiasm.bp.", hap, ".p_ctg.gfa.fasta")
  }
  
  if(mypath %in% df2){return(mypath)}
  
  return("")
}
df0$hap1 <- lapply(df0$sample, function(x){
  find_asm_output(x, "hap1")
}) %>% unlist

df0$hap2 <- lapply(df0$sample, function(x){
  find_asm_output(x, "hap2")
}) %>% unlist

write.table(df0, "C:/docs/autism/asm_qc/asm_paths_20240617.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)


df3 <- readLines("C:/docs/autism/asm_qc/pbsv.grch38.files.txt")
df4 <- readLines("C:/docs/autism/asm_qc/pbsv.grch38_noY.files.txt")

find_pbsv_outputs <- function(sample_id, pbsv_str, line_src, pre_str){
  mypath <- paste0(pre_str, sample_id, "/pbsv_", sample_id, "_", pbsv_str,".vcf.gz")
  if(mypath %in% line_src){return(mypath)} else {return("")}
}

df0$PBSV_GRCH38_SV <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "sv", df3, "pbsv/GRCH38/results/")}) %>% unlist
df0$PBSV_GRCH38_DUP <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "dup", df3, "pbsv/GRCH38/results/")}) %>% unlist
df0$PBSV_GRCH38_BND <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "bnd", df3, "pbsv/GRCH38/results/")}) %>% unlist

df0$`PBSV_GRCH38-noY_SV` <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "sv", df4, "pbsv/GRCH38_noY/results/")}) %>% unlist
df0$`PBSV_GRCH38-noY_DUP` <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "dup", df4, "pbsv/GRCH38_noY/results/")}) %>% unlist
df0$`PBSV_GRCH38-noY_BND` <- lapply(df0$sample, function(x){find_pbsv_outputs(x, "bnd", df4, "pbsv/GRCH38_noY/results/")}) %>% unlist


df5 <- readLines("C:/docs/autism/asm_qc/files_fcs_old.txt")
df6 <- readLines("C:/docs/autism/asm_qc/files_fcs_batch01.txt")
df7 <- readLines("C:/docs/autism/asm_qc/files_fcs_batch02.txt")
find_fcs_outputs <- function(sample_id, hap){
  
  
  mypath <- paste0("assembly_qc/QC_results/contamination_screening/results/", sample_id,"_", hap,"/fasta/", sample_id, "_", hap, ".fasta")
  if(mypath %in% df6){return(mypath)}
  
  mypath <- paste0("assembly_qc_GROUP2/QC_results/contamination_screening/results/", sample_id,"_", hap,"/fasta/", sample_id, "_", hap, ".fasta")
  if(mypath %in% df7){return(mypath)}
  
  mypath <- paste0("assembly_qc/QC_results_old/contamination_screening/results/", sample_id,"_", hap,"/fasta/", sample_id, "_", hap, ".fasta")
  if(mypath %in% df5){return(mypath)}

  return("")
}
df0$asm_qc_hap1 <- lapply(df0$sample, function(x){find_fcs_outputs(x, "hap1")}) %>% unlist
df0$asm_qc_hap2 <- lapply(df0$sample, function(x){find_fcs_outputs(x, "hap2")}) %>% unlist


df8 <- read.table("C:/docs/autism/asm_qc/1_hifialn.grch38.man.tsv", header=TRUE)
df9 <- read.table("C:/docs/autism/asm_qc/1_hifialn.grch38_noY.man.tsv", header=TRUE)
df0$aln_grch38 <- lapply(df0$sample, function(x){
  if(x %in% df8$SAMPLE){
    df8$BAM[df8$SAMPLE==x] %>% str_remove("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/") %>% return  
  } else {
    return("")
  }
}) %>% unlist
df0$`aln_grch38-noY` <- lapply(df0$sample, function(x){
  if(x %in% df9$SAMPLE){
    df9$BAM[df9$SAMPLE==x] %>% str_remove("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/") %>% return  
  } else {
    return("")
  }
}) %>% unlist


df_comments <- df_comments[order(df_comments$sample), ]
df0 <- df0[order(df0$sample), ]
BOOL_CONTROL <- all(df_comments$sample == df0$sample)

if(BOOL_CONTROL){
  for(row in 1:nrow(df0)){
    for(col in 1:ncol(df0)){
      if(df0[row,col] == "" & df_comments[row,col]!="" & !grepl("/", df_comments[row,col])){
        df0[row,col] <- df_comments[row,col]
      }
    }
  }
  write.table(df0[order(df0$sample), ], "C:/docs/autism/asm_qc/autism_output_manifest_20240618.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
} else {
  print("warning new samples added!!")
}
```




```{r}
df_asmqc_01 <- read.table("C:/docs/autism/asm_qc/asmqc.batch01.man", header=TRUE)
df_asmqc_02 <- read.table("C:/docs/autism/asm_qc/asmqc.batch02.man", header=TRUE)

df_toasmqc <- df0[df0$asm_qc_hap1 == "" & df0$hap1 !="", ]

```

```{r}
my_comments <- df0$asm_qc_hap1[!grepl("/", df0$asm_qc_hap1)] %>% u
df_pav_man <- df0[!df0$asm_qc_hap1 %in% my_comments, c("sample", "asm_qc_hap1", "asm_qc_hap2")]
colnames(df_pav_man) <- c("NAME", "HAP1", "HAP2")
df_pav_man$HAP1 <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_pav_man$HAP1)
df_pav_man$HAP2 <- paste0("/net/eichler/vol28/projects/autism_genome_assembly/nobackups/", df_pav_man$HAP2)
write(df_pav_man, "C:/docs/autism/asm_qc/pav.man.refresh.20240618.tab")
```






/net/eichler/vol28/projects/autism_genome_assembly/nobackups/assembly_qc/QC_results/contamination_screening/results/12832_p1_hap1/fasta/12832_p1_hap1.fasta




software/pipelines/ss_analyses/202004
- align stranseq data to the refrence (bwa or winnowmap)
run breakpointr and/or saarclust

requests: 
- add strandphsaser tool
- add automatic library selection
- add strandseq pool genotyping



strandseq protocol
- merge bams 
  - samtools merge -@4 merged.bam {input.bam}
- call veraibale sites with bcftools
  - bcftools mpileup etc
- 
- get watson crick (WC) regions in each strandseq seq cell
  - exportRegions(datapath= {breakpointr.data})
  - trying to ask which sample differes the lest from the given cell haplotype, then that is the sample of origin for the given cell. 
  
  
will need to meet with david seperately later to go over pipeline?????????????????????????
