---
title: "jiadong"
output: html_document
date: "2025-01-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

```



```{r}
CURR_GENOME <- "t2t"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/jiadong/results/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    if(x == "chrX"){
      to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
        BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]
        BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
        BOOL_OL <- BOOL_START & BOOL_END
        return(which(BOOL_OL))
      }) %>% unlist %>% u
      # print(to_remove)
      temp_chr <- temp_chr[-to_remove, ]
    }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("_.")
df00$role <- "sample"
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- "sample"

df00$family <- df00$sample %>% str_extract("^.....(?=_)")


write.table(df00, "C:/docs/jiadong/t2t_coverage_ggplot.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/jiadong/t2t_coverage_main.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)



df02 <- df00
df02$chr <- as.factor(df02$chr)
g2 <- ggplot(df02) +
  geom_boxplot(aes(x=position, y = coverage, group=chr), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(aes(x = position, y = coverage, colour = Role, shape = HAP), width = 0.25, height = 0.004, alpha=0.3) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=1:24, labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(1,24)) + 
  ggtitle("Haplotype Assembly CHM13 Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/jiadong/coverage_t2t_window.pdf", g2, width = 15, height = 7, units = "in")
```

```{r}
df0 <- fread("C:/docs/jiadong/man.jiadong.tab") %>% as.data.frame()
df1 <- fread("C:/docs/jiadong/hgsvc_paf.fofn", header=FALSE) %>% as.data.frame()
df2 <- fread("C:/docs/jiadong/hprc_paf.fofn", header=FALSE) %>% as.data.frame()


df1$SAMPLE <- df1$V1 %>% bname %>% str_remove(".paf$") %>% str_replace("_", "_hap") %>% paste0(., "_HGSVC")
df1$ROLE <- df1$SAMPLE %>% str_extract("hap.")
df1$FAMILY <- df1$SAMPLE %>% str_remove("_hap.") 
df1$HAP1 <- df1$V1
df1$HAP2 <- df1$V1
df1$V1 <- NULL
df1$cohort <- "HGSVC"

df2$SAMPLE <- df2$V1 %>% bname %>% str_remove(".paf$") %>% str_replace("_", "_hap") %>% paste0(., "_HPRC")
df2$ROLE <- df2$SAMPLE %>% str_extract("hap.")
df2$FAMILY <- df2$SAMPLE %>% str_remove("_hap.")
df2$HAP1 <- df2$V1
df2$HAP2 <- df2$V1
df2$V1 <- NULL
df2$cohort <- "HPRC"

df0$cohort <- "ONT"


df_final <- rbind(df2, rbind(df1, df0))
df_colors <- data.frame(color = "nan", family = df_final$FAMILY, cohort = df_final$cohort)
df_colors$color[df_colors$cohort=="HGSVC"] <- "xkcd:green"
df_colors$color[df_colors$cohort=="HPRC"] <- "xkcd:blue"
df_colors$color[df_colors$cohort=="ONT"] <- "xkcd:orange"

df_colors$cohort <- NULL
df_final$cohort <- NULL
df_final <- df_final[order(df_final$FAMILY, df_final$ROLE), ]

write.table(df_final, "C:/docs/jiadong/manifest.all.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
write.table(df_colors, "C:/docs/jiadong/colors.all.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```


```{r}
CURR_GENOME <- "t2t"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/jiadong/coverage_df/results/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    if(x == "chrX"){
      to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
        BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]
        BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
        BOOL_OL <- BOOL_START & BOOL_END
        return(which(BOOL_OL))
      }) %>% unlist %>% u
      # print(to_remove)
      temp_chr <- temp_chr[-to_remove, ]
    }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- "sample"
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("HGSVC", df00$sample), "HGSVC", 
                    ifelse(grepl("HPRC", df00$sample), 
                           "HPRC", 
                           "ONT"))

write.table(df00, "C:/docs/jiadong/coverage_df/t2t_coverage_ggplot.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
write.table(df_print, "C:/docs/jiadong/coverage_df/t2t_coverage_main.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)



df02 <- df00
df02$chr <- as.factor(df02$chr)
STEP <- 0.4
df02$group_position <- df02$position*2 + ifelse(df02$Role=="HPRC", -STEP, ifelse(df02$Role=="HGSVC", 0, STEP))
df02 <- df02[sample(seq_down(df02)), ]
df02$cohort_group <- paste0(df02$chr, "-", df02$Role)
g2 <- ggplot(df02) +
  geom_jitter(aes(x = group_position, y = coverage, colour = Role, shape = HAP), width = 0.1, height = 0.004, alpha=0.8) +
  geom_boxplot(aes(x=group_position, y = coverage, group=cohort_group), alpha = 0.5, outlier.shape = NA) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=2*(1:24), labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(2.7,47.5)) + 
  ggtitle("Haplotype Assembly CHM13 Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
ggsave("C:/docs/jiadong/coverage_df/coverage_t2t_window_shift_with_box.pdf", g2, width = 18, height = 7, units = "in")
```



```{r}
df0 <- fread("C:/docs/jiadong/UW_ONT_paf.fofn", header=FALSE) %>% as.data.frame()
df0$SAMPLE <- df0$V1 %>% bname %>% str_replace("_", "_hap") %>% str_replace(".paf", "_ONT")
df0$ROLE <- df0$SAMPLE %>% str_extract("hap.")
df0$FAMILY <- df0$SAMPLE %>% str_extract("^.*?(?=_)")
df0$HAP1 <- df0$V1
df0$HAP2 <- df0$V1
df0$V1 <- NULL
df0$COHORT <- "ONT"

df1 <- fread("C:/docs/jiadong/hprc_paf.fofn", header=FALSE) %>% as.data.frame()
df1$SAMPLE <- df1$V1 %>% bname %>% str_replace("_", "_hap") %>% str_replace(".paf", "_HPRC")
df1$ROLE <- df1$SAMPLE %>% str_extract("hap.")
df1$FAMILY <- df1$SAMPLE %>% str_extract("^.*?(?=_)")
df1$HAP1 <- df1$V1
df1$HAP2 <- df1$V1
df1$V1 <- NULL
df1$COHORT <- "HPRC"

df2 <- fread("C:/docs/jiadong/hgsvc_paf.fofn", header=FALSE) %>% as.data.frame()
df2$SAMPLE <- df2$V1 %>% bname %>% str_replace("_", "_hap") %>% str_replace(".paf", "_HGSVC")
df2$ROLE <- df2$SAMPLE %>% str_extract("hap.")
df2$FAMILY <- df2$SAMPLE %>% str_extract("^.*?(?=_)")
df2$HAP1 <- df2$V1
df2$HAP2 <- df2$V1
df2$V1 <- NULL
df2$COHORT <- "HGSVC"

df3 <- fread("C:/docs/jiadong/target_samples.txt", header=FALSE, sep="\t") %>% as.data.frame()
df3$COHORT <- df3$V2 %>% str_remove(",.*")

df4 <- rbind(df0[df0$FAMILY %in% df3$V1, ], 
             df1[df1$FAMILY %in% df3$V1[df3$COHORT=="HPRC"], ], 
             df2[df2$FAMILY %in% df3$V1[df3$COHORT == "HGSVC"], ])
df4$FAMILY <- df4$FAMILY %>% paste0("_", df4$COHORT)

my_colors <-  c("xkcd:blue", "xkcd:orange", "xkcd:green")
my_cohorts <- c("HPRC", "ONT", "HGSVC")
names(my_colors) <- my_cohorts


df5 <- data.frame(color = my_colors[df4$COHORT], 
                  family = df4$FAMILY)

write.table(df4, "C:/docs/jiadong/subset_figure.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
write.table(df5, "C:/docs/jiadong/subset_colors.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)

```

