---
title: "locityper"
output: html_document
date: "2024-05-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
MY_GENE_REGEX <- "cytochrome P450 2D6 \\(CYP2D6\\) gene"
files <- list.files("C:/docs/locityper/results/", pattern=".blast.table$", recursive=TRUE, full.names=TRUE)
bool_match <- lapply(files, function(x){
  temp1 <- readLines(x)
  temp2 <- temp1[grep("^>", temp1)]
  # return(temp2 )
  grepl(MY_GENE_REGEX, temp1) %>% any %>% return
}) %>% unlist

df1 <- data.frame(
  sample = files %>% str_extract("(?<=results/).*?(?=/)"),
  CYP2D6_match = bool_match
)
```

```{r}
MY_GENE_REGEX <- "(Homo sapiens PMS1 homolog 2)|(PMS2)"
files <- list.files("C:/docs/locityper/results/", pattern=".blast.table$", recursive=TRUE, full.names=TRUE)
bool_match <- lapply(files, function(x){
  temp1 <- readLines(x)
  grepl(MY_GENE_REGEX, temp1) %>% any %>% return
}) %>% unlist

df1 <- data.frame(
  sample = files %>% str_extract("(?<=results/).*?(?=/)"),
  PMS2_match = bool_match
)
```

```{r}
library(jsonlite)
files <- list.files("C:/docs/locityper/CYP2D6/", pattern = "res.json.gz", full.names=TRUE, recursive = TRUE)

list_jsons <- lapply(files, function(x){
  read_json(x) %>% return
})

df0 <- lapply(seq_along(list_jsons), function(x){
  temp_sample <- files[x] %>% str_extract("(?<=sample_set/).*?(?=.CYP2D6)")
  gt <- list_jsons[[x]]$genotype
  
  which_match <- lapply(list_jsons[[x]]$options, function(x){
    x$genotype == gt
  }) %>% unlist %>% which
  
  data.frame(sample=temp_sample,
             genotype = gt,
             lik_mean = list_jsons[[x]]$options[[which_match]]$lik_mean, 
             lik_sd = list_jsons[[x]]$options[[which_match]]$lik_sd, 
             prob = list_jsons[[x]]$options[[which_match]]$prob, 
             log10_prob = list_jsons[[x]]$options[[which_match]]$log10_prob, 
             dist_to_primary = list_jsons[[x]]$options[[which_match]]$dist_to_primary)
  
}) %>% do.call(rbind, .)

```

```{r}
df0 <- fread("C:/docs/locityper/muc5ac_haplogroups_manifest.txt") %>% as.data.frame()
df1 <- fread("C:/docs/aou/locityper.MUC5AC.tsv") %>% as.data.frame()


all_genotypes <- df1$genotype %>% str_split(",") %>% unlist %>% u
  
df0$is_avail <- df0$haplotype %in% all_genotypes

missing_samples <- df0$haplotype[df0$is_avail == FALSE]
write.table(missing_samples, "C:/docs/aou/missing_haplotypes.txt", quote=FALSE, row.names = FALSE, col.names = FALSE)
```

