---
title: "smaht"
output: html_document
date: "2025-07-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(xlsx)
```



```{r}
df00 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "TissueSample")
df01 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "(Tissue)")
df02 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "Analyte")
df03 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "AnalytePreparation")
df04 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "PreparationKit")
df05 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "Treatment")
df06 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "Library")
df07 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "LibraryPreparation")
df08 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "Sequencing")
df09 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "FileSet")
df10 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "UnalignedReads")
df11 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "Software")
df12 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "AlignedReads")
df13 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "VariantCalls")
df14 <- read.xlsx2("C:/docs/smaht/smht020/SMHT020 - ONT Metadata.xlsx", sheetName = "SupplementaryFile")
df_ont_data <- read_sheet("https://docs.google.com/spreadsheets/d/1hK770z1egV_wvnzGAVmBE2jyLzHjYd_YJFTWs9bb1Jo/edit?gid=0#gid=0", sheet="Per-Run Data") %>% as.data.frame()
fout <- function(this_data_frame, path="C:/docs/smaht/smht020/new.tsv"){
  write.table(this_data_frame, "C:/docs/smaht/smht020/new.tsv", col.names = FALSE, row.names = FALSE, quote=FALSE)
}
```


```{r}
df_temp <- data.frame(description = df00$submitted_id %>% str_extract("PS[^_]+"))
write.table(df_temp, "C:/docs/smaht/smht020/new.tsv", col.names = FALSE, row.names = FALSE, quote=FALSE)
```

```{r}
df1 <- fread("C:/docs/smaht/smht020/manifest.tsv") %>% as.data.frame()
df1$NAME <- df1$INPUT %>% bname
df1$NAME2 <- df1$NAME %>% toupper() %>% str_replace("_PASS.BAM$", "_BAM") %>% str_replace("_PASS.FASTQ.GZ$", "_FASTQ")
df1$NAME3 <- paste0("UWSC_UNALIGNED-READS_FILTERED_", df1$NAME2)
write.table(df1$NAME3, "C:/docs/smaht/smht020/new.tsv", col.names = FALSE, row.names = FALSE, quote=FALSE)
write.table(df1$NAME, "C:/docs/smaht/smht020/new.tsv", col.names = FALSE, row.names = FALSE, quote=FALSE)

df2 <- fread("C:/docs/smaht/smht020/md5_summary.txt", header=FALSE) %>% as.data.frame()
df1$md5 <- df1$INPUT %>% lapply(function(x){
  df2$V1[df2$V2==x]
}) %>% unlist
fout(df1$md5)

df1$flow_cell <- df1$NAME %>% str_extract("(?<=_.._)[^_]+(?=_........_)")
fout(df1$flow_cell)

df3 <- fread("C:/docs/smaht/smht020/SMHT020.qc_summary.tsv") %>% as.data.frame()
df3 <- df3[df3$FILTER=="filt", ]
df1$n50 <- lapply(df1$NAME, function(x){
  which_runid <- lapply(df3$RUNID, function(y){
    grepl(y, x)
  }) %>% unlist %>% which
  df3$N50_K[which_runid] * 1000
}) %>% unlist
fout(df1$n50)

df1$file_format <- ifelse(grepl("bam$", df1$INPUT), "bam", "fastq_gz")
fout(df1$file_format)

df09$PSID <- df09$submitted_id %>% str_extract("PS[^-_]+(?=_)")
df1$PSID <- df1$NAME %>% str_extract("PS[^-_]+(?=_)")
df1$file_sets <- df1$PSID %>% lapply(function(x){
  if(x %in% df09$PSID){
    return(df09$submitted_id[df09$PSID==x])  
  }
  return("")
}) %>% unlist
fout(df1$file_sets)

df1$derived_from <- ifelse(grepl("bam", df1$NAME), "", df1$NAME3) %>% str_replace("_FASTQ$", "_BAM")
fout(df1$derived_from)

df1$software <- ifelse(df1$file_format=="bam", "UWSC_SOFTWARE_DORADO_102_NVIDIA_A100", "UWSC_SOFTWARE_DORADO_102_NVIDIA_A100|UWSC_SOFTWARE_SAMTOOLS_V1.19.2|UWSC_SOFTWARE_BGZIP_V1.16
")
fout(df1$software)

df_temp <- fread("C:/docs/smaht/smht020/new.tsv") %>% as.data.frame()
df_temp$val <- ifelse(df_temp$file_format=="bam", "UWSC_SOFTWARE_DORADO_102_NVIDIA_A100", "UWSC_SOFTWARE_DORADO_102_NVIDIA_A100|UWSC_SOFTWARE_SAMTOOLS_V1.19.2|UWSC_SOFTWARE_BGZIP_V1.16")
fout(df_temp$val)
```





```{r}
df_temp <- read.xlsx2("C:/docs/smaht/smht020/SMHT020_test.xlsx", sheetName = "(Tissue)")
df_temp$external_id <- 1:39
write.xlsx2(df_temp, "C:/docs/smaht/smht020/SMHT020_test.xlsx", sheetName = "(Tissue)", append = TRUE)

df_temp1 <- read.table("C:/docs/smaht/smht020/new.tsv")
df_temp1$V2 <- 1:6
write.table(df_temp1, "C:/docs/smaht/smht020/new.tsv")
```




ls -lh | awk9 | grep SDs.bed$  | xargs -I {} /bin/bash -c "echo {}; bedtools sort -i {} | bedtools merge > {}.sort.merge"
```{r}
files <- list.files("C:/docs/smaht/ShareSd/", full.names = TRUE, recursive = TRUE, pattern = ".SDs.bed.sort.merge")
samples <- files %>% bname %>% str_remove(".hap.*") %>% u


df00 <- data.frame(SAMPLE = samples,
                   hap1 = lapply(samples, function(x){
                     curr_bed_file <- fread(files[grepl("hap1", files) & grepl(x, files)], header=FALSE) %>% as.data.frame()
                     curr_bed_file$length <- curr_bed_file$V3 - curr_bed_file$V2
                     sum(curr_bed_file$length) %>% return()
                   }) %>% unlist,
                   hap2 = lapply(samples, function(x){
                     curr_bed_file <- fread(files[grepl("hap2", files) & grepl(x, files)], header=FALSE) %>% as.data.frame()
                     curr_bed_file$length <- curr_bed_file$V3 - curr_bed_file$V2
                     sum(curr_bed_file$length) %>% return()
                   }) %>% unlist
                   )

df00$hap1_mb <- df00$hap1 / 1000000
df00$hap2_mb <- df00$hap2 / 1000000

df00$hap1 <- NULL
df00$hap2 <- NULL

```


function to split peri, acro, and remaining SDs

```{r}
files <- list.files("C:/docs/smaht/ShareSd/", full.names = TRUE, recursive = TRUE, pattern = ".SDs.bed$")

for(file in files){
  curr_path <- file
  df00 <- fread(curr_path) %>% as.data.frame()
  df00$is_acro <- df00$acro==1 | df00$acro2==1 & df00$peri!=1 & df00$peri2!=1
  df00$is_peri <- df00$peri==1 | df00$peri2==1 & df00$acro!=1 & df00$acro2!=1
  df00$is_remaining <- df00$is_acro==FALSE & df00$is_peri==FALSE
  curr_sample <- curr_path %>% bname %>% str_remove(".contig.SDs.bed")
  write.table(df00[df00$is_acro, ], paste0("C:/docs/smaht/ShareSd/", curr_sample, ".acro.bed"), col.names=TRUE, row.names = FALSE, quote=FALSE, sep="\t")
  write.table(df00[df00$is_peri, ], paste0("C:/docs/smaht/ShareSd/", curr_sample, ".peri.bed"), col.names=TRUE, row.names = FALSE, quote=FALSE, sep="\t")
  write.table(df00[df00$is_remaining, ], paste0("C:/docs/smaht/ShareSd/", curr_sample, ".remaining.bed"), col.names=TRUE, row.names = FALSE, quote=FALSE, sep="\t")  
}

# ls -lh | awk9 | grep -P "(acro|peri|remaining).bed$" |  xargs -I {} /bin/bash -c "echo {}; bedtools sort -i {} | bedtools merge > {}.sort.merge"

files <- list.files("C:/docs/smaht/ShareSd/", full.names = TRUE, recursive = TRUE, pattern = "(acro|peri|remaining).bed.sort.merge")

df01 <- lapply(files, function(x){
  df_temp <- fread(x, header=FALSE) %>% as.data.frame()
  df_temp$length <- df_temp$V3 - df_temp$V2
  class <- x %>% str_extract("acro|peri|remaining") %>% str_to_title()
  df_rval <- data.frame(SAMPLE=bname(x) %>% str_extract("^.*(?=.hap)"),
                        HAP = str_extract(x, "hap."),
                        COHORT = "SMAHT", 
                        CHR = class,
                        length = df_temp$length %>% sum)
  return(df_rval)
}) %>% do.call(rbind, .)

SAMPLES <- u(df01$SAMPLE)
HAPS <- u(df01$HAP)
df02 <- expand.grid(SAMPLES, HAPS) %>% as.data.frame()
df02 <- df02[order(df02$Var1), ]
df02$total <- 0
for(sample in SAMPLES){
  for(hap in HAPS){
    df02$total[df02$Var1==sample & df02$Var2==hap] <- df01$length[df01$SAMPLE==sample & df01$HAP==hap] %>% sum  
  }
}


```

