---
title: "public test"
output: html_document
date: "2025-04-16"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
source("C:/docs/src/R/my_libraries.R")
```







```{r}
CURR_GENOME <- "t2t"
DF_EXCLUDE <- fread("C:/docs/autism/coverage/exclude_regions_chrX.tsv.txt") %>% as.data.frame()
DF_EXCLUDE <- DF_EXCLUDE[DF_EXCLUDE$genome == CURR_GENOME, ]

files <- list.files("C:/docs/jiadong/coverage_df/results/", recursive = TRUE, full.names = TRUE, pattern = ".tsv")
CHR_ORDER <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22", "chrX", "chrY")
COLNAMES <- c("chr", "start", "end", "n_features", "n_bases", "length", "percent")

df00 <- lapply(files, function(x){
  df_temp <- fread(x) %>% as.data.frame()
  colnames(df_temp) <- COLNAMES
  
  temp_by_chr <- lapply(df_temp$chr %>% u, function(x){
    temp_chr <- df_temp[df_temp$chr == x, ]
    if(x == "chrX"){
      to_remove <- lapply(seq_down(DF_EXCLUDE), function(y){
        BOOL_START <- temp_chr$start <= DF_EXCLUDE$end[y]
        BOOL_END <- temp_chr$end >= DF_EXCLUDE$start[y]
        BOOL_OL <- BOOL_START & BOOL_END
        return(which(BOOL_OL))
      }) %>% unlist %>% u
      # print(to_remove)
      temp_chr <- temp_chr[-to_remove, ]
    }
    pass_percent <- temp_chr$percent >= 0.95
    pass_bins <- temp_chr$n_features <= 3
    data.frame(chr = x,
               coverage = sum(pass_percent & pass_bins)/nrow(temp_chr))
  }) %>% do.call(rbind, .)
  
  temp_by_chr$SOURCE <- x
  return(temp_by_chr)
}) %>% do.call(rbind, .)
df00$sample <- df00$SOURCE %>% bname %>% str_remove(".tsv")
df00$HAP <- df00$sample %>% str_extract("hap.")
df00$role <- "sample"
df00$SOURCE <- NULL
df00 <- df00[!grepl("_|M", df00$chr), ]
df00$position <- ifelse(grepl("chrX", df00$chr), 23, ifelse(grepl("chrY", df00$chr), 24, str_remove(df00$chr, "chr"))) %>% as.numeric()
df00$Role <- ifelse(grepl("HGSVC", df00$sample), "HGSVC", 
                    ifelse(grepl("HPRC", df00$sample), 
                           "HPRC", 
                           "ONT"))

write.table(df00, "C:/docs/jiadong/coverage_df/t2t_coverage_ggplot.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
sampleToChr <- HashMap$new()
temp1 <- lapply(seq_down(df00), function(x){
  curr_sample <- df00$sample[x]
  if(sampleToChr$containsKey(curr_sample)==FALSE){sampleToChr$put(curr_sample, HashMap$new())}
  curr_chr <- df00$chr[x]
  curr_cov <- df00$coverage[x]
  sampleToChr$get(curr_sample)$put(curr_chr, curr_cov)
})
df_print <- lapply(sampleToChr$keySet(), function(curr_sample){
  curr_row <- lapply(CHR_ORDER, function(curr_chr){
    sampleToChr$get(curr_sample)$get(curr_chr)
  }) %>% unlist
  names(curr_row) <- CHR_ORDER
  rval <- c(curr_sample, curr_row)
  names(rval)[1] <- "sample"
  return(rval)
}) %>% do.call(rbind, .) %>% as.data.frame()
# write.table(df_print, "C:/docs/jiadong/coverage_df/t2t_coverage_main.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)



df02 <- df00
df02$chr <- as.factor(df02$chr)
STEP <- 0.4
df02$group_position <- df02$position*2 + ifelse(df02$Role=="HPRC", -STEP, ifelse(df02$Role=="HGSVC", 0, STEP))
df02 <- df02[sample(seq_down(df02)), ]
df02$cohort_group <- paste0(df02$chr, "-", df02$Role)
g2 <- ggplot(df02) +
  geom_jitter(aes(x = group_position, y = coverage, colour = Role, shape = HAP), width = 0.1, height = 0.004, alpha=0.8) +
  geom_boxplot(aes(x=group_position, y = coverage, group=cohort_group), alpha = 0.5, outlier.shape = NA) +
  scale_y_continuous(breaks=seq(0,1,.2), labels = seq(0,100,20)) + 
  scale_x_continuous(breaks=2*(1:24), labels = CHR_ORDER) + 
  ylab("Percent Contig Coverage") + 
  xlab("Chromosome") +
  coord_cartesian(ylim=c(0,1),xlim=c(2.7,47.5)) + 
  ggtitle("Haplotype Assembly CHM13 Coverage by Chromosome") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) ; g2
# ggsave("C:/docs/jiadong/coverage_df/coverage_t2t_window_shift_with_box.pdf", g2, width = 18, height = 7, units = "in")

print(g2)
```


Issue: memory will be a problem, better to chunk when possible then print chunks to a file

```{r}
getExtension <- function(file) strsplit(file, ".", fixed=T)[[1]][-1]
# df0 <- fread("C:/docs/website/project_autism_genome_assembly.tsv") %>% as.data.frame()
df0 <- fread("C:/docs/website/project_long_read_archive.tsv") %>% as.data.frame()
# df0$folders <- df0$V1 %>% dirname %>% str_split("/")
df1 <- df0[df0$V6=="pod5" | df0$V6=="fast5", ]
df1$V1 <- df1$V1 %>%  str_remove("^/net/eichler/vol28/projects/long_read_archive")

t2 <- Sys.time()
bool_gb <- grepl("GB", df1$V5)
bool_mb <- grepl("MB", df1$V5)
bool_tb <- grepl("TB", df1$V5)
df1$size_GB <- 0
df1$size_GB[bool_gb] <- as.numeric(str_remove(df1$V5[bool_gb], " GB"))
df1$size_GB[bool_mb] <- as.numeric(str_remove(df1$V5[bool_mb], " MB"))/1000
df1$size_GB[bool_tb] <- as.numeric(str_remove(df1$V5[bool_tb], " TB"))*1000
print(Sys.time() - t2)


subfolderize <- function(x){
  folders <- x %>% str_split("/") %>% unlist %>% grepv(".", .)
  folders <- folders[1:(length(folders)-1)]
  return(folders)  
}



t3 <- Sys.time()
dirToSize <- HashMap$new()
for(index in seq_down(df1)){
  curr_dir <- df1$V1[index]
  curr_size <- df1$size_GB[index]
  if(is.na(curr_dir)){next}
  dirs <- lapply(seq(1, str_count(curr_dir, "/")-1), function(x){
    str_extract(curr_dir, paste0("(/[^/]+){",x ,"}"))
  }) %>% unlist
  for(dir in dirs){
    if(!dirToSize$containsKey(dir)){
      dirToSize$put(dir, 0)
    } 
    dirToSize$put(dir, dirToSize$get(dir) + curr_size)
  }
}
print(Sys.time() - t3)

temp1 <- dirToSize$toString() 
df2 <- data.frame(dir = names(temp1),
                  size_gb = unlist(temp1))
df2 <- df2[order(df2$size_gb, decreasing = TRUE), ]
rownames(df2) <- seq_down(df2)


# fwrite(df2, "C:/docs/website/project_long_read_archive_dir.tsv", col.names = TRUE, row.names=FALSE, quote=FALSE, sep="\t")


t4 <- Sys.time()
THRESHOLD_FRACTION <- 0.7
TO_REMOVE_HASH <- HashMap$new()
for(dir in df2$dir){
  TO_REMOVE_HASH$put(dir, "FALSE")
}
temp1 <- lapply(df2$dir, function(curr_dir){
  curr_size <- dirToSize$get(curr_dir)
  dirs <- lapply(seq(1, str_count(curr_dir, "/")), function(x){
    str_extract(curr_dir, paste0("(/[^/]+){",x ,"}"))
  }) %>% unlist
  for(dir in dirs){
    if(curr_dir!=dir & (curr_size >= THRESHOLD_FRACTION * dirToSize$get(dir))){
      TO_REMOVE_HASH$put(dir, "TRUE")
    }
  }
})
print(Sys.time() - t4)


temp2 <- TO_REMOVE_HASH$toString() 
df3 <- data.frame(dir = names(temp2),
                  toRemove = unlist(temp2) %>% as.logical(), 
                  stringsAsFactors = FALSE)
df3 <- df3[order(df3$dir), ]
rownames(df3) <- seq_down(df3)
df3$size <- lapply(df3$dir, function(x){dirToSize$get(x)}) %>% unlist

df4 <- df3[df3$toRemove==FALSE, ]


df01 <- fread("C:/docs/website/ont_basecall_clinical.tsv") %>% as.data.frame()
df02 <- fread("C:/docs/website/ont_basecall_pop.tsv") %>% as.data.frame()
df03 <- fread("C:/docs/website/ont_basecall_nhp.tsv") %>% as.data.frame()

df04 <- rbind(df01, df02, df03)

t5 <- Sys.time()
dirToTypeIndex <- HashMap$new()
df04$dir_match <- lapply(seq_down(df04), function(x){
  curr_sample <- df04$SAMPLE[x]
  curr_RUNID <- df04$RUN_ID[x]
  bool_match <- grepl(curr_sample, df2$dir) & grepl(curr_RUNID, df2$dir)
  rvals <- df2$dir[which(bool_match)]
  if(any(bool_match)){
    for(rval in rvals){
      dirToTypeIndex$put(rval, x)
    }
    return(rvals)
  } else {
    return("")
  }
})
print(Sys.time() - t5)

df4$SAMPLE <- lapply(df4$dir, function(x){
  if(dirToTypeIndex$containsKey(x)){
    df04$SAMPLE[dirToTypeIndex$get(x)]
  } else {""}
}) %>% unlist

df4$SEQ_TYPE <- lapply(df4$dir, function(x){
  if(dirToTypeIndex$containsKey(x)){
    df04$SEQ_TYPE[dirToTypeIndex$get(x)]
  } else {""}
}) %>% unlist

df4$PROFILE <- lapply(df4$dir, function(x){
  if(dirToTypeIndex$containsKey(x)){
    df04$PROFILE[dirToTypeIndex$get(x)]
  } else {""}
}) %>% unlist

df4$RUN_ID <- lapply(df4$dir, function(x){
  if(dirToTypeIndex$containsKey(x)){
    df04$RUN_ID[dirToTypeIndex$get(x)]
  } else {""}
}) %>% unlist

df4$toRemove <- NULL

df4 <- df4[order(df4$size, decreasing = TRUE), ]
# write.table(df4, "C:/docs/website/project_long_read_archive_dir_matched.tsv", sep="\t", col.names = TRUE, row.names = FALSE, quote=FALSE)
```




```{r}
library(ggrepel)
t8 <- Sys.time()
logLines <- readLines("C:/docs/website/df_h_project.logs.txt")
date_indexes <- !grepl("\t", logLines)
date_indexes_which <- date_indexes %>% which
write.table(logLines[!date_indexes], "C:/docs/website/df_h_temp.txt", row.names = FALSE, col.names = FALSE, quote=FALSE, sep=" ")
# df00 <- logLines[!date_indexes] %>% str_split("\t") %>% do.call(rbind, .) %>% as.data.frame()
df00 <- fread("C:/docs/website/df_h_temp.txt", header=FALSE) %>% as.data.frame()
df00$date <- lapply(1:(length(date_indexes_which)), function(x){
  start <- date_indexes_which[x]+1
  if(x != length(date_indexes_which)){
    end <- date_indexes_which[x+1]-1  
  } else {
    end <- length(logLines)
  }
  date <- logLines[date_indexes_which[x]] %>% as.POSIXct(format="%a %b %d %H:%M:%S") # not perfect
  rval <- rep(date, length(start:end))
  return(rval)
}) %>% unlist
print(Sys.time() - t8)
df00$percent <- df00$V5 %>% str_remove("%") %>% as.numeric()
df00$label <- ifelse(df00$date == max(df00$date), as.character(df00$V1), NA_character_)
df00$date <- df00$date %>% as.POSIXct(., origin="1970-01-01")
```



```{r}
g1 <- ggplot(df00, aes(x=date, y=percent, group=V1, color=V1)) + 
  geom_line() +
  coord_cartesian(xlim = c(min(df00$date), max(df00$date)+140000), ylim = c(0, 120)) +
  geom_label_repel(aes(label = label),nudge_x = 140000, hjust = "right" ,na.rm = TRUE, box.padding = 0.3, max.overlaps=200) + 
  theme(legend.position="none") 
g1
ggsave("C:/docs/website/project_size_tracking.pdf", g1, width = 22, height = 30, units = "in", limitsize = FALSE)
```


```{r}
df00$curr_size_ <- df00$V2 %>% str_remove("T")

df1$size_GB[bool_gb] <- as.numeric(str_remove(df1$V5[bool_gb], " GB"))
df1$size_GB[bool_mb] <- as.numeric(str_remove(df1$V5[bool_mb], " MB"))/1000
df1$size_GB[bool_tb] <- as.numeric(str_remove(df1$V5[bool_tb], " TB"))

```

```{r}
temp1 <- c(100, 90, 80, 70, 70, 200, 90, 90, 90, 100, 75, 360, 90, 100, 200, 75, 105)
```


33233333

# Faster method
remove all date lines
str_split entire vector at once
add in date lines as column, using rep(., each=NNN)
  - NNN = dynamically calculated values?
  
  




TODO: also check nobackups/backups files


